{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ed244934-4bae-4026-9139-e46033e5c60b",
   "metadata": {},
   "source": [
    "# AB compartment analysis\n",
    "#### HiCExplorer hicPCA was unsatisfactory prob due to the fact that the eigenvector is calculated genome-wide\n",
    "\n",
    "## PART I: CALLING COMPARTMENTS WITH FAN-C\n",
    "#### Correlation matrix\n",
    "\n",
    "Compartments are derived from a correlation matrix, in which each entry i, j corresponds to the Pearson correlation between row i and column j of the (Hi-C) matrix.\n",
    "\n",
    "The fanc compartments command can produce a correlation matrix (AB compartment) object from a FAN-C matrix file. By default, compartment matrices are calculated on a per-chromosome basis, since each chromosome might be normalised differently.\n",
    "\n",
    "#### AB Eigenvector\n",
    "\n",
    "The eigenvector of the correlation matrix is used to derive compartment type and strength for each matrix bin. Generally, regions with positive values are assigned the ‘A’, regions with negative values the ‘B’ compartment. \n",
    "fanc compartments outputs the first eigenvector by default. \n",
    "\n",
    "Specifically, you can supply a FASTA file with the genomic sequence to fanc compartments using the -g <fasta_file> argument. This is generally a good idea: fanc compartments then calculates the average GC content of regions with positive and those with negative eigenvector entries. As GC content has previously been shown to correlate well with compartmentalisation, the eigenvector is oriented in such a way that negative entries correspond to ‘B’ (low GC content) and positive entries to ‘A’ (high GC content).\n",
    "\n",
    "#### AB domains\n",
    "\n",
    "Consecutive matrix bins with the same eigenvector sign are considered part of a “domain”. You can use the -d <domain_file> option to write the AB domains to a BED file.\n",
    "Again, it is recommended to supply the genome for a better assignment of ‘A’ and ‘B’ to each domain. \n",
    "\n",
    "### What I did\n",
    "Call compartments in Azucena genome (az_mrg) at resolutions 10/80/160/320/640/1280kb (fanc_pc.sh)\n",
    "\n",
    "Plot corr matrices, eigenvectors individually for each chromosome (fancplot.sh)\n",
    "\n",
    "AB compartment coords are in BED format\n",
    "\n",
    "Then compare the sign of the eigenvector for the chr01 with the resluts from The Plant Journal (2018) 94, 1141–1156 (Dong et al., 2018).\n",
    "\n",
    "I will proceed with resolution 80kb, 160kb, 640kb, 1280kb to validation\n",
    "\n",
    "Compartment coordinates are in fanc_Xkb.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dd16639-e5e0-4ae6-a76a-210b5116bda2",
   "metadata": {},
   "source": [
    "# PART II: VALIDATING COMPARTMENT CALLS\n",
    "#### Comparing relative enrichment of gene and transposable element (TE) density, H3K4me3, H3K18ac, H3K27me3, H3K27ac, DNA Methylation, GC content, gene expression (PRO-seq) in A/B compartments\n",
    "\n",
    "The chromosome-wide method, which have been successfully applied to the smaller plant genomes such as Arabidopsis and rice to identify global compartments.\n",
    "\n",
    "Regions classified as local A compartments are euchromatin regions with higher gene density and active epigenetic marks, while those grouped as local B compartments are heterochromatins with higher TE density and repressive marks. A compartments are associated with active chromatin marks as well as polycomb mark H3K27me3, while B compartments display heterochromatin features such as low gene density and high DNA methylation level."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac311cd-5747-4ead-9aaf-2c01f2d67212",
   "metadata": {},
   "source": [
    "We used expression data and showed that the median expression values per A compartment were higher than for B compartment for 160kb res\n",
    "\n",
    "I will proceed with 160kb resolution\n",
    "\n",
    "A and B compartment coordinates are in A.bed and B.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bf59641-4c00-4e1e-a73d-2d6e13801c79",
   "metadata": {},
   "source": [
    "# PART III: TAD/nonTAD bodies distribution between A/B compartments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1d54706-d5b4-4912-9ae6-d903750fda40",
   "metadata": {},
   "source": [
    "For this we chose TAD and nonTAD domains with at least five genes expressed, giving us a total of 1,247 domains (678 TAD domains and 569 nonTAD domains)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316fd324-e1e7-4580-8890-028c3f3b4f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "First, we analyzed the how many TADs were found within A/B compartment \n",
    "#All TADs: az_TADs_5kb_two_tools_80.bed 1207 TADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f94ed524-fae3-45df-825f-92e8be80a752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of regions: 89326207\n"
     ]
    }
   ],
   "source": [
    "# Total length of all TADs\n",
    "total_length = 0\n",
    "with open('az_TADs_5kb_two_tools_80.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end,e,r,t,y,u,i = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of regions:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8ceea046-0aff-44f9-ace0-c9e46741a677",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of A compartment: 127445834\n"
     ]
    }
   ],
   "source": [
    "#The total length of the A compartment in Azucena:\n",
    "total_length = 0\n",
    "with open('A.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of A compartment:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b4ed249b-6fb2-430b-84e6-17ef02a43250",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of B compartment: 252181719\n"
     ]
    }
   ],
   "source": [
    "#The total length of the B compartment in Azucena:\n",
    "total_length = 0\n",
    "with open('B.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of B compartment:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9f9f2df6-b853-4cb2-a1c3-4abb6019b9f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total genome length: 379627553\n"
     ]
    }
   ],
   "source": [
    "#Total Azucena genome length:\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Initialize a variable to store the total length\n",
    "total_length = 0\n",
    "\n",
    "# Open and read the FASTA file\n",
    "with open('azucena.fna', 'r') as fasta_file:\n",
    "    for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        # Add the length of each sequence to the total length\n",
    "        total_length += len(record.seq)\n",
    "\n",
    "print(\"Total genome length:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "728c4a88-d745-454a-98c4-565f14427101",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TADs in A compartment: 406\n",
      "Number of TADs in B compartment: 801\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(b_tads.bed)>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load BED files with PyBedTools\n",
    "tads = pybedtools.BedTool('az_TADs_5kb_two_tools_80.bed')\n",
    "a_compartments = pybedtools.BedTool('A.bed')\n",
    "b_compartments = pybedtools.BedTool('B.bed')\n",
    "\n",
    "#tads are in a compartment\n",
    "a_tads = tads.intersect(a_compartments, wa=True, f=0.5)\n",
    "print(\"Number of TADs in A compartment:\", a_tads.count())\n",
    "#tads are in b compartment\n",
    "b_tads = tads.intersect(b_compartments, wa=True, f=0.5)\n",
    "print(\"Number of TADs in B compartment:\", b_tads.count())\n",
    "\n",
    "a_tads.saveas('a_tads.bed')\n",
    "b_tads.saveas('b_tads.bed')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5939cd-b99f-4a8f-befb-3da6551456fa",
   "metadata": {},
   "source": [
    "Are TADs enriched in A/B compartment?\n",
    "\n",
    "Total genome length: 379627553\n",
    "Total length of A compartment: 127445834\n",
    "Total length of B compartment: 252181719\n",
    "Total TADs: 1207\n",
    "Number of TADs in A compartment: 406\n",
    "Number of TADs in B compartment: 801\n",
    "\n",
    "First, calculate the expected number of TADs in each compartment based on the proportion of the genome each compartment represents.\n",
    "Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\n",
    "Then use Chi-Squared Test to compare observed and expected values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7f10a52b-96e6-4886-8fb8-5905f9c4e6ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared statistic: 0.002345864861710905\n",
      "P-value: 0.9613702601232621\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Observed counts\n",
    "observed = [406, 801]\n",
    "# Calculate expected counts\n",
    "expected_A = (127445834 / 379627553) * 1207\n",
    "expected_B = (252181719 / 379627553) * 1207\n",
    "expected = [expected_A, expected_B]\n",
    "\n",
    "# Perform Chi-Squared Test\n",
    "chi2_stat, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "print(\"Chi-squared statistic:\", chi2_stat)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cf90c6b-dfcb-441c-b654-2e18e5acc672",
   "metadata": {},
   "source": [
    "### Conclusion: there is no statistically significant difference between the observed number of all TADs in the A and B compartments compared to what would be expected based on the lengths of the compartments. The median lengths of TADs in A/B compartment is the same = 65kb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c2673ca4-dc03-474a-9856-38eb99212091",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of TADs in A compartment: 65000.0\n"
     ]
    }
   ],
   "source": [
    "#This test is valid if the median length of TADs in A/B compartment is same. Calculating median lengths of TADs in compartments\n",
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('a_tads.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of TADs in A compartment:\", median_length)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "711daa0d-5172-430f-989c-4a342eebf2c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of TADs in B compartment: 65000.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('b_tads.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of TADs in B compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18612fa8-5016-4283-9197-d9c4fd535d34",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "623 nonTADs.bed\n"
     ]
    }
   ],
   "source": [
    "#same analysis for all nonTADs\n",
    "#list of all nonTADs\n",
    "!awk '{print $1, $2, $3}' OFS=\"\\t\" nonTADs_sonal.bed > nonTADs.bed\n",
    "!wc -l nonTADs.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d51f26b-1025-4f2b-b812-77469d6ea71b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nonTADs in A compartment: 190\n",
      "Number of nonTADs in B compartment: 410\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(b_nontads.bed)>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#number of nonTADs in A/B compartment\n",
    "tads = pybedtools.BedTool('nonTADs.bed')\n",
    "a_compartments = pybedtools.BedTool('A.bed')\n",
    "b_compartments = pybedtools.BedTool('B.bed')\n",
    "\n",
    "#tads are in a compartment\n",
    "a_tads = tads.intersect(a_compartments, wa=True, f=0.5)\n",
    "print(\"Number of nonTADs in A compartment:\", a_tads.count())\n",
    "#tads are in b compartment\n",
    "b_tads = tads.intersect(b_compartments, wa=True, f=0.5)\n",
    "print(\"Number of nonTADs in B compartment:\", b_tads.count())\n",
    "\n",
    "a_tads.saveas('a_nontads.bed')\n",
    "b_tads.saveas('b_nontads.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43254ffb-6190-4457-ba3f-8be5f16eee55",
   "metadata": {},
   "outputs": [],
   "source": [
    "Are nonTADs enriched in A/B compartment?\n",
    "\n",
    "Total genome length: 379627553\n",
    "Total length of A compartment: 127445834\n",
    "Total length of B compartment: 252181719\n",
    "Total TADs: 623\n",
    "Number of nonTADs in A compartment: 190\n",
    "Number of nonTADs in B compartment: 410\n",
    "\n",
    "First, calculate the expected number of TADs in each compartment based on the proportion of the genome each compartment represents.\n",
    "Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\n",
    "Then use Chi-Squared Test to compare observed and expected values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ecd73b4f-7470-4eb0-aa9f-e963d2f56840",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared statistic: 0.9759818142702185\n",
      "P-value: 0.3231928489732535\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Observed counts\n",
    "observed = [190, 410]\n",
    "# Calculate expected counts\n",
    "expected_A = (127445834 / 379627553) * 600\n",
    "expected_B = (252181719 / 379627553) * 600\n",
    "expected = [expected_A, expected_B]\n",
    "\n",
    "# Perform Chi-Squared Test\n",
    "chi2_stat, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "print(\"Chi-squared statistic:\", chi2_stat)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87f29eb3-1a22-46d8-87b1-e98fc76494f4",
   "metadata": {},
   "source": [
    "### Conclusion: there is no statistically significant difference between the observed number of all nonTADs in the A and B compartments compared to what would be expected based on the lengths of the compartments. Median lengths are similar: 273 and 297kb, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e0f8906c-9640-414a-8276-9004a8150916",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in A compartment: 273082.5\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('a_nontads.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in A compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d7d4229c-3e65-4cc0-8ea3-55cc31bccf0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in B compartment: 297498.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('b_nontads.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in B compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bd5d4d69-3e4c-4a87-8945-91287d6cf275",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "678 tads_exp.bed\n",
      "569 nontads_exp.bed\n"
     ]
    }
   ],
   "source": [
    "#analyzing only the TADs used for expression analysis\n",
    "!awk '$2 == \"TAD\" {print $1, $4, $5}' OFS=\"\\t\" TADs-nonTADs_geneExp.bed > tads_exp.bed\n",
    "!awk '$2 == \"nonTAD\" {print $1, $4, $5}' OFS=\"\\t\" TADs-nonTADs_geneExp.bed > nontads_exp.bed\n",
    "!wc -l tads_exp.bed\n",
    "!wc -l nontads_exp.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3773de4a-1a2a-4835-8366-beeac6a5a3e2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nonTADs exp in A compartment: 178\n",
      "Number of nonTADs exp in B compartment: 370\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(b_nontads_exp.bed)>"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#number of nonTADs_exp in A/B compartment\n",
    "tads = pybedtools.BedTool('nontads_exp.bed')\n",
    "a_compartments = pybedtools.BedTool('A.bed')\n",
    "b_compartments = pybedtools.BedTool('B.bed')\n",
    "\n",
    "#tads are in a compartment\n",
    "a_tads = tads.intersect(a_compartments, wa=True, f=0.5)\n",
    "print(\"Number of nonTADs exp in A compartment:\", a_tads.count())\n",
    "#tads are in b compartment\n",
    "b_tads = tads.intersect(b_compartments, wa=True, f=0.5)\n",
    "print(\"Number of nonTADs exp in B compartment:\", b_tads.count())\n",
    "\n",
    "a_tads.saveas('a_nontads_exp.bed')\n",
    "b_tads.saveas('b_nontads_exp.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d6d853ae-7560-4a0f-9207-9308c461fb7c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid non-printable character U+00A0 (2332527967.py, line 11)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[41], line 11\u001b[0;36m\u001b[0m\n\u001b[0;31m    Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid non-printable character U+00A0\n"
     ]
    }
   ],
   "source": [
    "Are nonTADs enriched in A/B compartment?\n",
    "\n",
    "Total genome length: 379627553\n",
    "Total length of A compartment: 127445834\n",
    "Total length of B compartment: 252181719\n",
    "Total TADs: 569\n",
    "Number of nonTADs exp in A compartment: 178\n",
    "Number of nonTADs exp in B compartment: 370\n",
    "\n",
    "First, calculate the expected number of TADs in each compartment based on the proportion of the genome each compartment represents.\n",
    "Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\n",
    "Then use Chi-Squared Test to compare observed and expected values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d2ff72c6-558f-4535-ac17-d1e6c1a8e721",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared statistic: 0.29169890151692\n",
      "P-value: 0.5891338855996818\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Observed counts\n",
    "observed = [178, 370]\n",
    "# Calculate expected counts\n",
    "expected_A = (127445834 / 379627553) * 548\n",
    "expected_B = (252181719 / 379627553) * 548\n",
    "expected = [expected_A, expected_B]\n",
    "\n",
    "# Perform Chi-Squared Test\n",
    "chi2_stat, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "print(\"Chi-squared statistic:\", chi2_stat)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11a668b9-572a-4dbe-ac3d-838ec92612f7",
   "metadata": {},
   "source": [
    "### Conclusion: there is no statistically significant difference between the observed number of nonTADs used in expression analysis in the A and B compartments compared to what would be expected based on the lengths of the compartments. Median lengths are similar: 297 and 335kb, respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "de8058d0-fa4e-4bef-83fd-916ff1a7b135",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in A compartment: 296868.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('a_nontads_exp.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in A compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d2fa95ed-f82f-4d6e-a5b8-4a61fb4632b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in B compartment: 334998.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('b_nontads_exp.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in B compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "04014315-210f-4f20-8b6c-5a144d1877bc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TADs exp in A compartment: 246\n",
      "Number of TADs exp in B compartment: 432\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File A.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t2240001\t2400000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n",
      "***** WARNING: File B.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t2240000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(b_tads_exp.bed)>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#number of TADs_exp in A/B compartment\n",
    "tads = pybedtools.BedTool('tads_exp.bed')\n",
    "a_compartments = pybedtools.BedTool('A.bed')\n",
    "b_compartments = pybedtools.BedTool('B.bed')\n",
    "\n",
    "#tads are in a compartment\n",
    "a_tads = tads.intersect(a_compartments, wa=True, f=0.5)\n",
    "print(\"Number of TADs exp in A compartment:\", a_tads.count())\n",
    "#tads are in b compartment\n",
    "b_tads = tads.intersect(b_compartments, wa=True, f=0.5)\n",
    "print(\"Number of TADs exp in B compartment:\", b_tads.count())\n",
    "\n",
    "a_tads.saveas('a_tads_exp.bed')\n",
    "b_tads.saveas('b_tads_exp.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3c12b151-c0b8-4cf8-81f8-91338131ad19",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid non-printable character U+00A0 (2332527967.py, line 11)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[41], line 11\u001b[0;36m\u001b[0m\n\u001b[0;31m    Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\u001b[0m\n\u001b[0m            ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid non-printable character U+00A0\n"
     ]
    }
   ],
   "source": [
    "Are nonTADs enriched in A/B compartment?\n",
    "\n",
    "Total genome length: 379627553\n",
    "Total length of A compartment: 127445834\n",
    "Total length of B compartment: 252181719\n",
    "Total TADs: 678\n",
    "Number of TADs exp in A compartment: 246\n",
    "Number of TADs exp in B compartment: 432\n",
    "\n",
    "First, calculate the expected number of TADs in each compartment based on the proportion of the genome each compartment represents.\n",
    "Expected TADs in compartment=(Length of compartment/Total genome length)×Total TADs\n",
    "Then use Chi-Squared Test to compare observed and expected values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "454ba550-1549-4e5f-a844-22083605d8d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared statistic: 2.2359118241662577\n",
      "P-value: 0.13483706057400863\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Observed counts\n",
    "observed = [246, 432]\n",
    "# Calculate expected counts\n",
    "expected_A = (127445834 / 379627553) * 678\n",
    "expected_B = (252181719 / 379627553) * 678\n",
    "expected = [expected_A, expected_B]\n",
    "\n",
    "# Perform Chi-Squared Test\n",
    "chi2_stat, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "print(\"Chi-squared statistic:\", chi2_stat)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91626fb3-ee5a-43c0-845b-3d7d5f8bb62c",
   "metadata": {},
   "source": [
    "### Conclusion: there is no statistically significant difference between the observed number of TADs used in expression analysis in the A and B compartments compared to what would be expected based on the lengths of the compartments. Median lengths are the same: 75kb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c796d70d-9257-4b6f-9cd8-a6aae753917e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in A compartment: 75000.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('a_tads_exp.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in A compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f28194fb-f992-4f5d-84e4-7ffd79bc7334",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median length of nonTADs in B compartment: 75000.0\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "import numpy as np\n",
    "\n",
    "# Load the BED file with pybedtools\n",
    "bed_file = pybedtools.BedTool('b_tads_exp.bed')\n",
    "\n",
    "# Calculate lengths of each region using .length attribute\n",
    "lengths = [feature.length for feature in bed_file]\n",
    "\n",
    "# Calculate the median length using numpy\n",
    "median_length = np.median(lengths)\n",
    "\n",
    "print(\"Median length of nonTADs in B compartment:\", median_length)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47d7448f-5fa3-46c0-b1ab-9b5adb29f3dd",
   "metadata": {},
   "source": [
    "# PART IV: DIFFERENT CONSERVATION GROUPS OF TADS DISTRIBUTION IN A/B COMPARTMENTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3ff9ba87-b642-47d4-9eed-af3dfd7d6801",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "643 NPB_TADs_1.bed\n",
      "438 NPB_TADs_2.bed\n",
      "221 NPB_TADs_3.bed\n",
      "98 NPB_TADs_4.bed\n",
      "24 NPB_TADs_5.bed\n"
     ]
    }
   ],
   "source": [
    "#TAD conservation groups\n",
    "!wc -l NPB_TADs_1.bed\n",
    "!wc -l NPB_TADs_2.bed\n",
    "!wc -l NPB_TADs_3.bed\n",
    "!wc -l NPB_TADs_4.bed\n",
    "!wc -l NPB_TADs_5.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "841f53c6-da1e-4a74-9696-aa7a15fe6cef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 52742146\n"
     ]
    }
   ],
   "source": [
    "#calling compartments in NPB genome\n",
    "!sbatch fancplot.sh\n",
    "#compartments are in fanc_npb_160kb.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "504c2992-cb5d-4930-9fbf-26122670eb78",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make .bed files with NPB A and B compartment coordinates\n",
    "!awk '$4 == \"A\" {print $1, $2, $3}' OFS=\"\\t\" fanc_npb_160kb.bed > An.bed\n",
    "!awk '$4 == \"B\" {print $1, $2, $3}' OFS=\"\\t\" fanc_npb_160kb.bed > Bn.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "812461d6-a216-4813-b2be-315275bac718",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#number of TADs in A/B compartment\n",
    "import pybedtools\n",
    "\n",
    "a_compartments = pybedtools.BedTool('An.bed')\n",
    "b_compartments = pybedtools.BedTool('Bn.bed')\n",
    "\n",
    "# Dictionaries to store the BedTool objects\n",
    "tad_files = {}\n",
    "a_tad_files = {}\n",
    "b_tad_files = {}\n",
    "\n",
    "# Loop through 1 to 5 to create file names dynamically\n",
    "for i in range(1, 6):\n",
    "    # Generate the filename and create a BedTool object\n",
    "    file_name = f'NPB_TADs_{i}.bed'\n",
    "    tad_files[f'tads_{i}'] = pybedtools.BedTool(file_name)\n",
    "    \n",
    "    # Perform intersection for A and B compartments\n",
    "    a_tads = tad_files[f'tads_{i}'].intersect(a_compartments, wa=True, f=0.5)\n",
    "    b_tads = tad_files[f'tads_{i}'].intersect(b_compartments, wa=True, f=0.5)\n",
    "    \n",
    "    # Save to new files and store in dictionaries\n",
    "    a_tads_bed = f'a_tads_{i}.bed'\n",
    "    b_tads_bed = f'b_tads_{i}.bed'\n",
    "    \n",
    "    a_tads.saveas(a_tads_bed)\n",
    "    b_tads.saveas(b_tads_bed)\n",
    "    \n",
    "    # Store the intersected BedTool objects\n",
    "    a_tad_files[f'a_tads_{i}'] = a_tads\n",
    "    b_tad_files[f'b_tads_{i}'] = b_tads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caf27d52-3818-422b-82f1-b5d62058a626",
   "metadata": {},
   "outputs": [],
   "source": [
    "First, we analyzed the how many TADs were found within A/B compartment \n",
    "#All TADs: az_TADs_5kb_two_tools_80.bed 1207 TADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "085b8aca-2cc5-4b95-8831-bc5907291984",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of regions: 47793143\n"
     ]
    }
   ],
   "source": [
    "# Total length of all TADs\n",
    "total_length = 0\n",
    "with open('NPB_TADs_1.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end, i = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of regions:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "78a0c605-4758-48cd-a7b0-be5a158a3066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of A compartment: 121678697\n"
     ]
    }
   ],
   "source": [
    "#The total length of the A compartment in NPB:\n",
    "total_length = 0\n",
    "with open('An.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of A compartment:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f9521b2-e9fc-43c3-bb7e-0d4e7182109f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total length of B compartment: 251566822\n"
     ]
    }
   ],
   "source": [
    "#The total length of the B compartment in NPB:\n",
    "total_length = 0\n",
    "with open('Bn.bed', 'r') as bed_file:\n",
    "    for line in bed_file:\n",
    "        # Split each line into chromosome, start, and end\n",
    "        _, start, end = line.strip().split()\n",
    "        # Calculate the length of each region and add it to total length\n",
    "        total_length += int(end) - int(start) + 1\n",
    "\n",
    "print(\"Total length of B compartment:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cefeb87f-4043-43fd-8637-346fa20d80bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total genome length: 373245519\n"
     ]
    }
   ],
   "source": [
    "#Total NPB genome length:\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Initialize a variable to store the total length\n",
    "total_length = 0\n",
    "\n",
    "# Open and read the FASTA file\n",
    "with open('NPB.fna', 'r') as fasta_file:\n",
    "    for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        # Add the length of each sequence to the total length\n",
    "        total_length += len(record.seq)\n",
    "\n",
    "print(\"Total genome length:\", total_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99577af9-7313-478c-981e-67fbd9416d3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TADs in A compartment for NPB_TADs_1.bed: 211\n",
      "Number of TADs in B compartment for NPB_TADs_1.bed: 432\n",
      "Number of TADs in A compartment for NPB_TADs_2.bed: 143\n",
      "Number of TADs in B compartment for NPB_TADs_2.bed: 295\n",
      "Number of TADs in A compartment for NPB_TADs_3.bed: 72\n",
      "Number of TADs in B compartment for NPB_TADs_3.bed: 150\n",
      "Number of TADs in A compartment for NPB_TADs_4.bed: 46\n",
      "Number of TADs in B compartment for NPB_TADs_4.bed: 52\n",
      "Number of TADs in A compartment for NPB_TADs_5.bed: 12\n",
      "Number of TADs in B compartment for NPB_TADs_5.bed: 12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File An.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t3200001\t3360000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n",
      "***** WARNING: File Bn.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t1\t3200000\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pybedtools\n",
    "\n",
    "a_compartments = pybedtools.BedTool('An.bed')\n",
    "b_compartments = pybedtools.BedTool('Bn.bed')\n",
    "\n",
    "# Dictionaries to store the BedTool objects\n",
    "tad_files = {}\n",
    "a_tad_files = {}\n",
    "b_tad_files = {}\n",
    "\n",
    "# Loop through 1 to 5 to create file names dynamically\n",
    "for i in range(1, 6):\n",
    "    # Generate the filename and create a BedTool object\n",
    "    file_name = f'NPB_TADs_{i}.bed'\n",
    "    tad_files[f'tads_{i}'] = pybedtools.BedTool(file_name)\n",
    "    \n",
    "    # Perform intersection for A and B compartments\n",
    "    a_tads = tad_files[f'tads_{i}'].intersect(a_compartments, wa=True, f=0.5)\n",
    "    b_tads = tad_files[f'tads_{i}'].intersect(b_compartments, wa=True, f=0.5)\n",
    "    \n",
    "    # Save to new files and store in dictionaries\n",
    "    a_tads_bed = f'a_tads_{i}.bed'\n",
    "    b_tads_bed = f'b_tads_{i}.bed'\n",
    "    \n",
    "    a_tads.saveas(a_tads_bed)\n",
    "    b_tads.saveas(b_tads_bed)\n",
    "    \n",
    "    # Store the intersected BedTool objects\n",
    "    a_tad_files[f'a_tads_{i}'] = a_tads\n",
    "    b_tad_files[f'b_tads_{i}'] = b_tads\n",
    "    \n",
    "    # Print the number of TADs in A and B compartments\n",
    "    num_a_tads = a_tads.count()\n",
    "    num_b_tads = b_tads.count()\n",
    "    print(f\"Number of TADs in A compartment for {file_name}: {num_a_tads}\")\n",
    "    print(f\"Number of TADs in B compartment for {file_name}: {num_b_tads}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "eada0f6e-071f-47c3-b14f-fd2555ca1d12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chi-squared statistic: 0.01349627384899706\n",
      "P-value: 0.9075150666113703\n"
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Observed counts\n",
    "observed = [211, 432]\n",
    "# Calculate expected counts\n",
    "expected_A = (121678697 / 373245519) * 643\n",
    "expected_B = (251566822 / 373245519) * 643\n",
    "expected = [expected_A, expected_B]\n",
    "\n",
    "# Perform Chi-Squared Test\n",
    "chi2_stat, p_value = stats.chisquare(f_obs=observed, f_exp=expected)\n",
    "\n",
    "print(\"Chi-squared statistic:\", chi2_stat)\n",
    "print(\"P-value:\", p_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "219186fc-5ca6-4b3e-8b82-da2b1d776276",
   "metadata": {},
   "source": [
    "### Conclusion: there is no statistically significant difference between the observed number of TADs from conservation groups 1,2,3,5 in the A and B compartments compared to what would be expected based on the lengths of the compartments. But there is an enrichment of TADs group 4 in the B compartment (52 over 46 in A compartment)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bbdd826-116c-4685-8d08-7215b725a9d8",
   "metadata": {},
   "source": [
    "# PART V: GROUPING TADS/NONTADS BY MEDIAN DISTANCE BETWEEN GENES WITHIN A TAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "46a1f43d-4ea3-4603-a407-9260fb3a0013",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "678 tads_exp.bed\n",
      "569 nontads_exp.bed\n"
     ]
    }
   ],
   "source": [
    "#tads/nontads used for CV gene expression analysis\n",
    "!wc -l tads_exp.bed\n",
    "!wc -l nontads_exp.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "79215606-992d-48c9-8958-93cc42735aa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!awk '{print $1, $2, $3}' OFS=\"\\t\" az_genes6.bed > az_genes.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cb8496f5-12cf-49e8-ab5f-a237c57325dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File az_genes.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t37926\t50143\n",
      "\n",
      "***** WARNING: File az_genes.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t37926\t50143\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(tads_exp_genes.bed)>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pybedtools\n",
    "\n",
    "# Load TADs and genes\n",
    "tads = pybedtools.BedTool('tads_exp.bed')\n",
    "genes = pybedtools.BedTool('az_genes.bed')\n",
    "\n",
    "intersected_genes = tads.intersect(genes, wa=True, wb=True, F=1)  \n",
    "intersected_genes.saveas('tads_exp_genes.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb912bb0-e92c-460c-84cb-2e5928d47760",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 678 unique TADs.\n",
      "TAD: chr01:10775000-10840000 has 4 genes.\n",
      "TAD: chr01:11370000-11475000 has 10 genes.\n",
      "TAD: chr01:11510000-11630000 has 11 genes.\n",
      "TAD: chr01:11905000-11975000 has 6 genes.\n",
      "TAD: chr01:12690000-12870000 has 12 genes.\n",
      "TAD: chr01:130000-195000 has 10 genes.\n",
      "TAD: chr01:13465000-13515000 has 4 genes.\n",
      "TAD: chr01:15195000-15255000 has 7 genes.\n",
      "TAD: chr01:15870000-15990000 has 8 genes.\n",
      "TAD: chr01:1835000-1925000 has 10 genes.\n",
      "TAD: chr01:195000-280000 has 11 genes.\n",
      "TAD: chr01:19730000-19830000 has 14 genes.\n",
      "TAD: chr01:21780000-21820000 has 6 genes.\n",
      "TAD: chr01:21820000-21915000 has 7 genes.\n",
      "TAD: chr01:22960000-23015000 has 5 genes.\n",
      "TAD: chr01:23630000-23725000 has 9 genes.\n",
      "TAD: chr01:23725000-23795000 has 7 genes.\n",
      "TAD: chr01:24130000-24210000 has 7 genes.\n",
      "TAD: chr01:24210000-24275000 has 8 genes.\n",
      "TAD: chr01:24555000-24635000 has 11 genes.\n",
      "TAD: chr01:24670000-24710000 has 10 genes.\n",
      "TAD: chr01:24710000-24755000 has 4 genes.\n",
      "TAD: chr01:24860000-24930000 has 11 genes.\n",
      "TAD: chr01:24930000-25040000 has 12 genes.\n",
      "TAD: chr01:25040000-25125000 has 9 genes.\n",
      "TAD: chr01:25125000-25165000 has 7 genes.\n",
      "TAD: chr01:25300000-25375000 has 10 genes.\n",
      "TAD: chr01:25795000-25830000 has 4 genes.\n",
      "TAD: chr01:25855000-25910000 has 3 genes.\n",
      "TAD: chr01:25910000-25975000 has 9 genes.\n",
      "TAD: chr01:26140000-26225000 has 9 genes.\n",
      "TAD: chr01:26225000-26297500 has 9 genes.\n",
      "TAD: chr01:26695000-26790000 has 11 genes.\n",
      "TAD: chr01:26920000-27060000 has 14 genes.\n",
      "TAD: chr01:27060000-27140000 has 10 genes.\n",
      "TAD: chr01:2740000-2805000 has 9 genes.\n",
      "TAD: chr01:27840000-27895000 has 6 genes.\n",
      "TAD: chr01:2805000-2940000 has 17 genes.\n",
      "TAD: chr01:28335000-28405000 has 9 genes.\n",
      "TAD: chr01:29015000-29060000 has 6 genes.\n",
      "TAD: chr01:29060000-29095000 has 6 genes.\n",
      "TAD: chr01:29350000-29450000 has 13 genes.\n",
      "TAD: chr01:29495000-29537500 has 4 genes.\n",
      "TAD: chr01:30650000-30705000 has 11 genes.\n",
      "TAD: chr01:30895000-30955000 has 13 genes.\n",
      "TAD: chr01:30955000-31040000 has 12 genes.\n",
      "TAD: chr01:31750000-31877500 has 13 genes.\n",
      "TAD: chr01:31935000-32010000 has 12 genes.\n",
      "TAD: chr01:32820000-32895000 has 13 genes.\n",
      "TAD: chr01:33070000-33195000 has 15 genes.\n",
      "TAD: chr01:33235000-33265000 has 5 genes.\n",
      "TAD: chr01:33265000-33335000 has 10 genes.\n",
      "TAD: chr01:33335000-33390000 has 5 genes.\n",
      "TAD: chr01:3555000-3650000 has 19 genes.\n",
      "TAD: chr01:35920000-35980000 has 6 genes.\n",
      "TAD: chr01:35980000-36075000 has 9 genes.\n",
      "TAD: chr01:36315000-36395000 has 12 genes.\n",
      "TAD: chr01:36540000-36585000 has 10 genes.\n",
      "TAD: chr01:36585000-36670000 has 11 genes.\n",
      "TAD: chr01:36800000-36850000 has 5 genes.\n",
      "TAD: chr01:36850000-36900000 has 8 genes.\n",
      "TAD: chr01:37355000-37435000 has 11 genes.\n",
      "TAD: chr01:37940000-37980000 has 9 genes.\n",
      "TAD: chr01:37980000-38025000 has 11 genes.\n",
      "TAD: chr01:38075000-38180000 has 15 genes.\n",
      "TAD: chr01:38615000-38700000 has 12 genes.\n",
      "TAD: chr01:38700000-38775000 has 11 genes.\n",
      "TAD: chr01:38775000-38840000 has 13 genes.\n",
      "TAD: chr01:38885000-38965000 has 15 genes.\n",
      "TAD: chr01:40077500-40120000 has 5 genes.\n",
      "TAD: chr01:40120000-40175000 has 9 genes.\n",
      "TAD: chr01:40215000-40260000 has 8 genes.\n",
      "TAD: chr01:40500000-40625000 has 22 genes.\n",
      "TAD: chr01:40685000-40740000 has 6 genes.\n",
      "TAD: chr01:40740000-40825000 has 12 genes.\n",
      "TAD: chr01:40825000-40865000 has 5 genes.\n",
      "TAD: chr01:41530000-41630000 has 17 genes.\n",
      "TAD: chr01:41630000-41705000 has 11 genes.\n",
      "TAD: chr01:41705000-41760000 has 4 genes.\n",
      "TAD: chr01:41760000-41845000 has 14 genes.\n",
      "TAD: chr01:43260000-43325000 has 13 genes.\n",
      "TAD: chr01:43475000-43570000 has 11 genes.\n",
      "TAD: chr01:43700000-43740000 has 5 genes.\n",
      "TAD: chr01:4945000-5035000 has 10 genes.\n",
      "TAD: chr01:5395000-5445000 has 4 genes.\n",
      "TAD: chr01:5445000-5505000 has 4 genes.\n",
      "TAD: chr01:5660000-5730000 has 7 genes.\n",
      "TAD: chr01:5850000-5910000 has 8 genes.\n",
      "TAD: chr01:6660000-6720000 has 6 genes.\n",
      "TAD: chr01:6720000-6860000 has 18 genes.\n",
      "TAD: chr01:7030000-7105000 has 7 genes.\n",
      "TAD: chr01:7105000-7185000 has 6 genes.\n",
      "TAD: chr01:7695000-7790000 has 15 genes.\n",
      "TAD: chr01:85000-130000 has 5 genes.\n",
      "TAD: chr01:9285000-9375000 has 10 genes.\n",
      "TAD: chr01:9375000-9435000 has 7 genes.\n",
      "TAD: chr01:9435000-9540000 has 11 genes.\n",
      "TAD: chr02:10055000-10145000 has 8 genes.\n",
      "TAD: chr02:11970000-12180000 has 23 genes.\n",
      "TAD: chr02:12300000-12440000 has 9 genes.\n",
      "TAD: chr02:12740000-12865000 has 16 genes.\n",
      "TAD: chr02:12865000-12960000 has 7 genes.\n",
      "TAD: chr02:13275000-13410000 has 7 genes.\n",
      "TAD: chr02:13995000-14115000 has 14 genes.\n",
      "TAD: chr02:15535000-15630000 has 8 genes.\n",
      "TAD: chr02:1555000-1620000 has 11 genes.\n",
      "TAD: chr02:15630000-15700000 has 10 genes.\n",
      "TAD: chr02:15945000-16030000 has 7 genes.\n",
      "TAD: chr02:18935000-19025000 has 8 genes.\n",
      "TAD: chr02:190000-305000 has 9 genes.\n",
      "TAD: chr02:19150000-19225000 has 8 genes.\n",
      "TAD: chr02:19675000-19730000 has 7 genes.\n",
      "TAD: chr02:19730000-19855000 has 14 genes.\n",
      "TAD: chr02:19855000-19965000 has 9 genes.\n",
      "TAD: chr02:19995000-20035000 has 9 genes.\n",
      "TAD: chr02:20080000-20165000 has 8 genes.\n",
      "TAD: chr02:20980000-21040000 has 5 genes.\n",
      "TAD: chr02:21040000-21095000 has 6 genes.\n",
      "TAD: chr02:21485000-21540000 has 6 genes.\n",
      "TAD: chr02:22615000-22700000 has 10 genes.\n",
      "TAD: chr02:22700000-22780000 has 12 genes.\n",
      "TAD: chr02:23695000-23745000 has 11 genes.\n",
      "TAD: chr02:24020000-24075000 has 8 genes.\n",
      "TAD: chr02:24075000-24150000 has 8 genes.\n",
      "TAD: chr02:24350000-24440000 has 9 genes.\n",
      "TAD: chr02:25125000-25165000 has 7 genes.\n",
      "TAD: chr02:25360000-25515000 has 12 genes.\n",
      "TAD: chr02:26720000-26755000 has 6 genes.\n",
      "TAD: chr02:26755000-26820000 has 10 genes.\n",
      "TAD: chr02:26820000-26870000 has 7 genes.\n",
      "TAD: chr02:27055000-27245000 has 17 genes.\n",
      "TAD: chr02:27245000-27340000 has 12 genes.\n",
      "TAD: chr02:27400000-27455000 has 5 genes.\n",
      "TAD: chr02:27455000-27505000 has 8 genes.\n",
      "TAD: chr02:27580000-27860000 has 36 genes.\n",
      "TAD: chr02:29025000-29095000 has 8 genes.\n",
      "TAD: chr02:29145000-29175000 has 5 genes.\n",
      "TAD: chr02:29175000-29285000 has 16 genes.\n",
      "TAD: chr02:29285000-29370000 has 6 genes.\n",
      "TAD: chr02:29487500-29540000 has 8 genes.\n",
      "TAD: chr02:29700000-29780000 has 11 genes.\n",
      "TAD: chr02:30475000-30545000 has 7 genes.\n",
      "TAD: chr02:305000-420000 has 11 genes.\n",
      "TAD: chr02:30630000-30740000 has 16 genes.\n",
      "TAD: chr02:30740000-30825000 has 14 genes.\n",
      "TAD: chr02:30910000-31035000 has 20 genes.\n",
      "TAD: chr02:31135000-31245000 has 18 genes.\n",
      "TAD: chr02:31405000-31470000 has 8 genes.\n",
      "TAD: chr02:31470000-31540000 has 7 genes.\n",
      "TAD: chr02:31560000-31650000 has 9 genes.\n",
      "TAD: chr02:31680000-31765000 has 12 genes.\n",
      "TAD: chr02:31945000-32005000 has 11 genes.\n",
      "TAD: chr02:32005000-32085000 has 11 genes.\n",
      "TAD: chr02:32085000-32170000 has 9 genes.\n",
      "TAD: chr02:32495000-32610000 has 19 genes.\n",
      "TAD: chr02:32610000-32685000 has 8 genes.\n",
      "TAD: chr02:33320000-33385000 has 5 genes.\n",
      "TAD: chr02:33445000-33490000 has 9 genes.\n",
      "TAD: chr02:33580000-33655000 has 9 genes.\n",
      "TAD: chr02:33655000-33720000 has 9 genes.\n",
      "TAD: chr02:33925000-33985000 has 8 genes.\n",
      "TAD: chr02:33985000-34050000 has 7 genes.\n",
      "TAD: chr02:3405000-3525000 has 19 genes.\n",
      "TAD: chr02:34245000-34345000 has 15 genes.\n",
      "TAD: chr02:34430000-34495000 has 11 genes.\n",
      "TAD: chr02:34495000-34540000 has 5 genes.\n",
      "TAD: chr02:35115000-35190000 has 11 genes.\n",
      "TAD: chr02:3525000-3615000 has 10 genes.\n",
      "TAD: chr02:35295000-35465000 has 24 genes.\n",
      "TAD: chr02:35690000-35730000 has 5 genes.\n",
      "TAD: chr02:35730000-35800000 has 13 genes.\n",
      "TAD: chr02:35800000-35910000 has 18 genes.\n",
      "TAD: chr02:35910000-35965000 has 10 genes.\n",
      "TAD: chr02:35965000-36050000 has 17 genes.\n",
      "TAD: chr02:3615000-3680000 has 9 genes.\n",
      "TAD: chr02:3680000-3790000 has 17 genes.\n",
      "TAD: chr02:3790000-3920000 has 15 genes.\n",
      "TAD: chr02:4035000-4075000 has 8 genes.\n",
      "TAD: chr02:4175000-4220000 has 6 genes.\n",
      "TAD: chr02:5055000-5115000 has 6 genes.\n",
      "TAD: chr02:5115000-5320000 has 24 genes.\n",
      "TAD: chr02:5320000-5400000 has 11 genes.\n",
      "TAD: chr02:570000-650000 has 8 genes.\n",
      "TAD: chr02:5760000-5845000 has 9 genes.\n",
      "TAD: chr02:5845000-5930000 has 10 genes.\n",
      "TAD: chr02:6000000-6110000 has 12 genes.\n",
      "TAD: chr02:6595000-6645000 has 5 genes.\n",
      "TAD: chr02:7715000-7770000 has 6 genes.\n",
      "TAD: chr02:7770000-7870000 has 9 genes.\n",
      "TAD: chr02:8885000-8935000 has 8 genes.\n",
      "TAD: chr03:10125000-10245000 has 17 genes.\n",
      "TAD: chr03:10755000-10795000 has 7 genes.\n",
      "TAD: chr03:11235000-11275000 has 6 genes.\n",
      "TAD: chr03:11855000-11925000 has 8 genes.\n",
      "TAD: chr03:12940000-13020000 has 11 genes.\n",
      "TAD: chr03:13365000-13435000 has 7 genes.\n",
      "TAD: chr03:14430000-14515000 has 11 genes.\n",
      "TAD: chr03:15010000-15050000 has 10 genes.\n",
      "TAD: chr03:15050000-15175000 has 10 genes.\n",
      "TAD: chr03:15490000-15605000 has 13 genes.\n",
      "TAD: chr03:15855000-15920000 has 8 genes.\n",
      "TAD: chr03:1615000-1705000 has 11 genes.\n",
      "TAD: chr03:19435000-19495000 has 7 genes.\n",
      "TAD: chr03:24110000-24160000 has 6 genes.\n",
      "TAD: chr03:24160000-24230000 has 11 genes.\n",
      "TAD: chr03:24280000-24365000 has 5 genes.\n",
      "TAD: chr03:24575000-24680000 has 11 genes.\n",
      "TAD: chr03:26295000-26390000 has 5 genes.\n",
      "TAD: chr03:26955000-27020000 has 7 genes.\n",
      "TAD: chr03:27055000-27095000 has 5 genes.\n",
      "TAD: chr03:27465000-27540000 has 10 genes.\n",
      "TAD: chr03:27540000-27585000 has 6 genes.\n",
      "TAD: chr03:27710000-27770000 has 3 genes.\n",
      "TAD: chr03:28155000-28237500 has 11 genes.\n",
      "TAD: chr03:2905000-2965000 has 6 genes.\n",
      "TAD: chr03:29220000-29295000 has 9 genes.\n",
      "TAD: chr03:2965000-3015000 has 5 genes.\n",
      "TAD: chr03:30145000-30175000 has 6 genes.\n",
      "TAD: chr03:30220000-30295000 has 12 genes.\n",
      "TAD: chr03:30550000-30580000 has 5 genes.\n",
      "TAD: chr03:30730000-30820000 has 10 genes.\n",
      "TAD: chr03:30820000-30860000 has 8 genes.\n",
      "TAD: chr03:31265000-31310000 has 9 genes.\n",
      "TAD: chr03:3150000-3220000 has 9 genes.\n",
      "TAD: chr03:32120000-32190000 has 8 genes.\n",
      "TAD: chr03:32190000-32230000 has 6 genes.\n",
      "TAD: chr03:3220000-3255000 has 8 genes.\n",
      "TAD: chr03:32755000-32810000 has 10 genes.\n",
      "TAD: chr03:32810000-32895000 has 7 genes.\n",
      "TAD: chr03:33425000-33475000 has 5 genes.\n",
      "TAD: chr03:33875000-33955000 has 13 genes.\n",
      "TAD: chr03:34115000-34165000 has 5 genes.\n",
      "TAD: chr03:34165000-34255000 has 17 genes.\n",
      "TAD: chr03:34740000-34825000 has 12 genes.\n",
      "TAD: chr03:35185000-35280000 has 9 genes.\n",
      "TAD: chr03:35465000-35510000 has 6 genes.\n",
      "TAD: chr03:35845000-35920000 has 12 genes.\n",
      "TAD: chr03:35920000-36010000 has 8 genes.\n",
      "TAD: chr03:36400000-36460000 has 9 genes.\n",
      "TAD: chr03:36460000-36515000 has 9 genes.\n",
      "TAD: chr03:36515000-36600000 has 12 genes.\n",
      "TAD: chr03:36745000-36790000 has 10 genes.\n",
      "TAD: chr03:36790000-36850000 has 9 genes.\n",
      "TAD: chr03:36960000-36995000 has 9 genes.\n",
      "TAD: chr03:36995000-37110000 has 17 genes.\n",
      "TAD: chr03:37110000-37170000 has 8 genes.\n",
      "TAD: chr03:3720000-3810000 has 21 genes.\n",
      "TAD: chr03:37925000-37975000 has 8 genes.\n",
      "TAD: chr03:4085000-4150000 has 11 genes.\n",
      "TAD: chr03:4400000-4460000 has 8 genes.\n",
      "TAD: chr03:4460000-4515000 has 6 genes.\n",
      "TAD: chr03:4725000-4800000 has 9 genes.\n",
      "TAD: chr03:5040000-5115000 has 6 genes.\n",
      "TAD: chr03:5115000-5175000 has 6 genes.\n",
      "TAD: chr03:5175000-5220000 has 4 genes.\n",
      "TAD: chr03:5220000-5280000 has 7 genes.\n",
      "TAD: chr03:5770000-5815000 has 7 genes.\n",
      "TAD: chr03:605000-640000 has 4 genes.\n",
      "TAD: chr03:6620000-6670000 has 6 genes.\n",
      "TAD: chr03:6670000-6740000 has 9 genes.\n",
      "TAD: chr03:8355000-8400000 has 5 genes.\n",
      "TAD: chr03:8515000-8595000 has 10 genes.\n",
      "TAD: chr03:8790000-8840000 has 9 genes.\n",
      "TAD: chr03:8840000-8930000 has 15 genes.\n",
      "TAD: chr03:8930000-8965000 has 9 genes.\n",
      "TAD: chr03:910000-970000 has 8 genes.\n",
      "TAD: chr03:9505000-9555000 has 12 genes.\n",
      "TAD: chr03:9690000-9795000 has 15 genes.\n",
      "TAD: chr03:970000-1095000 has 19 genes.\n",
      "TAD: chr04:10205000-10350000 has 13 genes.\n",
      "TAD: chr04:10350000-10465000 has 6 genes.\n",
      "TAD: chr04:11970000-12060000 has 9 genes.\n",
      "TAD: chr04:1225000-1270000 has 5 genes.\n",
      "TAD: chr04:1462500-1580000 has 6 genes.\n",
      "TAD: chr04:14950000-15055000 has 11 genes.\n",
      "TAD: chr04:18045000-18135000 has 10 genes.\n",
      "TAD: chr04:18135000-18235000 has 7 genes.\n",
      "TAD: chr04:19315000-19350000 has 4 genes.\n",
      "TAD: chr04:19825000-19880000 has 9 genes.\n",
      "TAD: chr04:19880000-19940000 has 6 genes.\n",
      "TAD: chr04:20150000-20205000 has 5 genes.\n",
      "TAD: chr04:20695000-20790000 has 11 genes.\n",
      "TAD: chr04:21685000-21730000 has 10 genes.\n",
      "TAD: chr04:21730000-21825000 has 14 genes.\n",
      "TAD: chr04:21825000-21895000 has 10 genes.\n",
      "TAD: chr04:22655000-22715000 has 12 genes.\n",
      "TAD: chr04:22715000-22770000 has 7 genes.\n",
      "TAD: chr04:23235000-23345000 has 11 genes.\n",
      "TAD: chr04:23605000-23660000 has 5 genes.\n",
      "TAD: chr04:23955000-24110000 has 25 genes.\n",
      "TAD: chr04:24230000-24305000 has 8 genes.\n",
      "TAD: chr04:24430000-24495000 has 9 genes.\n",
      "TAD: chr04:24540000-24630000 has 3 genes.\n",
      "TAD: chr04:24630000-24700000 has 9 genes.\n",
      "TAD: chr04:24700000-24770000 has 11 genes.\n",
      "TAD: chr04:25105000-25295000 has 38 genes.\n",
      "TAD: chr04:26655000-26700000 has 7 genes.\n",
      "TAD: chr04:26700000-26755000 has 8 genes.\n",
      "TAD: chr04:26925000-27005000 has 8 genes.\n",
      "TAD: chr04:27735000-27785000 has 8 genes.\n",
      "TAD: chr04:28245000-28317500 has 9 genes.\n",
      "TAD: chr04:28400000-28470000 has 7 genes.\n",
      "TAD: chr04:29045000-29120000 has 8 genes.\n",
      "TAD: chr04:29825000-29860000 has 4 genes.\n",
      "TAD: chr04:31615000-31740000 has 14 genes.\n",
      "TAD: chr04:32272500-32360000 has 12 genes.\n",
      "TAD: chr04:32510000-32600000 has 11 genes.\n",
      "TAD: chr04:33325000-33385000 has 8 genes.\n",
      "TAD: chr04:33385000-33525000 has 24 genes.\n",
      "TAD: chr04:34000000-34055000 has 6 genes.\n",
      "TAD: chr04:4027500-4110000 has 11 genes.\n",
      "TAD: chr04:5235000-5365000 has 12 genes.\n",
      "TAD: chr04:7847500-7900000 has 5 genes.\n",
      "TAD: chr05:13755000-13905000 has 11 genes.\n",
      "TAD: chr05:13905000-14020000 has 7 genes.\n",
      "TAD: chr05:17525000-17705000 has 13 genes.\n",
      "TAD: chr05:17755000-17820000 has 8 genes.\n",
      "TAD: chr05:18380000-18465000 has 9 genes.\n",
      "TAD: chr05:18815000-18885000 has 11 genes.\n",
      "TAD: chr05:19705000-19775000 has 11 genes.\n",
      "TAD: chr05:20270000-20380000 has 11 genes.\n",
      "TAD: chr05:20565000-20665000 has 14 genes.\n",
      "TAD: chr05:20800000-20875000 has 4 genes.\n",
      "TAD: chr05:20875000-20950000 has 10 genes.\n",
      "TAD: chr05:21540000-21585000 has 4 genes.\n",
      "TAD: chr05:21730000-21790000 has 11 genes.\n",
      "TAD: chr05:21880000-21935000 has 4 genes.\n",
      "TAD: chr05:22085000-22145000 has 8 genes.\n",
      "TAD: chr05:22145000-22225000 has 12 genes.\n",
      "TAD: chr05:22575000-22635000 has 7 genes.\n",
      "TAD: chr05:22635000-22725000 has 9 genes.\n",
      "TAD: chr05:2305000-2430000 has 16 genes.\n",
      "TAD: chr05:23410000-23465000 has 9 genes.\n",
      "TAD: chr05:23655000-23725000 has 10 genes.\n",
      "TAD: chr05:24005000-24085000 has 6 genes.\n",
      "TAD: chr05:24085000-24115000 has 4 genes.\n",
      "TAD: chr05:24260000-24330000 has 9 genes.\n",
      "TAD: chr05:2430000-2470000 has 5 genes.\n",
      "TAD: chr05:24330000-24410000 has 6 genes.\n",
      "TAD: chr05:24580000-24650000 has 9 genes.\n",
      "TAD: chr05:24650000-24715000 has 11 genes.\n",
      "TAD: chr05:25080000-25130000 has 9 genes.\n",
      "TAD: chr05:25130000-25175000 has 6 genes.\n",
      "TAD: chr05:25175000-25235000 has 9 genes.\n",
      "TAD: chr05:25615000-25670000 has 12 genes.\n",
      "TAD: chr05:25670000-25715000 has 5 genes.\n",
      "TAD: chr05:25715000-25825000 has 15 genes.\n",
      "TAD: chr05:25825000-26195000 has 16 genes.\n",
      "TAD: chr05:26345000-26425000 has 12 genes.\n",
      "TAD: chr05:26860000-26985000 has 11 genes.\n",
      "TAD: chr05:27145000-27230000 has 10 genes.\n",
      "TAD: chr05:27230000-27315000 has 14 genes.\n",
      "TAD: chr05:27315000-27415000 has 17 genes.\n",
      "TAD: chr05:27540000-27605000 has 7 genes.\n",
      "TAD: chr05:27605000-27660000 has 8 genes.\n",
      "TAD: chr05:27660000-27750000 has 11 genes.\n",
      "TAD: chr05:27750000-27835000 has 9 genes.\n",
      "TAD: chr05:28210000-28240000 has 5 genes.\n",
      "TAD: chr05:28560000-28600000 has 5 genes.\n",
      "TAD: chr05:2870000-2940000 has 12 genes.\n",
      "TAD: chr05:28915000-28970000 has 7 genes.\n",
      "TAD: chr05:29405000-29490000 has 11 genes.\n",
      "TAD: chr05:29490000-29550000 has 11 genes.\n",
      "TAD: chr05:29965000-30005000 has 3 genes.\n",
      "TAD: chr05:30005000-30035000 has 2 genes.\n",
      "TAD: chr05:30035000-30092500 has 9 genes.\n",
      "TAD: chr05:30220000-30300000 has 10 genes.\n",
      "TAD: chr05:3110000-3165000 has 6 genes.\n",
      "TAD: chr05:3165000-3220000 has 6 genes.\n",
      "TAD: chr05:3220000-3300000 has 10 genes.\n",
      "TAD: chr05:4995000-5075000 has 5 genes.\n",
      "TAD: chr05:5170000-5260000 has 9 genes.\n",
      "TAD: chr05:665000-765000 has 13 genes.\n",
      "TAD: chr05:7030000-7120000 has 9 genes.\n",
      "TAD: chr05:870000-950000 has 15 genes.\n",
      "TAD: chr05:8722500-8890000 has 18 genes.\n",
      "TAD: chr05:95000-145000 has 4 genes.\n",
      "TAD: chr05:9820000-9910000 has 5 genes.\n",
      "TAD: chr06:10965000-11055000 has 8 genes.\n",
      "TAD: chr06:12720000-12765000 has 6 genes.\n",
      "TAD: chr06:13315000-13370000 has 5 genes.\n",
      "TAD: chr06:14882500-15007500 has 4 genes.\n",
      "TAD: chr06:1500000-1615000 has 16 genes.\n",
      "TAD: chr06:16935000-17010000 has 7 genes.\n",
      "TAD: chr06:190000-270000 has 14 genes.\n",
      "TAD: chr06:2015000-2100000 has 14 genes.\n",
      "TAD: chr06:22375000-22445000 has 10 genes.\n",
      "TAD: chr06:2255000-2335000 has 12 genes.\n",
      "TAD: chr06:23850000-23925000 has 6 genes.\n",
      "TAD: chr06:24615000-24675000 has 6 genes.\n",
      "TAD: chr06:25090000-25150000 has 8 genes.\n",
      "TAD: chr06:25150000-25200000 has 5 genes.\n",
      "TAD: chr06:26030000-26140000 has 13 genes.\n",
      "TAD: chr06:26140000-26375000 has 14 genes.\n",
      "TAD: chr06:26415000-26495000 has 8 genes.\n",
      "TAD: chr06:26820000-26895000 has 13 genes.\n",
      "TAD: chr06:270000-335000 has 6 genes.\n",
      "TAD: chr06:27470000-27570000 has 13 genes.\n",
      "TAD: chr06:27570000-27635000 has 10 genes.\n",
      "TAD: chr06:28450000-28560000 has 9 genes.\n",
      "TAD: chr06:28730000-28830000 has 11 genes.\n",
      "TAD: chr06:28830000-28880000 has 9 genes.\n",
      "TAD: chr06:28880000-28965000 has 15 genes.\n",
      "TAD: chr06:29905000-29965000 has 9 genes.\n",
      "TAD: chr06:29965000-30060000 has 11 genes.\n",
      "TAD: chr06:30060000-30135000 has 11 genes.\n",
      "TAD: chr06:30445000-30480000 has 7 genes.\n",
      "TAD: chr06:30480000-30540000 has 9 genes.\n",
      "TAD: chr06:30690000-30760000 has 9 genes.\n",
      "TAD: chr06:3280000-3485000 has 31 genes.\n",
      "TAD: chr06:4070000-4170000 has 18 genes.\n",
      "TAD: chr06:4170000-4280000 has 9 genes.\n",
      "TAD: chr06:4520000-4570000 has 4 genes.\n",
      "TAD: chr06:470000-535000 has 12 genes.\n",
      "TAD: chr06:4775000-4815000 has 10 genes.\n",
      "TAD: chr06:4815000-4885000 has 10 genes.\n",
      "TAD: chr06:4885000-4970000 has 6 genes.\n",
      "TAD: chr06:5155000-5200000 has 8 genes.\n",
      "TAD: chr06:685000-745000 has 8 genes.\n",
      "TAD: chr06:7290000-7425000 has 16 genes.\n",
      "TAD: chr06:745000-870000 has 10 genes.\n",
      "TAD: chr06:7790000-7850000 has 9 genes.\n",
      "TAD: chr06:7850000-7910000 has 9 genes.\n",
      "TAD: chr06:8040000-8100000 has 7 genes.\n",
      "TAD: chr06:8195000-8270000 has 8 genes.\n",
      "TAD: chr06:900000-1030000 has 24 genes.\n",
      "TAD: chr06:9285000-9360000 has 5 genes.\n",
      "TAD: chr06:9360000-9415000 has 7 genes.\n",
      "TAD: chr06:9675000-9815000 has 8 genes.\n",
      "TAD: chr07:10090000-10165000 has 7 genes.\n",
      "TAD: chr07:11280000-11365000 has 7 genes.\n",
      "TAD: chr07:12460000-12560000 has 8 genes.\n",
      "TAD: chr07:1250000-1335000 has 12 genes.\n",
      "TAD: chr07:12560000-12630000 has 6 genes.\n",
      "TAD: chr07:13290000-13460000 has 14 genes.\n",
      "TAD: chr07:1335000-1370000 has 6 genes.\n",
      "TAD: chr07:14150000-14265000 has 9 genes.\n",
      "TAD: chr07:15460000-15610000 has 7 genes.\n",
      "TAD: chr07:16035000-16115000 has 9 genes.\n",
      "TAD: chr07:16410000-16510000 has 6 genes.\n",
      "TAD: chr07:16935000-17015000 has 8 genes.\n",
      "TAD: chr07:17015000-17145000 has 4 genes.\n",
      "TAD: chr07:17145000-17245000 has 10 genes.\n",
      "TAD: chr07:17730000-17910000 has 10 genes.\n",
      "TAD: chr07:19605000-19680000 has 8 genes.\n",
      "TAD: chr07:19715000-19800000 has 10 genes.\n",
      "TAD: chr07:20250000-20320000 has 11 genes.\n",
      "TAD: chr07:20350000-20480000 has 14 genes.\n",
      "TAD: chr07:2047500-2180000 has 25 genes.\n",
      "TAD: chr07:20480000-20555000 has 6 genes.\n",
      "TAD: chr07:20675000-20780000 has 12 genes.\n",
      "TAD: chr07:21315000-21385000 has 10 genes.\n",
      "TAD: chr07:21690000-21755000 has 3 genes.\n",
      "TAD: chr07:2180000-2230000 has 10 genes.\n",
      "TAD: chr07:2230000-2422500 has 19 genes.\n",
      "TAD: chr07:22465000-22500000 has 6 genes.\n",
      "TAD: chr07:22500000-22580000 has 4 genes.\n",
      "TAD: chr07:22730000-22815000 has 14 genes.\n",
      "TAD: chr07:23585000-23650000 has 9 genes.\n",
      "TAD: chr07:23690000-23750000 has 11 genes.\n",
      "TAD: chr07:23750000-23830000 has 8 genes.\n",
      "TAD: chr07:23830000-23895000 has 5 genes.\n",
      "TAD: chr07:23895000-24005000 has 15 genes.\n",
      "TAD: chr07:24005000-24075000 has 7 genes.\n",
      "TAD: chr07:24075000-24140000 has 4 genes.\n",
      "TAD: chr07:24540000-24600000 has 9 genes.\n",
      "TAD: chr07:25215000-25310000 has 11 genes.\n",
      "TAD: chr07:2530000-2620000 has 8 genes.\n",
      "TAD: chr07:26497500-26565000 has 12 genes.\n",
      "TAD: chr07:27195000-27265000 has 8 genes.\n",
      "TAD: chr07:27460000-27535000 has 4 genes.\n",
      "TAD: chr07:28785000-28860000 has 7 genes.\n",
      "TAD: chr07:28860000-28955000 has 12 genes.\n",
      "TAD: chr07:28955000-29090000 has 17 genes.\n",
      "TAD: chr07:3570000-3675000 has 12 genes.\n",
      "TAD: chr07:3720000-3805000 has 15 genes.\n",
      "TAD: chr07:3960000-4040000 has 8 genes.\n",
      "TAD: chr07:4130000-4220000 has 9 genes.\n",
      "TAD: chr07:4220000-4285000 has 5 genes.\n",
      "TAD: chr07:4360000-4460000 has 14 genes.\n",
      "TAD: chr07:4665000-4705000 has 5 genes.\n",
      "TAD: chr07:4975000-5085000 has 12 genes.\n",
      "TAD: chr07:5480000-5640000 has 23 genes.\n",
      "TAD: chr07:5860000-5925000 has 4 genes.\n",
      "TAD: chr07:595000-640000 has 6 genes.\n",
      "TAD: chr07:640000-690000 has 6 genes.\n",
      "TAD: chr07:7140000-7215000 has 8 genes.\n",
      "TAD: chr07:7215000-7295000 has 8 genes.\n",
      "TAD: chr07:7295000-7410000 has 10 genes.\n",
      "TAD: chr08:11395000-11630000 has 17 genes.\n",
      "TAD: chr08:12240000-12330000 has 7 genes.\n",
      "TAD: chr08:14870000-14935000 has 6 genes.\n",
      "TAD: chr08:17745000-17830000 has 8 genes.\n",
      "TAD: chr08:20135000-20205000 has 5 genes.\n",
      "TAD: chr08:20915000-20990000 has 11 genes.\n",
      "TAD: chr08:21035000-21100000 has 6 genes.\n",
      "TAD: chr08:21577500-21690000 has 10 genes.\n",
      "TAD: chr08:21920000-21975000 has 9 genes.\n",
      "TAD: chr08:23665000-23760000 has 10 genes.\n",
      "TAD: chr08:23760000-23835000 has 9 genes.\n",
      "TAD: chr08:23895000-23970000 has 7 genes.\n",
      "TAD: chr08:24020000-24100000 has 10 genes.\n",
      "TAD: chr08:24705000-24735000 has 5 genes.\n",
      "TAD: chr08:24820000-24870000 has 6 genes.\n",
      "TAD: chr08:24870000-24960000 has 9 genes.\n",
      "TAD: chr08:24960000-25080000 has 19 genes.\n",
      "TAD: chr08:25080000-25155000 has 7 genes.\n",
      "TAD: chr08:25155000-25240000 has 13 genes.\n",
      "TAD: chr08:25425000-25535000 has 9 genes.\n",
      "TAD: chr08:26115000-26180000 has 12 genes.\n",
      "TAD: chr08:26835000-26880000 has 9 genes.\n",
      "TAD: chr08:26880000-26940000 has 7 genes.\n",
      "TAD: chr08:27045000-27135000 has 14 genes.\n",
      "TAD: chr08:27505000-27550000 has 6 genes.\n",
      "TAD: chr08:27550000-27640000 has 9 genes.\n",
      "TAD: chr08:28060000-28105000 has 7 genes.\n",
      "TAD: chr08:28740000-28770000 has 5 genes.\n",
      "TAD: chr08:28770000-28820000 has 5 genes.\n",
      "TAD: chr08:28820000-28875000 has 11 genes.\n",
      "TAD: chr08:3305000-3415000 has 11 genes.\n",
      "TAD: chr08:3470000-3525000 has 4 genes.\n",
      "TAD: chr08:3525000-3600000 has 8 genes.\n",
      "TAD: chr08:3600000-3695000 has 10 genes.\n",
      "TAD: chr08:3695000-3795000 has 11 genes.\n",
      "TAD: chr08:3840000-3920000 has 13 genes.\n",
      "TAD: chr08:5535000-5755000 has 19 genes.\n",
      "TAD: chr08:60000-115000 has 9 genes.\n",
      "TAD: chr08:760000-905000 has 9 genes.\n",
      "TAD: chr08:7850000-7915000 has 6 genes.\n",
      "TAD: chr08:8160000-8220000 has 7 genes.\n",
      "TAD: chr08:8220000-8330000 has 12 genes.\n",
      "TAD: chr08:905000-975000 has 9 genes.\n",
      "TAD: chr08:9820000-9965000 has 14 genes.\n",
      "TAD: chr09:10760000-10880000 has 12 genes.\n",
      "TAD: chr09:12525000-12585000 has 9 genes.\n",
      "TAD: chr09:13005000-13125000 has 9 genes.\n",
      "TAD: chr09:13125000-13280000 has 16 genes.\n",
      "TAD: chr09:13280000-13365000 has 14 genes.\n",
      "TAD: chr09:14010000-14060000 has 6 genes.\n",
      "TAD: chr09:14140000-14185000 has 10 genes.\n",
      "TAD: chr09:14650000-14730000 has 8 genes.\n",
      "TAD: chr09:15300000-15455000 has 19 genes.\n",
      "TAD: chr09:15455000-15495000 has 4 genes.\n",
      "TAD: chr09:15950000-16035000 has 11 genes.\n",
      "TAD: chr09:16385000-16445000 has 9 genes.\n",
      "TAD: chr09:16445000-16515000 has 5 genes.\n",
      "TAD: chr09:16515000-16590000 has 10 genes.\n",
      "TAD: chr09:17635000-17670000 has 5 genes.\n",
      "TAD: chr09:17925000-17975000 has 5 genes.\n",
      "TAD: chr09:18150000-18255000 has 12 genes.\n",
      "TAD: chr09:18510000-18555000 has 7 genes.\n",
      "TAD: chr09:18555000-18620000 has 11 genes.\n",
      "TAD: chr09:19090000-19220000 has 8 genes.\n",
      "TAD: chr09:19265000-19365000 has 13 genes.\n",
      "TAD: chr09:19365000-19455000 has 11 genes.\n",
      "TAD: chr09:19455000-19535000 has 9 genes.\n",
      "TAD: chr09:19810000-19920000 has 16 genes.\n",
      "TAD: chr09:19980000-20057500 has 10 genes.\n",
      "TAD: chr09:20057500-20130000 has 14 genes.\n",
      "TAD: chr09:20130000-20200000 has 13 genes.\n",
      "TAD: chr09:20275000-20375000 has 12 genes.\n",
      "TAD: chr09:20375000-20450000 has 7 genes.\n",
      "TAD: chr09:21365000-21435000 has 13 genes.\n",
      "TAD: chr09:21985000-22020000 has 5 genes.\n",
      "TAD: chr09:22360000-22400000 has 9 genes.\n",
      "TAD: chr09:22400000-22490000 has 15 genes.\n",
      "TAD: chr09:22490000-22610000 has 8 genes.\n",
      "TAD: chr09:2875000-2980000 has 6 genes.\n",
      "TAD: chr09:4300000-4420000 has 7 genes.\n",
      "TAD: chr09:5035000-5185000 has 21 genes.\n",
      "TAD: chr09:6280000-6355000 has 9 genes.\n",
      "TAD: chr09:6355000-6540000 has 16 genes.\n",
      "TAD: chr09:8460000-8515000 has 3 genes.\n",
      "TAD: chr10:10255000-10382500 has 16 genes.\n",
      "TAD: chr10:11740000-11815000 has 6 genes.\n",
      "TAD: chr10:11960000-12040000 has 5 genes.\n",
      "TAD: chr10:12430000-12480000 has 6 genes.\n",
      "TAD: chr10:14245000-14385000 has 12 genes.\n",
      "TAD: chr10:16765000-16840000 has 9 genes.\n",
      "TAD: chr10:17005000-17115000 has 14 genes.\n",
      "TAD: chr10:17290000-17370000 has 8 genes.\n",
      "TAD: chr10:18915000-18980000 has 7 genes.\n",
      "TAD: chr10:19620000-19735000 has 9 genes.\n",
      "TAD: chr10:19735000-19775000 has 5 genes.\n",
      "TAD: chr10:19775000-19865000 has 7 genes.\n",
      "TAD: chr10:19960000-20055000 has 12 genes.\n",
      "TAD: chr10:20345000-20400000 has 8 genes.\n",
      "TAD: chr10:20555000-20600000 has 7 genes.\n",
      "TAD: chr10:20600000-20670000 has 8 genes.\n",
      "TAD: chr10:20935000-20985000 has 9 genes.\n",
      "TAD: chr10:20985000-21045000 has 8 genes.\n",
      "TAD: chr10:21910000-21985000 has 12 genes.\n",
      "TAD: chr10:22025000-22060000 has 4 genes.\n",
      "TAD: chr10:22450000-22525000 has 11 genes.\n",
      "TAD: chr10:22525000-22585000 has 11 genes.\n",
      "TAD: chr10:22755000-22815000 has 8 genes.\n",
      "TAD: chr10:22815000-22885000 has 7 genes.\n",
      "TAD: chr10:22885000-22965000 has 14 genes.\n",
      "TAD: chr10:23105000-23190000 has 10 genes.\n",
      "TAD: chr10:24000000-24035000 has 8 genes.\n",
      "TAD: chr10:24035000-24280000 has 27 genes.\n",
      "TAD: chr10:24405000-24470000 has 6 genes.\n",
      "TAD: chr10:4885000-5135000 has 11 genes.\n",
      "TAD: chr11:10535000-10585000 has 5 genes.\n",
      "TAD: chr11:14045000-14155000 has 13 genes.\n",
      "TAD: chr11:14155000-14315000 has 13 genes.\n",
      "TAD: chr11:142500-285000 has 20 genes.\n",
      "TAD: chr11:1470000-1585000 has 17 genes.\n",
      "TAD: chr11:1585000-1650000 has 8 genes.\n",
      "TAD: chr11:16375000-16470000 has 9 genes.\n",
      "TAD: chr11:17145000-17245000 has 10 genes.\n",
      "TAD: chr11:17245000-17315000 has 7 genes.\n",
      "TAD: chr11:1725000-1855000 has 19 genes.\n",
      "TAD: chr11:17315000-17515000 has 18 genes.\n",
      "TAD: chr11:17765000-18150000 has 21 genes.\n",
      "TAD: chr11:1855000-1935000 has 11 genes.\n",
      "TAD: chr11:18890000-19000000 has 9 genes.\n",
      "TAD: chr11:1975000-2040000 has 5 genes.\n",
      "TAD: chr11:20890000-20965000 has 8 genes.\n",
      "TAD: chr11:2090000-2185000 has 20 genes.\n",
      "TAD: chr11:21500000-21635000 has 9 genes.\n",
      "TAD: chr11:21635000-21715000 has 8 genes.\n",
      "TAD: chr11:2185000-2255000 has 7 genes.\n",
      "TAD: chr11:2280000-2385000 has 16 genes.\n",
      "TAD: chr11:23605000-23665000 has 6 genes.\n",
      "TAD: chr11:24365000-24425000 has 7 genes.\n",
      "TAD: chr11:24425000-24530000 has 19 genes.\n",
      "TAD: chr11:24530000-24605000 has 4 genes.\n",
      "TAD: chr11:25500000-25580000 has 8 genes.\n",
      "TAD: chr11:2565000-2610000 has 8 genes.\n",
      "TAD: chr11:26150000-26220000 has 15 genes.\n",
      "TAD: chr11:26790000-26830000 has 4 genes.\n",
      "TAD: chr11:26830000-26980000 has 17 genes.\n",
      "TAD: chr11:27015000-27120000 has 15 genes.\n",
      "TAD: chr11:3015000-3080000 has 4 genes.\n",
      "TAD: chr11:4395000-4485000 has 13 genes.\n",
      "TAD: chr11:6395000-6490000 has 10 genes.\n",
      "TAD: chr11:65000-142500 has 8 genes.\n",
      "TAD: chr11:690000-845000 has 19 genes.\n",
      "TAD: chr11:6995000-7080000 has 5 genes.\n",
      "TAD: chr11:7080000-7155000 has 6 genes.\n",
      "TAD: chr11:7155000-7230000 has 7 genes.\n",
      "TAD: chr11:7480000-7545000 has 7 genes.\n",
      "TAD: chr11:7750000-7795000 has 3 genes.\n",
      "TAD: chr12:1145000-1255000 has 16 genes.\n",
      "TAD: chr12:1605000-1690000 has 13 genes.\n",
      "TAD: chr12:18340000-18425000 has 8 genes.\n",
      "TAD: chr12:20220000-20285000 has 9 genes.\n",
      "TAD: chr12:2100000-2200000 has 12 genes.\n",
      "TAD: chr12:22350000-22420000 has 7 genes.\n",
      "TAD: chr12:2270000-2350000 has 8 genes.\n",
      "TAD: chr12:22790000-22880000 has 13 genes.\n",
      "TAD: chr12:23030000-23170000 has 20 genes.\n",
      "TAD: chr12:23170000-23290000 has 16 genes.\n",
      "TAD: chr12:23290000-23365000 has 8 genes.\n",
      "TAD: chr12:23465000-23730000 has 29 genes.\n",
      "TAD: chr12:2350000-2560000 has 26 genes.\n",
      "TAD: chr12:23730000-23840000 has 10 genes.\n",
      "TAD: chr12:25445000-25500000 has 7 genes.\n",
      "TAD: chr12:2560000-2605000 has 6 genes.\n",
      "TAD: chr12:26035000-26070000 has 5 genes.\n",
      "TAD: chr12:2605000-2690000 has 12 genes.\n",
      "TAD: chr12:26855000-26950000 has 12 genes.\n",
      "TAD: chr12:26950000-27045000 has 8 genes.\n",
      "TAD: chr12:27120000-27165000 has 8 genes.\n",
      "TAD: chr12:27165000-27215000 has 8 genes.\n",
      "TAD: chr12:2930000-2970000 has 5 genes.\n",
      "TAD: chr12:2970000-3135000 has 15 genes.\n",
      "TAD: chr12:3385000-3445000 has 5 genes.\n",
      "TAD: chr12:3445000-3545000 has 13 genes.\n",
      "TAD: chr12:3545000-3810000 has 30 genes.\n",
      "TAD: chr12:5940000-6220000 has 23 genes.\n",
      "TAD: chr12:6395000-6455000 has 4 genes.\n",
      "TAD: chr12:645000-800000 has 20 genes.\n",
      "TAD: chr12:6455000-6702500 has 24 genes.\n",
      "TAD: chr12:6735000-6800000 has 5 genes.\n",
      "TAD: chr12:6800000-6920000 has 11 genes.\n",
      "TAD: chr12:9935000-10240000 has 10 genes.\n",
      "Median distances calculated and saved to tad_exp_dist.bed.\n"
     ]
    }
   ],
   "source": [
    "#calculate median distance between genes per tad\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "\n",
    "# Load intersected TAD and gene data\n",
    "tad_gene_file = 'tads_exp_genes.bed'\n",
    "output_file = 'tad_exp_dist.bed'\n",
    "\n",
    "# Dictionary to hold TADs and their corresponding gene start-end coordinates\n",
    "tad_genes = defaultdict(list)\n",
    "\n",
    "# Read the TAD-gene intersections\n",
    "with open(tad_gene_file, 'r') as f:\n",
    "    for line in f:\n",
    "        parts = line.strip().split('\\t')\n",
    "        if len(parts) < 6:\n",
    "            print(f\"Skipping line, not enough columns: {line.strip()}\")\n",
    "            continue  # Skip lines that don't have the expected number of columns\n",
    "        \n",
    "        # Extract TAD information and gene coordinates\n",
    "        tad_chr = parts[0]\n",
    "        tad_start = int(parts[1])\n",
    "        tad_end = int(parts[2])\n",
    "        gene_start = int(parts[4])\n",
    "        gene_end = int(parts[5])\n",
    "        \n",
    "        # Group genes by TAD coordinates\n",
    "        tad_genes[(tad_chr, tad_start, tad_end)].append((gene_start, gene_end))\n",
    "\n",
    "# Debug: Print number of TADs found\n",
    "print(f\"Found {len(tad_genes)} unique TADs.\")\n",
    "\n",
    "# Prepare to write output\n",
    "with open(output_file, 'w') as out:\n",
    "    # For each unique TAD, calculate the median distance between genes\n",
    "    for (tad_chr, tad_start, tad_end), genes in tad_genes.items():\n",
    "        # Sort gene starts and ends\n",
    "        gene_starts = sorted(gene[0] for gene in genes)\n",
    "        gene_ends = sorted(gene[1] for gene in genes)\n",
    "\n",
    "        # Debug: Print genes associated with the current TAD\n",
    "        print(f\"TAD: {tad_chr}:{tad_start}-{tad_end} has {len(genes)} genes.\")\n",
    "\n",
    "        # Calculate distances between adjacent genes\n",
    "        distances = []\n",
    "        for i in range(len(gene_starts) - 1):\n",
    "            distance = gene_starts[i + 1] - gene_ends[i]  # distance = gene2_start - gene1_end\n",
    "            if distance > 0:  # Only consider positive distances\n",
    "                distances.append(distance)\n",
    "        \n",
    "        # Calculate median distance\n",
    "        if distances:  # Check if there are distances to calculate median\n",
    "            median_distance = np.median(distances)\n",
    "        else:\n",
    "            median_distance = 0  # If no distances, set to 0 or you can skip this TAD\n",
    "            print(f\"No valid distances for TAD {tad_chr}:{tad_start}-{tad_end}. Setting median distance to 0.\")\n",
    "\n",
    "        # Write the results to the output file\n",
    "        out.write(f\"{tad_chr}\\t{tad_start}\\t{tad_end}\\t{median_distance}\\n\")\n",
    "\n",
    "print(f\"Median distances calculated and saved to {output_file}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5817a27b-0a27-422b-821b-d8bbd65b1777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAq8AAAHUCAYAAAAUbMECAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABUnElEQVR4nO3deVxVdf7H8fdF4bKIuCCbCpLhvuRSptXgEu62OC1mlmbNTzMzU6fGnEkqB9LKbKZRW12mTK3ULNOEVFrUxtRKjcwpFC2RMBcUBJXv7w8fnPGyCQhcjr6ej8d9PLzf873nfO73Hi5vD+d7jsMYYwQAAADYgIe7CwAAAABKi/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/B6iZg/f74cDof18Pb2VkhIiHr06KH4+Hilp6cXek1sbKwcDkeZtpOVlaXY2Fht2LChTK8raltNmjTRwIEDy7SeC1m0aJFmzZpV5DKHw6HY2NgK3V5F+/TTT9W5c2f5+fnJ4XBoxYoV7i6pWN27d1f37t2t53v37pXD4dD8+fPdVlNBxe13I0aMKNN6Nm7cqNjYWB09erTiiqvmunfvrjZt2lToOmfPnl2t9g9UvqK+F0r6eSrt74UNGzbI4XCU+XdRUQr+/izu0aRJE5fXdezYUQ6HQ88//3yp1lua38sonZruLgAVa968eWrRooVOnz6t9PR0ffHFF5o+fbqef/55LVmyRDfeeKPV94EHHlDfvn3LtP6srCw99dRTkuQSXC6kPNsqj0WLFmnnzp0aP358oWWbNm1So0aNKr2G8jLG6I477lCzZs20cuVK+fn5qXnz5u4uq9RCQ0O1adMmNW3a1N2llGj58uWqXbt2mV6zceNGPfXUUxoxYoTq1KlTOYVdBmbPnq3AwMAy/+cB9lXU90JF/Dx17NhRmzZtUqtWrS66xgEDBmjTpk0ubV27dtVtt92miRMnWm1Op9P69zfffKPt27dLkt544w1NmjSp2PWX5fcySofweolp06aNOnfubD3/4x//qEcffVTXX3+9Bg8erD179ig4OFiS1KhRo0oPc1lZWfL19a2SbV3Itdde69btX8ivv/6q33//Xbfeeqt69erl7nLKzOl0VvsxlqQOHTq4uwTgslFZ3wu1a9eusPU2aNBADRo0KNQeHBxc7DZef/11SeeC76pVq7Rx40Z169atyL5l+b2M0uG0gctAeHi4XnjhBWVmZuqVV16x2ov6k+q6devUvXt31a9fXz4+PgoPD9cf//hHZWVlae/evdYP+FNPPWX9KST/KEr++rZt26bbbrtNdevWtf63XdIpCsuXL1e7du3k7e2tK664Qv/4xz9cluf/6WXv3r0u7QX/bNS9e3etWrVK+/btc/lTTb6iThvYuXOnbr75ZtWtW1fe3t666qqrtGDBgiK3884772jKlCkKCwtT7dq1deONN2r37t3FD/x5vvjiC/Xq1Uv+/v7y9fVVt27dtGrVKmt5bGysFe4ff/zxIv9EVVRNixYt0uOPP67Q0FDVqlVLgwYN0qFDh5SZman/+7//U2BgoAIDA3XffffpxIkTLuswxmj27Nm66qqr5OPjo7p16+q2227Tzz//XKjfjBkzFBERIW9vb3Xs2FGrV68uVFNRfx7873//q/vuu09RUVHy9fVVw4YNNWjQIO3YsaPCx3jVqlW66qqr5HQ6FRkZWeyf8gqeNpCXl6dp06apefPm8vHxUZ06ddSuXTu99NJLks59Nn/+858lSZGRkdZ+lb/fLVmyRL1791ZoaKh8fHzUsmVL/eUvf9HJkyddtjtixAjVqlVL//3vf9W/f3/VqlVLjRs31sSJE5WTk+PSNycnR08//bRatmwpb29v1a9fXz169NDGjRutPqX9/LZv366BAwcqKChITqdTYWFhGjBggA4cOFCqcf3888917bXXysfHRw0bNtTf/vY3nT171qVPbm6upk2bphYtWsjpdKpBgwa677779Ntvv7mM+65du5SUlOTyZ1hjjIKDg/XQQw9Zfc+ePau6devKw8NDhw4dstpnzpypmjVruvy5+euvv9ZNN92kevXqydvbWx06dNDSpUsLvY+0tDSNGjVKjRo1kpeXlyIjI/XUU0/pzJkzVp/8ffj555/XzJkzFRkZqVq1aqlr167avHlzqcbriy++UNeuXeXt7W2N1+uvv17kd9iSJUvUtWtX+fn5qVatWurTp491NC9fWfab0nwOUsnf88X585//rICAAJfP/uGHH5bD4dBzzz1ntR0+fFgeHh765z//6TKm+d8LF/p5yrdmzRp17NhRPj4+atGihd58802X5UWdNlCWsboYp06d0qJFi9SpUye9+OKLklSovgsp7vfyzz//rCFDhigsLExOp1PBwcHq1auXvvnmmwqr/5JgcEmYN2+ekWS2bNlS5PITJ06YGjVqmF69elltU6dONefvAikpKcbb29vExMSYFStWmA0bNpi3337b3HPPPebIkSPm1KlTZs2aNUaSuf/++82mTZvMpk2bzH//+1+X9UVERJjHH3/cJCQkmBUrVhS5LWOMiYiIMA0bNjTh4eHmzTffNB9//LG5++67jSTz3HPPFXpvKSkpLq9fv369kWTWr19vjDFm165d5rrrrjMhISFWbZs2bbL6SzJTp061nv/www/G39/fNG3a1CxcuNCsWrXK3HXXXUaSmT59eqHtNGnSxNx9991m1apV5p133jHh4eEmKirKnDlzpsTPZsOGDcbT09N06tTJLFmyxKxYscL07t3bOBwOs3jxYmOMMfv37zfLli0zkszDDz9sNm3aZLZt21bsOvNrioiIMCNGjDBr1qwxc+fONbVq1TI9evQwMTExZtKkSWbt2rVm+vTppkaNGubhhx92Wcef/vQn4+npaSZOnGjWrFljFi1aZFq0aGGCg4NNWlqa1S//s7v//vvN6tWrzauvvmoaNmxoQkJCTHR0tNUvJSXFSDLz5s2z2pKSkszEiRPNe++9Z5KSkszy5cvNLbfcYnx8fMwPP/xQYWOcmJhoatSoYa6//nqzbNky8+6775qrr77ahIeHF7nfDR8+3HoeHx9vatSoYaZOnWo+/fRTs2bNGjNr1iwTGxtrfTYPP/ywkWSWLVtm7VfHjh0zxhjzzDPPmBdffNGsWrXKbNiwwcydO9dERkaaHj16uGx3+PDhxsvLy7Rs2dI8//zzJjEx0Tz55JPG4XCYp556yup3+vRp06NHD1OzZk0zadIk8/HHH5uVK1eaJ554wrzzzjtl+vxOnDhh6tevbzp37myWLl1qkpKSzJIlS8zo0aPN999/X+KYRkdHm/r165uwsDDzj3/8w3zyySdm3LhxRpJ56KGHrH5nz541ffv2NX5+fuapp54yCQkJ5vXXXzcNGzY0rVq1MllZWcYYY7Zt22auuOIK06FDB2sM8/fxIUOGmGbNmlnr3Lx5s5FkfHx8zNtvv2219+vXz1xzzTXW83Xr1hkvLy9zww03mCVLlpg1a9aYESNGFNoPDx48aBo3bmwiIiLMK6+8YhITE80zzzxjnE6nGTFihNUvfx9u0qSJ6du3r1mxYoVZsWKFadu2ralbt645evRoiWP27bffGm9vb9OuXTuzePFis3LlStO/f3/TpEmTQt9hf//7343D4TAjR440H330kVm2bJnp2rWr8fPzM7t27bL6lXa/Ke3ncKHv+eLkf/dv3LjRamvRooXx8fExMTExVtuSJUuMJGv/Kvi9cKGfp4iICNOoUSPTqlUrs3DhQvPJJ5+Y22+/3UgySUlJ1nYKfv+XZaxKo+B+fr63337bSDL/+te/jDHGXH/99aZWrVomMzPTpV95fi83b97cXHnllebf//63SUpKMu+//76ZOHGiy/uEMYTXS8SFfkiMMSY4ONi0bNnSel4wUL733ntGkvnmm2+KXcdvv/1WKAQWXN+TTz5Z7LLzRUREGIfDUWh7MTExpnbt2ubkyZMu7+1C4dUYYwYMGGAiIiKKrL1g3UOGDDFOp9Okpqa69OvXr5/x9fW1flHlb6d///4u/ZYuXWokuQTkolx77bUmKCjI5YvtzJkzpk2bNqZRo0YmLy/PGPO/L/nzg3tx8msaNGiQS/v48eONJDNu3DiX9ltuucXUq1fPer5p0yYjybzwwgsu/fbv3298fHzMY489Zowx5siRI8bb29vceuutLv2+/PJLI+mC4bWgM2fOmNzcXBMVFWUeffTRQu+nvGPcpUsXExYWZrKzs62248ePm3r16l0wvA4cONBcddVVJa7/ueeeK3IfLCgvL8+cPn3aJCUlGUnm22+/tZYNHz7cSDJLly51eU3//v1N8+bNrecLFy40ksxrr71W7HZK+/l9/fXXRpL1n8iyiI6ONpLMBx984NL+pz/9yXh4eJh9+/YZY4x55513jCTz/vvvu/TbsmWLkWRmz55ttbVu3dpln8n3+uuvG0nWz+K0adNMixYtzE033WTuu+8+Y4wxubm5xs/PzzzxxBPW61q0aGE6dOhgTp8+7bK+gQMHmtDQUHP27FljjDGjRo0ytWrVsmrO9/zzzxtJVljM34fbtm3r8h+m//znP0aSy38einL77bcbPz8/89tvv1ltZ8+eNa1atXLZf1JTU03NmjUL/YcyMzPThISEmDvuuMNqK+1+U9rPoTTf80U5efKk8fLyMk8//bQxxpgDBw4YSebxxx83Pj4+5tSpU8aYc/tHWFiY9bqivhdK+nmKiIgw3t7eLp9Vdna2qVevnhk1apTVVlx4Lc1YlUZJ4bVnz57G29vbCvv5v6PeeOMNl35l/b2ckZFhJJlZs2aVqdbLEacNXEaMMSUuv+qqq+Tl5aX/+7//04IFCwr9+bG0/vjHP5a6b+vWrdW+fXuXtqFDh+r48ePatm1bubZfWuvWrVOvXr3UuHFjl/YRI0YoKyur0An8N910k8vzdu3aSZL27dtX7DZOnjypr776Srfddptq1apltdeoUUP33HOPDhw4UOo/ixel4Kzcli1bSjp3HlbB9t9//906deCjjz6Sw+HQsGHDdObMGesREhKi9u3bW3+K27Rpk06dOqW7777bZX3dunVTRETEBes7c+aM4uLi1KpVK3l5ealmzZry8vLSnj17lJycXKh/ecd4y5YtGjx4sLy9va12f39/DRo06II1XnPNNfr22281ZswYffLJJzp+/PgFX3O+n3/+WUOHDlVISIhq1KghT09PRUdHS1Kh9+hwOArV1K5dO5f3t3r1anl7e2vkyJHFbrO0n9+VV16punXr6vHHH9fcuXP1/fffl+m9+fv7F/pMhg4dqry8PH322WdWLXXq1NGgQYNcarnqqqsUEhJSqtng+RNWEhMTJUkJCQmKiYnRjTfeqISEBEnn9sWTJ09aff/73//qhx9+sPbN87fdv39/HTx40PrZ+uijj9SjRw+FhYW59OvXr58kKSkpyaWeAQMGqEaNGtbz0uyH+evp2bOnAgMDrTYPDw/dcccdLv0++eQTnTlzRvfee69LPd7e3oqOji40ZqXZb0r7OZT3e97X11ddu3Z1+Yzq1KmjP//5z8rNzdUXX3wh6dxneLETkK666iqFh4dbz729vdWsWbMLjr9UurG6GCkpKVq/fr0GDx5sTTa7/fbb5e/vX+ZTByTX38v16tVT06ZN9dxzz2nmzJnavn278vLyKqTuSw3h9TJx8uRJHT58WGFhYcX2adq0qRITExUUFKSHHnpITZs2VdOmTa1z/0orNDS01H1DQkKKbTt8+HCZtltWhw8fLrLW/DEquP369eu7PM+feZqdnV3sNo4cOSJjTJm2Uxb16tVzee7l5VVi+6lTpyRJhw4dss419PT0dHls3rxZGRkZLrWV9DmVZMKECfrb3/6mW265RR9++KG++uorbdmyRe3bty9y3Mo7xnl5eeWucfLkyXr++ee1efNm9evXT/Xr11evXr309ddfX/C1J06c0A033KCvvvpK06ZN04YNG7RlyxYtW7asyLp9fX1dAnb+e8z/XCTpt99+U1hYmDw8iv96Lu3nFxAQoKSkJF111VV64okn1Lp1a4WFhWnq1Kk6ffr0Bd9fUZNICv58Hjp0SEePHpWXl1ehWtLS0qxaShIREWF9/+T/xzE/vOb/By8xMVE+Pj7WpJj8c2EnTZpUaLtjxoyRJGvbhw4d0ocfflioX+vWrV365SvPfpg/JkWNWcG2/NqvvvrqQjUtWbKkUD2l2W9K+zlczPf8jTfeqM2bN+vkyZNKTExUz549Vb9+fXXq1EmJiYlKSUlRSkrKRYfXguOf/34vNP5S6cbqYrz55psyxui2227T0aNHdfToUZ0+fVo33XSTvvzyS/3www+lXlfB38sOh0Offvqp+vTpoxkzZqhjx45q0KCBxo0bp8zMzAqp/1LB1QYuE6tWrdLZs2cveHmrG264QTfccIPOnj2rr7/+Wv/85z81fvx4BQcHa8iQIaXaVlmuHZuWllZsW/4XWP4XUcET7kvzS7Ek9evX18GDBwu1//rrr5LkcvSkvPInnVT2dsoqMDBQDodDn3/+ucvlX/Llt+V/BsV9TiVNKpOkt956S/fee6/i4uJc2jMyMirsklN169aVw+EocV8qSc2aNTVhwgRNmDBBR48eVWJiop544gn16dNH+/fvl6+vb7GvXbdunX799Vdt2LDBOtoq6aKuB9ugQQN98cUXysvLKzbAlvbzk6S2bdtq8eLFMsbou+++0/z58/X000/Lx8dHf/nLX0qs5fzJUvkK/nwGBgaqfv36WrNmTZHr8Pf3L3Eb+Xr16qUPPvhASUlJysvLU/fu3eXv76+wsDAlJCQoMTFRN9xwg/Xe8n9uJk+erMGDBxe5zvxLzQUGBqpdu3b6+9//XmS/kv5TXxb169cvcczy5df+3nvvleovGKVRls+hvN/zvXr10t/+9jd99tln+vTTTzV16lSrfe3atYqMjLSeX4ry8vKsiWfF7XNvvvmmZsyYUar1FfV7OSIiQm+88YYk6ccff9TSpUsVGxur3NxczZ0796Lqv5Rw5PUykJqaqkmTJikgIECjRo0q1Wtq1KihLl266F//+pckWX/CL+0RiNLatWuXvv32W5e2RYsWyd/fXx07dpQkKyB99913Lv1WrlxZaH2l/d+5dO4LNj98nG/hwoXy9fWtkMuw+Pn5qUuXLlq2bJlLXXl5eXrrrbfUqFEjNWvW7KK3U1YDBw6UMUa//PKLOnfuXOjRtm1bSecuL+bt7a23337b5fUbN24s9Z/wCoarVatW6Zdffqmw9+Ln56drrrlGy5Ytczm6kpmZqQ8//LBM66pTp45uu+02PfTQQ/r999+t2eHF7ff5/1Er+B7Pnz1cVv369dOpU6dKvJh/aT+/grW2b99eL774ourUqVOq03IyMzML/ZwtWrRIHh4e+sMf/mDVcvjwYZ09e7bIWs6/VnFJP5833nijDh06pFmzZunaa6+1wlavXr20fPlybdmyxeWIXvPmzRUVFaVvv/22yO127tzZWsfAgQO1c+dONW3atMh+FRVeo6OjtW7dOpf/WOfl5endd9916denTx/VrFlTP/30U7G1l1VZPod8xX3PF+eaa65R7dq1NWvWLKWlpSkmJkbSuc9u+/btWrp0qVq1anXB8azo3yNV5ZNPPtGBAwf00EMPaf369YUerVu31sKFC12uYFGc0vxebtasmf7617+qbdu2lX4and1w5PUSs3PnTutcp/T0dH3++eeaN2+eatSooeXLlxd5Lbt8c+fO1bp16zRgwACFh4fr1KlT1jk8+b80/P39FRERoQ8++EC9evVSvXr1FBgYeMEjcMUJCwvTTTfdpNjYWIWGhuqtt95SQkKCpk+fbh3xuvrqq9W8eXNNmjRJZ86cUd26dbV8+XLrHKvztW3bVsuWLdOcOXPUqVMneXh4FPuLYOrUqda5cE8++aTq1aunt99+W6tWrdKMGTMUEBBQrvdUUHx8vGJiYtSjRw9NmjRJXl5emj17tnbu3Kl33nmnzHc5qwjXXXed/u///k/33Xefvv76a/3hD3+Qn5+fDh48qC+++EJt27bVgw8+qLp162rSpEmaNm2aHnjgAd1+++3av3+/YmNjS/Un+YEDB2r+/Plq0aKF2rVrp61bt+q5556r8Gv+PvPMM+rbt69iYmI0ceJEnT17VtOnT5efn59+//33El87aNAg6zqMDRo00L59+zRr1ixFREQoKipKkqww+NJLL2n48OHy9PRU8+bN1a1bN9WtW1ejR4/W1KlT5enpqbfffrvQf8jK4q677tK8efM0evRo7d69Wz169FBeXp6++uortWzZUkOGDCn15/fRRx9p9uzZuuWWW3TFFVfIGKNly5bp6NGjVvAoSf369fXggw8qNTVVzZo108cff6zXXntNDz74oHVO4pAhQ/T222+rf//+euSRR3TNNdfI09NTBw4c0Pr163XzzTfr1ltvtcZx8eLFWrJkia644gp5e3tbY9uzZ085HA6tXbvWuhGKdO67Z/jw4da/z/fKK6+oX79+6tOnj0aMGKGGDRvq999/V3JysrZt22aFxqeffloJCQnq1q2bxo0bp+bNm+vUqVPau3evPv74Y82dO7dC9skpU6boww8/VK9evTRlyhT5+Pho7ty51mXT8o+kN2nSRE8//bSmTJmin3/+WX379lXdunV16NAh/ec//5Gfn5/LGJRGaT+H0nzPF6dGjRqKjo7Whx9+qMjISOtSiNddd52cTqc+/fRTjRs37oK1FvfzVNqj9O7yxhtvqGbNmnriiSeKDOijRo3SuHHjtGrVKt18881We2l/L3/33XcaO3asbr/9dkVFRcnLy0vr1q3Td999d8G/klx23DVTDBUrf1Zj/sPLy8sEBQWZ6OhoExcXZ9LT0wu9puAVADZt2mRuvfVWExERYZxOp6lfv76Jjo42K1eudHldYmKi6dChg3E6nUaSNXM7f33nz7QtblvGnJtVOmDAAPPee++Z1q1bGy8vL9OkSRMzc+bMQq//8ccfTe/evU3t2rVNgwYNzMMPP2xWrVpVaLbp77//bm677TZTp04d43A4XLapIq6SsGPHDjNo0CATEBBgvLy8TPv27QvNls+f1fruu++6tJdmdn2+zz//3PTs2dP4+fkZHx8fc+2115oPP/ywyPWV5WoDBWsqbnZrcZ/Nm2++abp06WLV1bRpU3Pvvfear7/+2uqTl5dn4uPjTePGjY2Xl5dp166d+fDDD010dPQFrzZw5MgRc//995ugoCDj6+trrr/+evP5558Xem1FjPHKlStNu3btjJeXlwkPDzfPPvtssfvd+VcbeOGFF0y3bt1MYGCg9dr777/f7N271+V1kydPNmFhYcbDw8Nlv9u4caPp2rWr8fX1NQ0aNDAPPPCA2bZtW6G6hw8fbvz8/ArVXVSN2dnZ5sknnzRRUVHGy8vL1K9f3/Ts2dPlMkXGXPjz++GHH8xdd91lmjZtanx8fExAQIC55pprzPz58y84ntHR0aZ169Zmw4YNpnPnzsbpdJrQ0FDzxBNPFJrdf/r0afP888+b9u3bG29vb1OrVi3TokULM2rUKLNnzx6r3969e03v3r2Nv7+/dam383Xo0MFIMl9++aXV9ssvvxhJpn79+taVOc737bffmjvuuMMEBQUZT09PExISYnr27Gnmzp3r0u+3334z48aNM5GRkcbT09PUq1fPdOrUyUyZMsWcOHHCGFPyz2BR3x9F+fzzz02XLl2M0+k0ISEh5s9//rOZPn26kVToUlsrVqwwPXr0MLVr1zZOp9NERESY2267zSQmJlp9yrLflOZzKO33fHFeeuklI8n86U9/cmmPiYkxkgqtp7if4eJ+nvJ/LxRU3HdGwasNlHasLkQFrjbw22+/GS8vL3PLLbcU+5ojR44YHx8f6yowZf29fOjQITNixAjTokUL4+fnZ2rVqmXatWtnXnzxxQteLvBy4zDmAlPQAQBAufXu3Vt79+7Vjz/+6O5SgEsCpw0AAFBBJkyYoA4dOqhx48b6/fff9fbbbyshIcGahAPg4hFeAQCoIGfPntWTTz6ptLQ0ORwOtWrVSv/+9781bNgwd5cGXDI4bQAAAAC2waWyAAAAYBuEVwAAANgG4RUAAAC2cclP2MrLy9Ovv/4qf39/t1wMHgAAACUzxigzM1NhYWHF3ho73yUfXn/99Vc1btzY3WUAAADgAvbv33/BO95d8uE1/3Zz+/fvV+3atd1cDQAAAAo6fvy4GjduXKrbBF/y4TX/VIHatWsTXgEAAKqx0pziyYQtAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALbh1vDapEkTORyOQo+HHnpI0rm7LcTGxiosLEw+Pj7q3r27du3a5c6SAQAA4EZuDa9btmzRwYMHrUdCQoIk6fbbb5ckzZgxQzNnztTLL7+sLVu2KCQkRDExMcrMzHRn2QAAAHATt4bXBg0aKCQkxHp89NFHatq0qaKjo2WM0axZszRlyhQNHjxYbdq00YIFC5SVlaVFixa5s2wAAAC4SbU55zU3N1dvvfWWRo4cKYfDoZSUFKWlpal3795WH6fTqejoaG3cuLHY9eTk5Oj48eMuDwAAAFwaqk14XbFihY4ePaoRI0ZIktLS0iRJwcHBLv2Cg4OtZUWJj49XQECA9WjcuHGl1QwAAICqVW3C6xtvvKF+/fopLCzMpb3gPW6NMSXe93by5Mk6duyY9di/f3+l1AsAAICqV9PdBUjSvn37lJiYqGXLllltISEhks4dgQ0NDbXa09PTCx2NPZ/T6ZTT6ay8YgEAAOA21eLI67x58xQUFKQBAwZYbZGRkQoJCbGuQCCdOy82KSlJ3bp1c0eZAAAAcDO3H3nNy8vTvHnzNHz4cNWs+b9yHA6Hxo8fr7i4OEVFRSkqKkpxcXHy9fXV0KFD3VixfaSmpiojI6NQe2BgoMLDw91QEQAAwMVxe3hNTExUamqqRo4cWWjZY489puzsbI0ZM0ZHjhxRly5dtHbtWvn7+7uhUntJTU1Vi5YtlZ2VVWiZj6+vfkhOJsACAADbcRhjjLuLqEzHjx9XQECAjh07ptq1a7u7nCqzbds2derUSXdMm6OgyCirPT1lj5b+9UFt3bpVHTt2dGOFAAAA55Qlr7n9yCsqV1BklBq2bO/uMgAAACpEtZiwBQAAAJQG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbXCd12qKW7sCAAAURnithri1KwAAQNEIr9VQRkaGsrOyir21a0ZGBuEVAABclgiv1Ri3dgUAAHDFhC0AAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbNd1dAMouOTnZ5XlgYKDCw8PdVA0AAEDVIbzaSGbGITk8PDRs2DCXdh9fX/2QnEyABQAAlzy3nzbwyy+/aNiwYapfv758fX111VVXaevWrdZyY4xiY2MVFhYmHx8fde/eXbt27XJjxe6TnXlcJi9Pd0ybo7FvJ2rs24m6Y9ocZWdlKSMjw93lAQAAVDq3htcjR47ouuuuk6enp1avXq3vv/9eL7zwgurUqWP1mTFjhmbOnKmXX35ZW7ZsUUhIiGJiYpSZmem+wt0sKDJKDVu2V8OW7RUUGeXucgAAAKqMW08bmD59uho3bqx58+ZZbU2aNLH+bYzRrFmzNGXKFA0ePFiStGDBAgUHB2vRokUaNWpUVZcMAAAAN3LrkdeVK1eqc+fOuv322xUUFKQOHTrotddes5anpKQoLS1NvXv3ttqcTqeio6O1cePGIteZk5Oj48ePuzwAAABwaXBreP355581Z84cRUVF6ZNPPtHo0aM1btw4LVy4UJKUlpYmSQoODnZ5XXBwsLWsoPj4eAUEBFiPxo0bV+6bAAAAQJVxa3jNy8tTx44dFRcXpw4dOmjUqFH605/+pDlz5rj0czgcLs+NMYXa8k2ePFnHjh2zHvv376+0+gEAAFC13BpeQ0ND1apVK5e2li1bKjU1VZIUEhIiSYWOsqanpxc6GpvP6XSqdu3aLg8AAABcGtwaXq+77jrt3r3bpe3HH39URESEJCkyMlIhISFKSEiwlufm5iopKUndunWr0loBAADgfm692sCjjz6qbt26KS4uTnfccYf+85//6NVXX9Wrr74q6dzpAuPHj1dcXJyioqIUFRWluLg4+fr6aujQoe4sHQAAAG7g1vB69dVXa/ny5Zo8ebKefvppRUZGatasWbr77rutPo899piys7M1ZswYHTlyRF26dNHatWvl7+/vxsoBAADgDm6/PezAgQM1cODAYpc7HA7FxsYqNja26ooCAABAteT228MCAAAApUV4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG3UdHcBcI/k5GSX54GBgQoPD3dTNQAAAKVDeL3MZGYcksPDQ8OGDXNp9/H11Q/JyQRYAABQrRFeLzPZmcdl8vJ0x7Q5CoqMkiSlp+zR0r8+qIyMDMIrAACo1givl6mgyCg1bNne3WUAAACUCRO2AAAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbbg1vMbGxsrhcLg8QkJCrOXGGMXGxiosLEw+Pj7q3r27du3a5caKAQAA4E5uP/LaunVrHTx40Hrs2LHDWjZjxgzNnDlTL7/8srZs2aKQkBDFxMQoMzPTjRUDAADAXdweXmvWrKmQkBDr0aBBA0nnjrrOmjVLU6ZM0eDBg9WmTRstWLBAWVlZWrRokZurBgAAgDu4Pbzu2bNHYWFhioyM1JAhQ/Tzzz9LklJSUpSWlqbevXtbfZ1Op6Kjo7Vx48Zi15eTk6Pjx4+7PAAAAHBpcGt47dKlixYuXKhPPvlEr732mtLS0tStWzcdPnxYaWlpkqTg4GCX1wQHB1vLihIfH6+AgADr0bhx40p9DwAAAKg6bg2v/fr10x//+Ee1bdtWN954o1atWiVJWrBggdXH4XC4vMYYU6jtfJMnT9axY8esx/79+yuneAAAAFQ5t582cD4/Pz+1bdtWe/bssa46UPAoa3p6eqGjsedzOp2qXbu2ywMAAACXhmoVXnNycpScnKzQ0FBFRkYqJCRECQkJ1vLc3FwlJSWpW7dubqwSAAAA7lLTnRufNGmSBg0apPDwcKWnp2vatGk6fvy4hg8fLofDofHjxysuLk5RUVGKiopSXFycfH19NXToUHeWDQAAADdxa3g9cOCA7rrrLmVkZKhBgwa69tprtXnzZkVEREiSHnvsMWVnZ2vMmDE6cuSIunTporVr18rf39+dZQMAAMBN3BpeFy9eXOJyh8Oh2NhYxcbGVk1BAAAAqNaq1TmvAAAAQEkIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA23DrTQpQcZKTk0t8DgAAcCkgvNpcZsYhOTw8NGzYMHeXAgAAUOkIrzaXnXlcJi9Pd0ybo6DIKKt995efKmF2vBsrAwAAqHiE10tEUGSUGrZsbz1PT9njxmoAAAAqBxO2AAAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbdR0dwGXu9TUVGVkZLi0JScnu6kaAACA6o3w6kapqalq0bKlsrOy3F0KAACALRBe3SgjI0PZWVm6Y9ocBUVGWe27v/xUCbPj3VgZAABA9UR4rQaCIqPUsGV763l6yh43VgMAAFB9MWELAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG2UK7ympKRUdB0AAADABZUrvF555ZXq0aOH3nrrLZ06daqiawIAAACKVK7w+u2336pDhw6aOHGiQkJCNGrUKP3nP/+p6NoAAAAAF+UKr23atNHMmTP1yy+/aN68eUpLS9P111+v1q1ba+bMmfrtt98quk4AAADg4iZs1axZU7feequWLl2q6dOn66efftKkSZPUqFEj3XvvvTp48GBF1QkAAABcXHj9+uuvNWbMGIWGhmrmzJmaNGmSfvrpJ61bt06//PKLbr755oqqEwAAAFDN8rxo5syZmjdvnnbv3q3+/ftr4cKF6t+/vzw8zmXhyMhIvfLKK2rRokWFFgsAAIDLW7nC65w5czRy5Ejdd999CgkJKbJPeHi43njjjYsqDgAAADhfuU4b2LNnjyZPnlxscJUkLy8vDR8+vNTrjI+Pl8Ph0Pjx4602Y4xiY2MVFhYmHx8fde/eXbt27SpPyQAAALgElCu8zps3T++++26h9nfffVcLFiwo8/q2bNmiV199Ve3atXNpnzFjhmbOnKmXX35ZW7ZsUUhIiGJiYpSZmVmesgEAAGBz5Qqvzz77rAIDAwu1BwUFKS4urkzrOnHihO6++2699tprqlu3rtVujNGsWbM0ZcoUDR48WG3atNGCBQuUlZWlRYsWladsAAAA2Fy5wuu+ffsUGRlZqD0iIkKpqallWtdDDz2kAQMG6MYbb3RpT0lJUVpamnr37m21OZ1ORUdHa+PGjcWuLycnR8ePH3d5AAAA4NJQrvAaFBSk7777rlD7t99+q/r165d6PYsXL9a2bdsUHx9faFlaWpokKTg42KU9ODjYWlaU+Ph4BQQEWI/GjRuXuh4AAABUb+UKr0OGDNG4ceO0fv16nT17VmfPntW6dev0yCOPaMiQIaVax/79+/XII4/orbfekre3d7H9HA6Hy3NjTKG2802ePFnHjh2zHvv37y/dmwIAAEC1V65LZU2bNk379u1Tr169VLPmuVXk5eXp3nvvLfU5r1u3blV6ero6depktZ09e1afffaZXn75Ze3evVvSuSOwoaGhVp/09PRCR2PP53Q65XQ6y/O2AAAAUM2VK7x6eXlpyZIleuaZZ/Ttt9/Kx8dHbdu2VURERKnX0atXL+3YscOl7b777lOLFi30+OOP64orrlBISIgSEhLUoUMHSVJubq6SkpI0ffr08pQNAAAAmytXeM3XrFkzNWvWrFyv9ff3V5s2bVza/Pz8VL9+fat9/PjxiouLU1RUlKKiohQXFydfX18NHTr0YsoGAACATZUrvJ49e1bz58/Xp59+qvT0dOXl5bksX7duXYUU99hjjyk7O1tjxozRkSNH1KVLF61du1b+/v4Vsn4AAADYS7nC6yOPPKL58+drwIABatOmTYkTqMpiw4YNLs8dDodiY2MVGxtbIesHAACAvZUrvC5evFhLly5V//79K7oeAAAAoFjlulSWl5eXrrzyyoquBQAAAChRucLrxIkT9dJLL8kYU9H1AAAAAMUq12kDX3zxhdavX6/Vq1erdevW8vT0dFm+bNmyCikOAAAAOF+5wmudOnV06623VnQtAAAAQInKFV7nzZtX0XUAAAAAF1Suc14l6cyZM0pMTNQrr7yizMxMSdKvv/6qEydOVFhxAAAAwPnKdeR137596tu3r1JTU5WTk6OYmBj5+/trxowZOnXqlObOnVvRdQIAAADlO/L6yCOPqHPnzjpy5Ih8fHys9ltvvVWffvpphRUHAAAAnK/cVxv48ssv5eXl5dIeERGhX375pUIKAwAAAAoq15HXvLw8nT17tlD7gQMH5O/vf9FFAQAAAEUpV3iNiYnRrFmzrOcOh0MnTpzQ1KlTuWUsAAAAKk25Tht48cUX1aNHD7Vq1UqnTp3S0KFDtWfPHgUGBuqdd96p6BoBAAAASeUMr2FhYfrmm2/0zjvvaNu2bcrLy9P999+vu+++22UCF1ylpqYqIyPDep6cnOzGakqnYM35AgMDFR4e7oaKAADA5axc4VWSfHx8NHLkSI0cObIi67lkpaamqkXLlsrOynJ3KaVWUs0+vr76ITmZAAsAAKpUucLrwoULS1x+7733lquYS1lGRoays7J0x7Q5CoqMkiTt/vJTJcyOd3NlxSuqZklKT9mjpX99UBkZGYRXAABQpcoVXh955BGX56dPn1ZWVpa8vLzk6+tLeC1BUGSUGrZsL+lcCLSD82sGAABwp3JdbeDIkSMujxMnTmj37t26/vrrmbAFAACASlOu8FqUqKgoPfvss4WOygIAAAAVpcLCqyTVqFFDv/76a0WuEgAAALCU65zXlStXujw3xujgwYN6+eWXdd1111VIYQAAAEBB5Qqvt9xyi8tzh8OhBg0aqGfPnnrhhRcqoi4AAACgkHKF17y8vIquAwAAALigCj3nFQAAAKhM5TryOmHChFL3nTlzZnk2AQAAABRSrvC6fft2bdu2TWfOnFHz5s0lST/++KNq1Kihjh07Wv0cDkfFVAkAAAConOF10KBB8vf314IFC1S3bl1J525ccN999+mGG27QxIkTK7RIAAAAQCrnOa8vvPCC4uPjreAqSXXr1tW0adO42gAAAAAqTbnC6/Hjx3Xo0KFC7enp6crMzLzoogAAAICilCu83nrrrbrvvvv03nvv6cCBAzpw4IDee+893X///Ro8eHBF1wgAAABIKuc5r3PnztWkSZM0bNgwnT59+tyKatbU/fffr+eee65CCwQAAADylSu8+vr6avbs2Xruuef0008/yRijK6+8Un5+fhVdHwAAAGC5qJsUHDx4UAcPHlSzZs3k5+cnY0xF1QUAAAAUUq7wevjwYfXq1UvNmjVT//79dfDgQUnSAw88wGWyAAAAUGnKFV4fffRReXp6KjU1Vb6+vlb7nXfeqTVr1lRYcQAAAMD5ynXO69q1a/XJJ5+oUaNGLu1RUVHat29fhRSG6i85OdnleWBgoMLDw91UDQAAuByUK7yePHnS5YhrvoyMDDmdzosuCtVbZsYhOTw8NGzYMJd2H19f/ZCcTIAFAACVplynDfzhD3/QwoULrecOh0N5eXl67rnn1KNHjworDtVTduZxmbw83TFtjsa+naixbyfqjmlzlJ2VpYyMDHeXBwAALmHlOvL63HPPqXv37vr666+Vm5urxx57TLt27dLvv/+uL7/8sqJrRDUVFBmlhi3bu7sMAABwGSnXkddWrVrpu+++0zXXXKOYmBidPHlSgwcP1vbt29W0adOKrhEAAACQVI4jr6dPn1bv3r31yiuv6KmnnqqMmuAmBSdgFXwOAADgbmUOr56entq5c6ccDkdl1AM3KG4CFgAAQHVTrtMG7r33Xr3xxhsXvfE5c+aoXbt2ql27tmrXrq2uXbtq9erV1nJjjGJjYxUWFiYfHx91795du3btuujtwlVRE7DGvp2omDGT3V0aAACAi3JN2MrNzdXrr7+uhIQEde7cWX5+fi7LZ86cWar1NGrUSM8++6yuvPJKSdKCBQt08803a/v27WrdurVmzJihmTNnav78+WrWrJmmTZummJgY7d69W/7+/uUpHSUoOAErPWWPG6sBAAAorEzh9eeff1aTJk20c+dOdezYUZL0448/uvQpy+kEgwYNcnn+97//XXPmzNHmzZvVqlUrzZo1S1OmTNHgwYMlnQu3wcHBWrRokUaNGlWW0gEAAHAJKFN4jYqK0sGDB7V+/XpJ524H+49//EPBwcEXXcjZs2f17rvv6uTJk+ratatSUlKUlpam3r17W32cTqeio6O1cePGYsNrTk6OcnJyrOfHjx+/6NoAAABQPZTpnFdjjMvz1atX6+TJkxdVwI4dO1SrVi05nU6NHj1ay5cvV6tWrZSWliZJhYJxcHCwtawo8fHxCggIsB6NGze+qPoAAABQfZRrwla+gmG2PJo3b65vvvlGmzdv1oMPPqjhw4fr+++/t5YXPA3BGFPiqQmTJ0/WsWPHrMf+/fsvukYAAABUD2U6bcDhcBQKjhd7ySwvLy9rwlbnzp21ZcsWvfTSS3r88cclSWlpaQoNDbX6p6enl3iagtPplNPpvKiaAAAAUD2VKbwaYzRixAgrHJ46dUqjR48udLWBZcuWlbsgY4xycnIUGRmpkJAQJSQkqEOHDpLOXeUgKSlJ06dPL/f6AQAAYF9lCq/Dhw93eX6xF7V/4okn1K9fPzVu3FiZmZlavHixNmzYoDVr1sjhcGj8+PGKi4tTVFSUoqKiFBcXJ19fXw0dOvSitgsAAAB7KlN4nTdvXoVu/NChQ7rnnnt08OBBBQQEqF27dlqzZo1iYmIkSY899piys7M1ZswYHTlyRF26dNHatWu5xisAAMBlqlw3KagoF7pLl8PhUGxsrGJjY6umIAAAAFRrF3W1AQAAAKAqEV4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgGzXdXQAufampqcrIyHBpCwwMVHh4uJsqAgAAdkV4RaVKTU1Vi5YtlZ2V5dLu4+urH5KTCbAAAKBMCK+oVBkZGcrOytId0+YoKDJKkpSeskdL//qgMjIyCK8AAKBMCK+oEkGRUWrYsr27ywAAADbHhC0AAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtlHT3QXg0pKcnFzi84qUmpqqjIyMQu2BgYEKDw+vtO0CAAD3IbyiQmRmHJLDw0PDhg2rku2lpqaqRcuWys7KKrTMx9dXPyQnE2ABALgEEV5RIbIzj8vk5emOaXMUFBllte/+8lMlzI6v8O1lZGQoOyur0PbSU/Zo6V8fVEZGBuEVAIBLkFvPeY2Pj9fVV18tf39/BQUF6ZZbbtHu3btd+hhjFBsbq7CwMPn4+Kh79+7atWuXmyrGhQRFRqlhy/bWo25Y5QbIgts7P8gCAIBLj1vDa1JSkh566CFt3rxZCQkJOnPmjHr37q2TJ09afWbMmKGZM2fq5Zdf1pYtWxQSEqKYmBhlZma6sXIAAAC4g1tPG1izZo3L83nz5ikoKEhbt27VH/7wBxljNGvWLE2ZMkWDBw+WJC1YsEDBwcFatGiRRo0aVWidOTk5ysnJsZ4fP368ct8EAAAAqky1ulTWsWPHJEn16tWTJKWkpCgtLU29e/e2+jidTkVHR2vjxo1FriM+Pl4BAQHWo3HjxpVfOAAAAKpEtQmvxhhNmDBB119/vdq0aSNJSktLkyQFBwe79A0ODraWFTR58mQdO3bMeuzfv79yCwcAAECVqTZXGxg7dqy+++47ffHFF4WWORwOl+fGmEJt+ZxOp5xOZ6XUCAAAAPeqFkdeH374Ya1cuVLr169Xo0aNrPaQkBBJKnSUNT09vdDRWAAAAFz63BpejTEaO3asli1bpnXr1ikyMtJleWRkpEJCQpSQkGC15ebmKikpSd26davqcgEAAOBmbj1t4KGHHtKiRYv0wQcfyN/f3zrCGhAQIB8fHzkcDo0fP15xcXGKiopSVFSU4uLi5Ovrq6FDh7qzdFSAom4dy61dAQBASdwaXufMmSNJ6t69u0v7vHnzNGLECEnSY489puzsbI0ZM0ZHjhxRly5dtHbtWvn7+1dxtagoJd1Kllu7AgCAkrg1vBpjLtjH4XAoNjZWsbGxlV8QqkRxt5Ll1q4AAOBCqs3VBnD5yb+1KwAAQGlVi6sNAAAAAKVBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBt1HR3AUBBycnJhdoCAwMVHh7uhmoAAEB1QnhFtZGZcUgODw8NGzas0DIfX1/9kJxMgAUA4DJHeEW1kZ15XCYvT3dMm6OgyCirPT1lj5b+9UFlZGQQXgEAuMwRXlHtBEVGqWHL9u4uAwAAVENM2AIAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbXeYVtnH/b2KJuIQsAAC59hFdUeyXdNhYAAFxeCK+o9oq6bezuLz9Vwux4N1cGAACqGuEVtnH+bWPTU/a4uRoAAOAOTNgCAACAbRBeAQAAYBuEVwAAANgG4RUAAAC24dbw+tlnn2nQoEEKCwuTw+HQihUrXJYbYxQbG6uwsDD5+Pioe/fu2rVrl3uKBQAAgNu5NbyePHlS7du318svv1zk8hkzZmjmzJl6+eWXtWXLFoWEhCgmJkaZmZlVXCkAAACqA7deKqtfv37q169fkcuMMZo1a5amTJmiwYMHS5IWLFig4OBgLVq0SKNGjarKUgEAAFANVNtzXlNSUpSWlqbevXtbbU6nU9HR0dq4cWOxr8vJydHx48ddHgAAALg0VNvwmpaWJkkKDg52aQ8ODraWFSU+Pl4BAQHWo3HjxpVaJwAAAKpOtQ2v+RwOh8tzY0yhtvNNnjxZx44dsx779++v7BIBAABQRart7WFDQkIknTsCGxoaarWnp6cXOhp7PqfTKafTWen1AQAAoOpV2yOvkZGRCgkJUUJCgtWWm5urpKQkdevWzY2VAQAAwF3ceuT1xIkT+u9//2s9T0lJ0TfffKN69eopPDxc48ePV1xcnKKiohQVFaW4uDj5+vpq6NChbqwadpCcnFyoLTAwUOHh4W6oBgAAVBS3htevv/5aPXr0sJ5PmDBBkjR8+HDNnz9fjz32mLKzszVmzBgdOXJEXbp00dq1a+Xv7++uklHNZWYcksPDQ8OGDSu0zMfXVz8kJxNgAQCwMbeG1+7du8sYU+xyh8Oh2NhYxcbGVl1RsLXszOMyeXm6Y9ocBUVGWe3pKXu09K8PKiMjg/AKAICNVdsJW8DFCIqMUsOW7d1dBgAAqGDVdsIWAAAAUBBHXnFZKTiRKycnp8hLqxXXXpZJX6mpqcrIyCj36wEAQGGEV1wWipvI5fDwkMnLK9S/uPbSTvpKTU1Vi5YtlZ2VVa7XAwCAohFecVkoaiLX7i8/VcLs+EKTu4prL8ukr4yMDGVnZbmsg0ljAABcPMIrLivnT+RKT9lTqK2k9ovdHgAAuHhM2AIAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbXea0ERd0WtOBtSXHp4PMGAKDqEF4rWHG3BcWlic8bAICqRXitYEXdFlT63y1HcWnh8wYAoGoRXitJcbccxaWJzxsAgKrBhC0AAADYBkdegTI6fzIWE7MAAKhahFeglDIzDsnh4aFhw4a5uxQAAC5bhFeglLIzj8vk5blMzmJiFgAAVYvwCpTR+ZOzmJgFAEDVYsIWAAAAbIMjrwDKpag7i0lSYGCgwsPD3VARAOByQHgFUGYl3VnMx9dXPyQnE2ABAJWC8AqgzIq7s1h6yh4t/euDysjIILwCACoF4RVAuRW8sxgAAJWNCVsAAACwDcIrAAAAbIPTBoBqgJn7AACUDuEVcDNm7gMAUHqEV8DNmLkPAEDpEV6BaoKZ+wAAXBgTtgAAAGAbHHkFqlhycnKJzy+krJO7iuqfk5Mjp9NZqG9x7ZU1cay491JUHUxeAwBIhFegymRmHJLDw0PDhg0r9zrKOrmruP4ODw+ZvLxC6yiuvTImjpX0Xoqqg8lrAACJ8ApUmezM4zJ5eYUmZu3+8lMlzI4v1TrKOrmrqP752yuujqqaOFbceymqDiavAQDyEV6BKlZwYlZ6yp6LXkdZ+udvr7g6qnriWHWpAwBgD0zYAgAAgG1w5BWo5s6f0FXWyV2XYh04pywT9yriDm7V/S5wRdVX1bVV9zECimO3fZfwClRTFTHB61KqA/9Tlol7FXEHt+p+F7ji6qvK2qr7GAHFseO+S3gFqqmiJniVZXLXpVYH/qcsE/cq4g5u1f0ucEXVV9W1VfcxAopjx33XFuF19uzZeu6553Tw4EG1bt1as2bN0g033ODusoAqUdRkq8u5DvxPWSa1VcQEuOo+ia461FcdagDKw077brWfsLVkyRKNHz9eU6ZM0fbt23XDDTeoX79+Sk1NdXdpAAAAqGLVPrzOnDlT999/vx544AG1bNlSs2bNUuPGjTVnzhx3lwYAAIAqVq1PG8jNzdXWrVv1l7/8xaW9d+/e2rhxY5GvycnJUU5OjvX82LFjkqTjx49XXqHnOXHihCTpl+TvlJt10mr/be+eQu1FtZW13Y7rsGPN1f597/tJkrR161ZrH5Sk3bt3V4s6PDw8lFfgjllF1Vbs9opZb3HrLq69LH0rYh2Vtb1ix66IcSpL3+K2WdZ1VPU4F7mfV/E+UxHjXNbaqss67FhzRazDjjUX1X6hfffEiRNVkqHyt2GMuXBnU4398ssvRpL58ssvXdr//ve/m2bNmhX5mqlTpxpJPHjw4MGDBw8ePGz22L9//wXzYbU+8prP4XC4PDfGFGrLN3nyZE2YMMF6npeXp99//13169cv9jUX6/jx42rcuLH279+v2rVrV8o2LhWMVdkwXqXHWJUN41V6jFXpMVZlw3j9jzFGmZmZCgsLu2Dfah1eAwMDVaNGDaWlpbm0p6enKzg4uMjXOJ1OOZ1Ol7Y6depUVokuateufdnvfKXFWJUN41V6jFXZMF6lx1iVHmNVNozXOQEBAaXqV60nbHl5ealTp05KSEhwaU9ISFC3bt3cVBUAAADcpVofeZWkCRMm6J577lHnzp3VtWtXvfrqq0pNTdXo0aPdXRoAAACqWLUPr3feeacOHz6sp59+WgcPHlSbNm308ccfKyIiwt2lWZxOp6ZOnVrodAUUxliVDeNVeoxV2TBepcdYlR5jVTaMV/k4jCnNNQkAAAAA96vW57wCAAAA5yO8AgAAwDYIrwAAALANwisAAABsg/B6kWbPnq3IyEh5e3urU6dO+vzzz91dUqWLjY2Vw+FweYSEhFjLjTGKjY1VWFiYfHx81L17d+3atctlHTk5OXr44YcVGBgoPz8/3XTTTTpw4IBLnyNHjuiee+5RQECAAgICdM899+jo0aNV8RbL7bPPPtOgQYMUFhYmh8OhFStWuCyvyrFJTU3VoEGD5Ofnp8DAQI0bN065ubmV8bbL5UJjNWLEiEL72bXXXuvS53IZq/j4eF199dXy9/dXUFCQbrnlFut+5PnYt/6nNOPF/nXOnDlz1K5dO+si+V27dtXq1aut5exXri40XuxXVeSCN5BFsRYvXmw8PT3Na6+9Zr7//nvzyCOPGD8/P7Nv3z53l1appk6dalq3bm0OHjxoPdLT063lzz77rPH39zfvv/++2bFjh7nzzjtNaGioOX78uNVn9OjRpmHDhiYhIcFs27bN9OjRw7Rv396cOXPG6tO3b1/Tpk0bs3HjRrNx40bTpk0bM3DgwCp9r2X18ccfmylTppj333/fSDLLly93WV5VY3PmzBnTpk0b06NHD7Nt2zaTkJBgwsLCzNixYyt9DErrQmM1fPhw07dvX5f97PDhwy59Lpex6tOnj5k3b57ZuXOn+eabb8yAAQNMeHi4OXHihNWHfet/SjNe7F/nrFy50qxatcrs3r3b7N692zzxxBPG09PT7Ny50xjDflXQhcaL/apqEF4vwjXXXGNGjx7t0taiRQvzl7/8xU0VVY2pU6ea9u3bF7ksLy/PhISEmGeffdZqO3XqlAkICDBz5841xhhz9OhR4+npaRYvXmz1+eWXX4yHh4dZs2aNMcaY77//3kgymzdvtvps2rTJSDI//PBDJbyrilcwkFXl2Hz88cfGw8PD/PLLL1afd955xzidTnPs2LFKeb8Xo7jwevPNNxf7mst1rIwxJj093UgySUlJxhj2rQspOF7GsH+VpG7duub1119nvyql/PEyhv2qqnDaQDnl5uZq69at6t27t0t77969tXHjRjdVVXX27NmjsLAwRUZGasiQIfr5558lSSkpKUpLS3MZF6fTqejoaGtctm7dqtOnT7v0CQsLU5s2baw+mzZtUkBAgLp06WL1ufbaaxUQEGDb8a3Ksdm0aZPatGmjsLAwq0+fPn2Uk5OjrVu3Vur7rEgbNmxQUFCQmjVrpj/96U9KT0+3ll3OY3Xs2DFJUr169SSxb11IwfHKx/7l6uzZs1q8eLFOnjyprl27sl9dQMHxysd+Vfmq/R22qquMjAydPXtWwcHBLu3BwcFKS0tzU1VVo0uXLlq4cKGaNWumQ4cOadq0aerWrZt27dplvfeixmXfvn2SpLS0NHl5ealu3bqF+uS/Pi0tTUFBQYW2HRQUZNvxrcqxSUtLK7SdunXrysvLyzbj169fP91+++2KiIhQSkqK/va3v6lnz57aunWrnE7nZTtWxhhNmDBB119/vdq0aSOJfaskRY2XxP51vh07dqhr1646deqUatWqpeXLl6tVq1ZWUGK/clXceEnsV1WF8HqRHA6Hy3NjTKG2S02/fv2sf7dt21Zdu3ZV06ZNtWDBAuvE9PKMS8E+RfW/FMa3qsbG7uN35513Wv9u06aNOnfurIiICK1atUqDBw8u9nWX+liNHTtW3333nb744otCy9i3CituvNi//qd58+b65ptvdPToUb3//vsaPny4kpKSrOXsV66KG69WrVqxX1URThsop8DAQNWoUaPQ/3DS09ML/W/oUufn56e2bdtqz5491lUHShqXkJAQ5ebm6siRIyX2OXToUKFt/fbbb7Yd36ocm5CQkELbOXLkiE6fPm3b8QsNDVVERIT27Nkj6fIcq4cfflgrV67U+vXr1ahRI6udfatoxY1XUS7n/cvLy0tXXnmlOnfurPj4eLVv314vvfQS+1UxihuvolzO+1VlIryWk5eXlzp16qSEhASX9oSEBHXr1s1NVblHTk6OkpOTFRoaqsjISIWEhLiMS25urpKSkqxx6dSpkzw9PV36HDx4UDt37rT6dO3aVceOHdN//vMfq89XX32lY8eO2XZ8q3Jsunbtqp07d+rgwYNWn7Vr18rpdKpTp06V+j4ry+HDh7V//36FhoZKurzGyhijsWPHatmyZVq3bp0iIyNdlrNvubrQeBXlct6/CjLGKCcnh/2qlPLHqyjsV5WkCiaFXbLyL5X1xhtvmO+//96MHz/e+Pn5mb1797q7tEo1ceJEs2HDBvPzzz+bzZs3m4EDBxp/f3/rfT/77LMmICDALFu2zOzYscPcddddRV5apVGjRiYxMdFs27bN9OzZs8hLhbRr185s2rTJbNq0ybRt27baXyorMzPTbN++3Wzfvt1IMjNnzjTbt2+3Lp9WVWOTfxmVXr16mW3btpnExETTqFGjanUZlZLGKjMz00ycONFs3LjRpKSkmPXr15uuXbuahg0bXpZj9eCDD5qAgACzYcMGl0vwZGVlWX3Yt/7nQuPF/vU/kydPNp999plJSUkx3333nXniiSeMh4eHWbt2rTGG/aqgksaL/arqEF4v0r/+9S8TERFhvLy8TMeOHV0uxXKpyr/On6enpwkLCzODBw82u3btspbn5eWZqVOnmpCQEON0Os0f/vAHs2PHDpd1ZGdnm7Fjx5p69eoZHx8fM3DgQJOamurS5/Dhw+buu+82/v7+xt/f39x9993myJEjVfEWy239+vVGUqHH8OHDjTFVOzb79u0zAwYMMD4+PqZevXpm7Nix5tSpU5X59sukpLHKysoyvXv3Ng0aNDCenp4mPDzcDB8+vNA4XC5jVdQ4STLz5s2z+rBv/c+Fxov9639Gjhxp/Q5r0KCB6dWrlxVcjWG/Kqik8WK/qjoOY4ypuuO8AAAAQPlxzisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAlMOGDRvkcDh09OhRSdL8+fNVp04dt9ZUkMPh0IoVK9xdBgBUKMIrgEvOiBEj5HA4NHr06ELLxowZI4fDoREjRlToNu+88079+OOPFbrOouS/N4fDIU9PTwUHBysmJkZvvvmm8vLyXPoePHhQ/fr1K9V6CboA7ILwCuCS1LhxYy1evFjZ2dlW26lTp/TOO+8oPDy8wrfn4+OjoKCgCl9vUfr27auDBw9q7969Wr16tXr06KFHHnlEAwcO1JkzZ6x+ISEhcjqdVVITAFQVwiuAS1LHjh0VHh6uZcuWWW3Lli1T48aN1aFDB5e+xhjNmDFDV1xxhXx8fNS+fXu99957Ln0+/vhjNWvWTD4+PurRo4f27t3rsrzgaQM//fSTbr75ZgUHB6tWrVq6+uqrlZiY6PKaJk2aKC4uTiNHjpS/v7/Cw8P16quvXvC9OZ1OhYSEqGHDhurYsaOeeOIJffDBB1q9erXmz59v9Tv/aGpubq7Gjh2r0NBQeXt7q0mTJoqPj7fqkKRbb71VDofDel5R7+HAgQMaMmSI6tWrJz8/P3Xu3FlfffWVtfzDDz9Up06d5O3trSuuuEJPPfWUSwgHgPMRXgFcsu677z7NmzfPev7mm29q5MiRhfr99a9/1bx58zRnzhzt2rVLjz76qIYNG6akpCRJ0v79+zV48GD1799f33zzjR544AH95S9/KXHbJ06cUP/+/ZWYmKjt27erT58+GjRokFJTU136vfDCC+rcubO2b9+uMWPG6MEHH9QPP/xQ5vfas2dPtW/f3iWsn+8f//iHVq5cqaVLl2r37t166623rJC6ZcsWSdK8efN08OBB63lFvIcTJ04oOjpav/76q1auXKlvv/1Wjz32mHWKwyeffKJhw4Zp3Lhx+v777/XKK69o/vz5+vvf/17mMQBwmTAAcIkZPny4ufnmm81vv/1mnE6nSUlJMXv37jXe3t7mt99+MzfffLMZPny4McaYEydOGG9vb7Nx40aXddx///3mrrvuMsYYM3nyZNOyZUuTl5dnLX/88ceNJHPkyBFjjDHz5s0zAQEBJdbVqlUr889//tN6HhERYYYNG2Y9z8vLM0FBQWbOnDkXfG9FufPOO03Lli2t55LM8uXLjTHGPPzww6Znz54u7+F85/etyPfwyiuvGH9/f3P48OEi13fDDTeYuLg4l7Z///vfJjQ09IK1ALg81XRzdgaAShMYGKgBAwZowYIFMsZowIABCgwMdOnz/fff69SpU4qJiXFpz83NtU4vSE5O1rXXXiuHw2Et79q1a4nbPnnypJ566il99NFH+vXXX3XmzBllZ2cXOmrZrl07698Oh0MhISFKT08v1/s1xrjUeL4RI0YoJiZGzZs3V9++fTVw4ED17t270t/DN998ow4dOqhevXpFbmPr1q3asmWLy5HWs2fP6tSpU8rKypKvr2+p3juAywfhFcAlbeTIkRo7dqwk6V//+leh5fl/vl61apUaNmzosix/spMxpszb/fOf/6xPPvlEzz//vK688kr5+PjotttuU25urks/T09Pl+cOh6PQVQNKKzk5WZGRkUUu69ixo1JSUrR69WolJibqjjvu0I033ljo3N6Kfg8+Pj4l1pyXl6ennnpKgwcPLrTM29u7xNcCuDwRXgFc0vr27WuFrT59+hRa3qpVKzmdTqWmpio6OrrIdbRq1arQZaQ2b95c4nY///xzjRgxQrfeequkc+d+FpzkVZHWrVunHTt26NFHHy22T+3atXXnnXfqzjvv1G233aa+ffvq999/V7169eTp6amzZ89W+Hto166dXn/9dWs7BXXs2FG7d+/WlVdeWab1Arh8EV4BXNJq1Kih5ORk698F+fv7a9KkSXr00UeVl5en66+/XsePH9fGjRtVq1YtDR8+XKNHj9YLL7ygCRMmaNSoUdq6davLrP6iXHnllVq2bJkGDRokh8Ohv/3tb+U+olpQTk6O0tLSdPbsWR06dEhr1qxRfHy8Bg4cqHvvvbfI17z44osKDQ3VVVddJQ8PD7377rsKCQmxrpDQpEkTffrpp7ruuuvkdDpVt27dCnkPd911l+Li4nTLLbcoPj5eoaGh2r59u8LCwtS1a1c9+eSTGjhwoBo3bqzbb79dHh4e+u6777Rjxw5NmzbtYocKwCWIqw0AuOTVrl1btWvXLnb5M888oyeffFLx8fFq2bKl+vTpow8//ND6E3x4eLjef/99ffjhh2rfvr3mzp2ruLi4Erf54osvqm7duurWrZsGDRqkPn36qGPHjhXyftasWaPQ0FA1adJEffv21fr16/WPf/xDH3zwQZEBXZJq1aql6dOnq3Pnzrr66qu1d+9effzxx/LwOPdr4IUXXlBCQoLLpcQq4j14eXlp7dq1CgoKUv/+/dW2bVs9++yzVp19+vTRRx99pISEBF199dW69tprNXPmTEVERFzECAG4lDlMeU7mAgAAANyAI68AAACwDcIrAAAAbIPwCgAAANsgvAIAAMA2CK8AAACwDcIrAAAAbIPwCgAAANsgvAIAAMA2CK8AAACwDcIrAAAAbIPwCgAAANv4fyx1IA+zGRsPAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plotting distribution of median distances between genes per tad\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Initialize a list to store the values from column 4\n",
    "distances = []\n",
    "\n",
    "# Read the file and extract column 4 values\n",
    "with open('tad_exp_dist.bed', 'r') as file:\n",
    "    for line in file:\n",
    "        parts = line.strip().split('\\t')\n",
    "        if len(parts) >= 4:\n",
    "            distances.append(float(parts[3]))\n",
    "\n",
    "# Plot the histogram\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(distances, bins=100, color='skyblue', edgecolor='black')\n",
    "plt.xlabel('Median Distance')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Distribution of median distances between genes within TADs')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b1830666-8bd8-4d29-8359-7fec7c506d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data has been split into quartiles and saved to separate files: tad_exp_dist_Q1.bed, tad_exp_dist_Q2.bed, tad_exp_dist_Q3.bed, tad_exp_dist_Q4.bed.\n"
     ]
    }
   ],
   "source": [
    "#split TADs into 4 quartiles of median distances\n",
    "import pandas as pd\n",
    "\n",
    "# Load the data into a DataFrame\n",
    "data = pd.read_csv('tad_exp_dist.bed', sep='\\t', header=None)\n",
    "data.columns = ['chr', 'tad_start', 'tad_end', 'median_distance']\n",
    "\n",
    "# Calculate the quartiles for the 'median_distance' column\n",
    "quartiles = data['median_distance'].quantile([0.25, 0.5, 0.75])\n",
    "\n",
    "# Assign each row to a quartile\n",
    "data['quartile'] = pd.cut(data['median_distance'], \n",
    "                          bins=[-float('inf'), quartiles[0.25], quartiles[0.5], quartiles[0.75], float('inf')],\n",
    "                          labels=['Q1', 'Q2', 'Q3', 'Q4'])\n",
    "\n",
    "# Write each quartile to a separate file\n",
    "for q in ['Q1', 'Q2', 'Q3', 'Q4']:\n",
    "    quartile_data = data[data['quartile'] == q]\n",
    "    quartile_data.drop(columns=['quartile']).to_csv(f'tad_exp_dist_{q}.bed', sep='\\t', index=False, header=False)\n",
    "\n",
    "print(\"Data has been split into quartiles and saved to separate files: tad_exp_dist_Q1.bed, tad_exp_dist_Q2.bed, tad_exp_dist_Q3.bed, tad_exp_dist_Q4.bed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "711579c6-5a77-4f9b-94b7-5ad42559976a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "170 tad_exp_dist_Q1.bed\n",
      "169 tad_exp_dist_Q2.bed\n",
      "169 tad_exp_dist_Q3.bed\n",
      "170 tad_exp_dist_Q4.bed\n"
     ]
    }
   ],
   "source": [
    "!wc -l tad_exp_dist_Q1.bed\n",
    "!wc -l tad_exp_dist_Q2.bed\n",
    "!wc -l tad_exp_dist_Q3.bed\n",
    "!wc -l tad_exp_dist_Q4.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72c268eb-3c07-40f3-bf84-8895b569ebb6",
   "metadata": {},
   "source": [
    "### Same for nonTADs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9128f399-487b-425f-a213-ccc2f9d9a4db",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File az_genes.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t37926\t50143\n",
      "\n",
      "***** WARNING: File az_genes.bed has a record where naming convention (leading zero) is inconsistent with other files:\n",
      "chr10\t37926\t50143\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<BedTool(nontads_exp_genes.bed)>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pybedtools\n",
    "\n",
    "# Load TADs and genes\n",
    "tads = pybedtools.BedTool('nontads_exp.bed')\n",
    "genes = pybedtools.BedTool('az_genes.bed')\n",
    "\n",
    "intersected_genes = tads.intersect(genes, wa=True, wb=True, F=1)  \n",
    "intersected_genes.saveas('nontads_exp_genes.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b56e5271-f8f9-4cee-857a-6f05e1607b86",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 569 unique TADs.\n",
      "TAD: chr01:1-84999 has 15 genes.\n",
      "TAD: chr01:10400001-10474999 has 10 genes.\n",
      "TAD: chr01:10840001-11264999 has 48 genes.\n",
      "TAD: chr01:11705001-11904999 has 12 genes.\n",
      "TAD: chr01:12100001-12689999 has 47 genes.\n",
      "TAD: chr01:12870001-13464999 has 45 genes.\n",
      "TAD: chr01:13515001-13864999 has 30 genes.\n",
      "TAD: chr01:13930001-15194999 has 105 genes.\n",
      "TAD: chr01:15255001-15869999 has 58 genes.\n",
      "TAD: chr01:16065001-16459999 has 30 genes.\n",
      "TAD: chr01:16540001-18684999 has 117 genes.\n",
      "TAD: chr01:18720001-19194999 has 43 genes.\n",
      "TAD: chr01:1925001-2739999 has 91 genes.\n",
      "TAD: chr01:19265001-19729999 has 35 genes.\n",
      "TAD: chr01:19830001-19964999 has 9 genes.\n",
      "TAD: chr01:20040001-21779999 has 127 genes.\n",
      "TAD: chr01:21915001-22179999 has 18 genes.\n",
      "TAD: chr01:22255001-22794999 has 62 genes.\n",
      "TAD: chr01:23105001-23629999 has 71 genes.\n",
      "TAD: chr01:23795001-23984999 has 16 genes.\n",
      "TAD: chr01:24275001-24554999 has 25 genes.\n",
      "TAD: chr01:24755001-24859999 has 6 genes.\n",
      "TAD: chr01:25165001-25254999 has 7 genes.\n",
      "TAD: chr01:25375001-25744999 has 42 genes.\n",
      "TAD: chr01:25830001-25854999 has 5 genes.\n",
      "TAD: chr01:25975001-26139999 has 21 genes.\n",
      "TAD: chr01:26352501-26664999 has 43 genes.\n",
      "TAD: chr01:26790001-26919999 has 16 genes.\n",
      "TAD: chr01:27140001-27799999 has 69 genes.\n",
      "TAD: chr01:27895001-28274999 has 40 genes.\n",
      "TAD: chr01:280001-624999 has 54 genes.\n",
      "TAD: chr01:28405001-29014999 has 70 genes.\n",
      "TAD: chr01:29170001-29299999 has 16 genes.\n",
      "TAD: chr01:2940001-3554999 has 105 genes.\n",
      "TAD: chr01:29450001-29494999 has 5 genes.\n",
      "TAD: chr01:29590001-30199999 has 75 genes.\n",
      "TAD: chr01:30225001-30649999 has 60 genes.\n",
      "TAD: chr01:30705001-30894999 has 26 genes.\n",
      "TAD: chr01:31040001-31749999 has 95 genes.\n",
      "TAD: chr01:32050001-32819999 has 97 genes.\n",
      "TAD: chr01:32925001-33069999 has 19 genes.\n",
      "TAD: chr01:33390001-35919999 has 345 genes.\n",
      "TAD: chr01:36075001-36314999 has 34 genes.\n",
      "TAD: chr01:36425001-36539999 has 17 genes.\n",
      "TAD: chr01:3650001-4229999 has 75 genes.\n",
      "TAD: chr01:36670001-36799999 has 18 genes.\n",
      "TAD: chr01:36900001-37354999 has 53 genes.\n",
      "TAD: chr01:37470001-37939999 has 40 genes.\n",
      "TAD: chr01:38180001-38614999 has 56 genes.\n",
      "TAD: chr01:38965001-39144999 has 20 genes.\n",
      "TAD: chr01:39185001-40077499 has 113 genes.\n",
      "TAD: chr01:40260001-40469999 has 31 genes.\n",
      "TAD: chr01:40865001-40964999 has 8 genes.\n",
      "TAD: chr01:41125001-41194999 has 9 genes.\n",
      "TAD: chr01:41250001-41374999 has 15 genes.\n",
      "TAD: chr01:41410001-41529999 has 14 genes.\n",
      "TAD: chr01:41845001-42459999 has 68 genes.\n",
      "TAD: chr01:42520001-43259999 has 108 genes.\n",
      "TAD: chr01:4270001-4419999 has 16 genes.\n",
      "TAD: chr01:43325001-43429999 has 20 genes.\n",
      "TAD: chr01:43570001-43664999 has 11 genes.\n",
      "TAD: chr01:43740001-44011168 has 31 genes.\n",
      "TAD: chr01:4510001-4909999 has 53 genes.\n",
      "TAD: chr01:5035001-5319999 has 34 genes.\n",
      "TAD: chr01:5505001-5659999 has 15 genes.\n",
      "TAD: chr01:5730001-5849999 has 10 genes.\n",
      "TAD: chr01:5965001-6659999 has 76 genes.\n",
      "TAD: chr01:6860001-7029999 has 18 genes.\n",
      "TAD: chr01:705001-1799999 has 142 genes.\n",
      "TAD: chr01:7185001-7389999 has 20 genes.\n",
      "TAD: chr01:7530001-7694999 has 21 genes.\n",
      "TAD: chr01:7790001-9194999 has 172 genes.\n",
      "TAD: chr01:9595001-10244999 has 74 genes.\n",
      "TAD: chr02:1-189999 has 35 genes.\n",
      "TAD: chr02:10310001-10429999 has 6 genes.\n",
      "TAD: chr02:10475001-11969999 has 131 genes.\n",
      "TAD: chr02:12440001-12739999 has 28 genes.\n",
      "TAD: chr02:12960001-13274999 has 30 genes.\n",
      "TAD: chr02:13410001-13994999 has 23 genes.\n",
      "TAD: chr02:14145001-15534999 has 69 genes.\n",
      "TAD: chr02:15805001-15944999 has 17 genes.\n",
      "TAD: chr02:16030001-17554999 has 117 genes.\n",
      "TAD: chr02:1620001-3404999 has 242 genes.\n",
      "TAD: chr02:17735001-18854999 has 79 genes.\n",
      "TAD: chr02:19305001-19409999 has 11 genes.\n",
      "TAD: chr02:20165001-20979999 has 66 genes.\n",
      "TAD: chr02:21095001-21199999 has 13 genes.\n",
      "TAD: chr02:21235001-21484999 has 31 genes.\n",
      "TAD: chr02:21540001-22614999 has 108 genes.\n",
      "TAD: chr02:22780001-23694999 has 118 genes.\n",
      "TAD: chr02:23745001-23809999 has 11 genes.\n",
      "TAD: chr02:23875001-24019999 has 22 genes.\n",
      "TAD: chr02:24175001-24349999 has 17 genes.\n",
      "TAD: chr02:24440001-25124999 has 70 genes.\n",
      "TAD: chr02:25165001-25359999 has 21 genes.\n",
      "TAD: chr02:25515001-25574999 has 10 genes.\n",
      "TAD: chr02:25615001-25874999 has 18 genes.\n",
      "TAD: chr02:25925001-26719999 has 95 genes.\n",
      "TAD: chr02:26870001-27054999 has 27 genes.\n",
      "TAD: chr02:27505001-27579999 has 13 genes.\n",
      "TAD: chr02:27860001-28029999 has 17 genes.\n",
      "TAD: chr02:28125001-28214999 has 13 genes.\n",
      "TAD: chr02:28240001-28309999 has 5 genes.\n",
      "TAD: chr02:28435001-29024999 has 66 genes.\n",
      "TAD: chr02:29095001-29144999 has 5 genes.\n",
      "TAD: chr02:29370001-29487499 has 14 genes.\n",
      "TAD: chr02:29540001-29699999 has 13 genes.\n",
      "TAD: chr02:29780001-30474999 has 98 genes.\n",
      "TAD: chr02:30825001-30909999 has 11 genes.\n",
      "TAD: chr02:31035001-31134999 has 13 genes.\n",
      "TAD: chr02:31245001-31404999 has 21 genes.\n",
      "TAD: chr02:31765001-31944999 has 17 genes.\n",
      "TAD: chr02:32170001-32494999 has 42 genes.\n",
      "TAD: chr02:32685001-33319999 has 89 genes.\n",
      "TAD: chr02:33720001-33924999 has 21 genes.\n",
      "TAD: chr02:34050001-34244999 has 30 genes.\n",
      "TAD: chr02:34345001-34429999 has 12 genes.\n",
      "TAD: chr02:34540001-34744999 has 31 genes.\n",
      "TAD: chr02:34775001-35049999 has 41 genes.\n",
      "TAD: chr02:35190001-35294999 has 8 genes.\n",
      "TAD: chr02:35465001-35689999 has 36 genes.\n",
      "TAD: chr02:36050001-36468344 has 68 genes.\n",
      "TAD: chr02:4075001-4174999 has 14 genes.\n",
      "TAD: chr02:420001-569999 has 24 genes.\n",
      "TAD: chr02:4220001-5054999 has 95 genes.\n",
      "TAD: chr02:5400001-5729999 has 57 genes.\n",
      "TAD: chr02:6110001-6594999 has 70 genes.\n",
      "TAD: chr02:6645001-7254999 has 60 genes.\n",
      "TAD: chr02:7305001-7659999 has 34 genes.\n",
      "TAD: chr02:745001-1494999 has 124 genes.\n",
      "TAD: chr02:7870001-8219999 has 25 genes.\n",
      "TAD: chr02:8260001-8314999 has 4 genes.\n",
      "TAD: chr02:8345001-8829999 has 50 genes.\n",
      "TAD: chr02:9075001-10054999 has 73 genes.\n",
      "TAD: chr03:1-604999 has 91 genes.\n",
      "TAD: chr03:10245001-10754999 has 72 genes.\n",
      "TAD: chr03:10795001-11199999 has 51 genes.\n",
      "TAD: chr03:1095001-1359999 has 27 genes.\n",
      "TAD: chr03:11420001-11854999 has 35 genes.\n",
      "TAD: chr03:11925001-12044999 has 18 genes.\n",
      "TAD: chr03:12105001-12309999 has 20 genes.\n",
      "TAD: chr03:12355001-12759999 has 56 genes.\n",
      "TAD: chr03:12810001-12939999 has 19 genes.\n",
      "TAD: chr03:13020001-13209999 has 21 genes.\n",
      "TAD: chr03:13235001-13364999 has 15 genes.\n",
      "TAD: chr03:13435001-14189999 has 33 genes.\n",
      "TAD: chr03:1410001-1614999 has 30 genes.\n",
      "TAD: chr03:14335001-14429999 has 10 genes.\n",
      "TAD: chr03:14515001-14859999 has 39 genes.\n",
      "TAD: chr03:15175001-15489999 has 33 genes.\n",
      "TAD: chr03:15605001-15789999 has 21 genes.\n",
      "TAD: chr03:15920001-16124999 has 24 genes.\n",
      "TAD: chr03:16175001-16689999 has 47 genes.\n",
      "TAD: chr03:16730001-16899999 has 12 genes.\n",
      "TAD: chr03:17050001-17389999 has 15 genes.\n",
      "TAD: chr03:1705001-2904999 has 168 genes.\n",
      "TAD: chr03:17640001-17784999 has 12 genes.\n",
      "TAD: chr03:17832501-18309999 has 47 genes.\n",
      "TAD: chr03:18375001-18469999 has 11 genes.\n",
      "TAD: chr03:18580001-18884999 has 22 genes.\n",
      "TAD: chr03:18920001-19434999 has 24 genes.\n",
      "TAD: chr03:20210001-21639999 has 21 genes.\n",
      "TAD: chr03:21730001-21984999 has 7 genes.\n",
      "TAD: chr03:22025001-22189999 has 6 genes.\n",
      "TAD: chr03:22240001-24109999 has 172 genes.\n",
      "TAD: chr03:24365001-24574999 has 18 genes.\n",
      "TAD: chr03:24680001-25429999 has 66 genes.\n",
      "TAD: chr03:25560001-26164999 has 54 genes.\n",
      "TAD: chr03:26390001-26869999 has 53 genes.\n",
      "TAD: chr03:27190001-27289999 has 11 genes.\n",
      "TAD: chr03:27905001-28084999 has 21 genes.\n",
      "TAD: chr03:28237501-28434999 has 23 genes.\n",
      "TAD: chr03:28570001-29219999 has 66 genes.\n",
      "TAD: chr03:29295001-30144999 has 101 genes.\n",
      "TAD: chr03:3015001-3149999 has 21 genes.\n",
      "TAD: chr03:30380001-30549999 has 16 genes.\n",
      "TAD: chr03:30580001-30729999 has 20 genes.\n",
      "TAD: chr03:30985001-31264999 has 32 genes.\n",
      "TAD: chr03:31310001-31509999 has 21 genes.\n",
      "TAD: chr03:31570001-31969999 has 37 genes.\n",
      "TAD: chr03:32025001-32119999 has 15 genes.\n",
      "TAD: chr03:32230001-32754999 has 73 genes.\n",
      "TAD: chr03:3255001-3349999 has 9 genes.\n",
      "TAD: chr03:32895001-33424999 has 58 genes.\n",
      "TAD: chr03:33475001-33649999 has 25 genes.\n",
      "TAD: chr03:33690001-33874999 has 31 genes.\n",
      "TAD: chr03:3380001-3719999 has 49 genes.\n",
      "TAD: chr03:33955001-34114999 has 19 genes.\n",
      "TAD: chr03:34255001-34739999 has 72 genes.\n",
      "TAD: chr03:34825001-35184999 has 57 genes.\n",
      "TAD: chr03:35280001-35464999 has 27 genes.\n",
      "TAD: chr03:35510001-35809999 has 42 genes.\n",
      "TAD: chr03:36010001-36399999 has 58 genes.\n",
      "TAD: chr03:36600001-36744999 has 27 genes.\n",
      "TAD: chr03:36850001-36959999 has 17 genes.\n",
      "TAD: chr03:37170001-37924999 has 127 genes.\n",
      "TAD: chr03:37975001-38425022 has 60 genes.\n",
      "TAD: chr03:3810001-4084999 has 38 genes.\n",
      "TAD: chr03:4150001-4294999 has 22 genes.\n",
      "TAD: chr03:4325001-4399999 has 8 genes.\n",
      "TAD: chr03:4515001-4619999 has 15 genes.\n",
      "TAD: chr03:4835001-5039999 has 26 genes.\n",
      "TAD: chr03:5280001-5769999 has 70 genes.\n",
      "TAD: chr03:5815001-6619999 has 117 genes.\n",
      "TAD: chr03:640001-909999 has 34 genes.\n",
      "TAD: chr03:6740001-6899999 has 16 genes.\n",
      "TAD: chr03:6955001-7794999 has 102 genes.\n",
      "TAD: chr03:7845001-8354999 has 58 genes.\n",
      "TAD: chr03:8400001-8514999 has 12 genes.\n",
      "TAD: chr03:8595001-8789999 has 22 genes.\n",
      "TAD: chr03:8965001-9504999 has 58 genes.\n",
      "TAD: chr03:9555001-9689999 has 20 genes.\n",
      "TAD: chr03:9795001-10124999 has 47 genes.\n",
      "TAD: chr04:1-289999 has 15 genes.\n",
      "TAD: chr04:10540001-10739999 has 10 genes.\n",
      "TAD: chr04:10770001-11664999 has 70 genes.\n",
      "TAD: chr04:1130001-1224999 has 10 genes.\n",
      "TAD: chr04:11700001-11969999 has 8 genes.\n",
      "TAD: chr04:12110001-14184999 has 134 genes.\n",
      "TAD: chr04:1270001-1462499 has 10 genes.\n",
      "TAD: chr04:14225001-14369999 has 9 genes.\n",
      "TAD: chr04:15055001-17107499 has 163 genes.\n",
      "TAD: chr04:1580001-4027499 has 110 genes.\n",
      "TAD: chr04:17155001-17309999 has 20 genes.\n",
      "TAD: chr04:17492501-18044999 has 54 genes.\n",
      "TAD: chr04:18275001-19314999 has 118 genes.\n",
      "TAD: chr04:19350001-19824999 has 37 genes.\n",
      "TAD: chr04:19940001-20149999 has 19 genes.\n",
      "TAD: chr04:20205001-20694999 has 60 genes.\n",
      "TAD: chr04:20790001-21684999 has 132 genes.\n",
      "TAD: chr04:21935001-22219999 has 32 genes.\n",
      "TAD: chr04:22305001-22499999 has 27 genes.\n",
      "TAD: chr04:22545001-22654999 has 17 genes.\n",
      "TAD: chr04:22825001-23234999 has 60 genes.\n",
      "TAD: chr04:23345001-23604999 has 47 genes.\n",
      "TAD: chr04:23660001-23954999 has 37 genes.\n",
      "TAD: chr04:24110001-24229999 has 18 genes.\n",
      "TAD: chr04:24770001-25104999 has 44 genes.\n",
      "TAD: chr04:25295001-26654999 has 160 genes.\n",
      "TAD: chr04:26755001-26924999 has 22 genes.\n",
      "TAD: chr04:27070001-27144999 has 7 genes.\n",
      "TAD: chr04:27195001-27294999 has 9 genes.\n",
      "TAD: chr04:27365001-27734999 has 32 genes.\n",
      "TAD: chr04:27785001-28244999 has 54 genes.\n",
      "TAD: chr04:28317501-28399999 has 6 genes.\n",
      "TAD: chr04:28470001-29044999 has 83 genes.\n",
      "TAD: chr04:29220001-29824999 has 73 genes.\n",
      "TAD: chr04:29860001-31614999 has 225 genes.\n",
      "TAD: chr04:31740001-32272499 has 63 genes.\n",
      "TAD: chr04:32360001-32509999 has 22 genes.\n",
      "TAD: chr04:32600001-33324999 has 109 genes.\n",
      "TAD: chr04:33525001-33999999 has 56 genes.\n",
      "TAD: chr04:34055001-34192892 has 25 genes.\n",
      "TAD: chr04:350001-1024999 has 49 genes.\n",
      "TAD: chr04:4417501-4559999 has 6 genes.\n",
      "TAD: chr04:4650001-4959999 has 21 genes.\n",
      "TAD: chr04:5480001-7254999 has 132 genes.\n",
      "TAD: chr04:7367501-7777499 has 14 genes.\n",
      "TAD: chr04:8062501-9439999 has 48 genes.\n",
      "TAD: chr04:9460001-9919999 has 22 genes.\n",
      "TAD: chr05:1-94999 has 10 genes.\n",
      "TAD: chr05:13215001-13754999 has 38 genes.\n",
      "TAD: chr05:14095001-17159999 has 207 genes.\n",
      "TAD: chr05:145001-664999 has 77 genes.\n",
      "TAD: chr05:17225001-17339999 has 10 genes.\n",
      "TAD: chr05:17820001-18379999 has 36 genes.\n",
      "TAD: chr05:18465001-18814999 has 40 genes.\n",
      "TAD: chr05:18935001-19674999 has 73 genes.\n",
      "TAD: chr05:19775001-20269999 has 52 genes.\n",
      "TAD: chr05:20450001-20564999 has 9 genes.\n",
      "TAD: chr05:20665001-20799999 has 14 genes.\n",
      "TAD: chr05:20950001-21539999 has 47 genes.\n",
      "TAD: chr05:21585001-21729999 has 15 genes.\n",
      "TAD: chr05:21790001-21879999 has 6 genes.\n",
      "TAD: chr05:21935001-22084999 has 18 genes.\n",
      "TAD: chr05:22300001-22574999 has 32 genes.\n",
      "TAD: chr05:22725001-23409999 has 66 genes.\n",
      "TAD: chr05:23465001-23654999 has 25 genes.\n",
      "TAD: chr05:23725001-24004999 has 32 genes.\n",
      "TAD: chr05:24115001-24209999 has 8 genes.\n",
      "TAD: chr05:24410001-24534999 has 20 genes.\n",
      "TAD: chr05:2470001-2869999 has 54 genes.\n",
      "TAD: chr05:24770001-25079999 has 36 genes.\n",
      "TAD: chr05:25315001-25614999 has 41 genes.\n",
      "TAD: chr05:26195001-26344999 has 24 genes.\n",
      "TAD: chr05:26480001-26859999 has 58 genes.\n",
      "TAD: chr05:26985001-27089999 has 13 genes.\n",
      "TAD: chr05:27415001-27539999 has 16 genes.\n",
      "TAD: chr05:27835001-28209999 has 62 genes.\n",
      "TAD: chr05:28240001-28559999 has 43 genes.\n",
      "TAD: chr05:28600001-28869999 has 44 genes.\n",
      "TAD: chr05:28970001-29274999 has 40 genes.\n",
      "TAD: chr05:29315001-29404999 has 11 genes.\n",
      "TAD: chr05:29550001-29964999 has 69 genes.\n",
      "TAD: chr05:30300001-30954872 has 72 genes.\n",
      "TAD: chr05:3300001-4879999 has 109 genes.\n",
      "TAD: chr05:4930001-4994999 has 8 genes.\n",
      "TAD: chr05:5075001-5169999 has 7 genes.\n",
      "TAD: chr05:5260001-5734999 has 33 genes.\n",
      "TAD: chr05:5775001-6984999 has 97 genes.\n",
      "TAD: chr05:7180001-8722499 has 110 genes.\n",
      "TAD: chr05:765001-869999 has 13 genes.\n",
      "TAD: chr05:9000001-9819999 has 59 genes.\n",
      "TAD: chr05:950001-2304999 has 152 genes.\n",
      "TAD: chr05:9910001-13074999 has 99 genes.\n",
      "TAD: chr06:10045001-10519999 has 40 genes.\n",
      "TAD: chr06:1030001-1499999 has 51 genes.\n",
      "TAD: chr06:10595001-10852499 has 28 genes.\n",
      "TAD: chr06:10915001-10964999 has 5 genes.\n",
      "TAD: chr06:11055001-11839999 has 58 genes.\n",
      "TAD: chr06:11970001-12604999 has 58 genes.\n",
      "TAD: chr06:12765001-13004999 has 21 genes.\n",
      "TAD: chr06:13060001-13149999 has 7 genes.\n",
      "TAD: chr06:13370001-14632499 has 85 genes.\n",
      "TAD: chr06:14675001-14882499 has 14 genes.\n",
      "TAD: chr06:15155001-16039999 has 36 genes.\n",
      "TAD: chr06:16130001-16864999 has 43 genes.\n",
      "TAD: chr06:1615001-2014999 has 60 genes.\n",
      "TAD: chr06:17010001-17499999 has 41 genes.\n",
      "TAD: chr06:17930001-19389999 has 69 genes.\n",
      "TAD: chr06:19455001-19779999 has 25 genes.\n",
      "TAD: chr06:19825001-21304999 has 109 genes.\n",
      "TAD: chr06:21430001-21519999 has 6 genes.\n",
      "TAD: chr06:2150001-2254999 has 19 genes.\n",
      "TAD: chr06:21575001-22374999 has 61 genes.\n",
      "TAD: chr06:22940001-23164999 has 20 genes.\n",
      "TAD: chr06:23255001-23654999 has 43 genes.\n",
      "TAD: chr06:2335001-3254999 has 142 genes.\n",
      "TAD: chr06:23965001-24614999 has 65 genes.\n",
      "TAD: chr06:24675001-25089999 has 35 genes.\n",
      "TAD: chr06:25200001-25904999 has 65 genes.\n",
      "TAD: chr06:26495001-26714999 has 29 genes.\n",
      "TAD: chr06:26895001-27469999 has 59 genes.\n",
      "TAD: chr06:27635001-27994999 has 45 genes.\n",
      "TAD: chr06:28145001-28334999 has 16 genes.\n",
      "TAD: chr06:28385001-28449999 has 4 genes.\n",
      "TAD: chr06:28560001-28729999 has 29 genes.\n",
      "TAD: chr06:28965001-29784999 has 108 genes.\n",
      "TAD: chr06:29830001-29904999 has 11 genes.\n",
      "TAD: chr06:30135001-30444999 has 39 genes.\n",
      "TAD: chr06:30540001-30689999 has 22 genes.\n",
      "TAD: chr06:30760001-31189045 has 54 genes.\n",
      "TAD: chr06:335001-469999 has 17 genes.\n",
      "TAD: chr06:3485001-4069999 has 73 genes.\n",
      "TAD: chr06:4280001-4519999 has 28 genes.\n",
      "TAD: chr06:4570001-4774999 has 16 genes.\n",
      "TAD: chr06:5030001-5154999 has 13 genes.\n",
      "TAD: chr06:5200001-5739999 has 61 genes.\n",
      "TAD: chr06:535001-684999 has 19 genes.\n",
      "TAD: chr06:5780001-7289999 has 153 genes.\n",
      "TAD: chr06:7425001-7789999 has 37 genes.\n",
      "TAD: chr06:8270001-9284999 has 78 genes.\n",
      "TAD: chr06:9415001-9674999 has 13 genes.\n",
      "TAD: chr07:1-594999 has 57 genes.\n",
      "TAD: chr07:10165001-10304999 has 16 genes.\n",
      "TAD: chr07:10525001-10859999 has 16 genes.\n",
      "TAD: chr07:10925001-11159999 has 14 genes.\n",
      "TAD: chr07:11365001-12459999 has 53 genes.\n",
      "TAD: chr07:12955001-13289999 has 24 genes.\n",
      "TAD: chr07:13460001-14149999 has 24 genes.\n",
      "TAD: chr07:1370001-2047499 has 77 genes.\n",
      "TAD: chr07:14265001-14904999 has 45 genes.\n",
      "TAD: chr07:14950001-15234999 has 18 genes.\n",
      "TAD: chr07:15295001-15359999 has 5 genes.\n",
      "TAD: chr07:15610001-15779999 has 21 genes.\n",
      "TAD: chr07:15940001-16034999 has 11 genes.\n",
      "TAD: chr07:16510001-16934999 has 28 genes.\n",
      "TAD: chr07:17485001-17729999 has 19 genes.\n",
      "TAD: chr07:17910001-19534999 has 151 genes.\n",
      "TAD: chr07:19800001-19994999 has 18 genes.\n",
      "TAD: chr07:20030001-20249999 has 15 genes.\n",
      "TAD: chr07:20555001-20674999 has 12 genes.\n",
      "TAD: chr07:20780001-21224999 has 48 genes.\n",
      "TAD: chr07:21420001-21689999 has 26 genes.\n",
      "TAD: chr07:21810001-22464999 has 77 genes.\n",
      "TAD: chr07:22580001-22729999 has 21 genes.\n",
      "TAD: chr07:22815001-23519999 has 102 genes.\n",
      "TAD: chr07:24170001-24539999 has 42 genes.\n",
      "TAD: chr07:2422501-2529999 has 14 genes.\n",
      "TAD: chr07:24600001-25214999 has 80 genes.\n",
      "TAD: chr07:25310001-26497499 has 160 genes.\n",
      "TAD: chr07:2620001-3569999 has 106 genes.\n",
      "TAD: chr07:26565001-27194999 has 76 genes.\n",
      "TAD: chr07:27265001-27459999 has 35 genes.\n",
      "TAD: chr07:27535001-27869999 has 48 genes.\n",
      "TAD: chr07:27910001-28784999 has 103 genes.\n",
      "TAD: chr07:29090001-29327599 has 16 genes.\n",
      "TAD: chr07:3805001-3959999 has 20 genes.\n",
      "TAD: chr07:4040001-4129999 has 9 genes.\n",
      "TAD: chr07:4460001-4609999 has 17 genes.\n",
      "TAD: chr07:4900001-4974999 has 8 genes.\n",
      "TAD: chr07:5085001-5479999 has 39 genes.\n",
      "TAD: chr07:5640001-5859999 has 25 genes.\n",
      "TAD: chr07:5925001-7139999 has 85 genes.\n",
      "TAD: chr07:690001-1249999 has 69 genes.\n",
      "TAD: chr07:7565001-10089999 has 165 genes.\n",
      "TAD: chr08:1-59999 has 6 genes.\n",
      "TAD: chr08:1055001-3304999 has 238 genes.\n",
      "TAD: chr08:10692501-11394999 has 39 genes.\n",
      "TAD: chr08:115001-759999 has 85 genes.\n",
      "TAD: chr08:11630001-12239999 has 33 genes.\n",
      "TAD: chr08:12330001-12794999 has 28 genes.\n",
      "TAD: chr08:12960001-14347499 has 63 genes.\n",
      "TAD: chr08:14445001-14759999 has 27 genes.\n",
      "TAD: chr08:15150001-16754999 has 72 genes.\n",
      "TAD: chr08:16825001-17744999 has 59 genes.\n",
      "TAD: chr08:17830001-18569999 has 50 genes.\n",
      "TAD: chr08:18625001-20044999 has 103 genes.\n",
      "TAD: chr08:20265001-20914999 has 70 genes.\n",
      "TAD: chr08:21100001-21577499 has 42 genes.\n",
      "TAD: chr08:21690001-21839999 has 9 genes.\n",
      "TAD: chr08:22070001-22264999 has 23 genes.\n",
      "TAD: chr08:22295001-22704999 has 41 genes.\n",
      "TAD: chr08:22810001-23664999 has 77 genes.\n",
      "TAD: chr08:24100001-24704999 has 68 genes.\n",
      "TAD: chr08:24735001-24819999 has 10 genes.\n",
      "TAD: chr08:25240001-25424999 has 29 genes.\n",
      "TAD: chr08:25595001-26114999 has 58 genes.\n",
      "TAD: chr08:26180001-26724999 has 65 genes.\n",
      "TAD: chr08:26940001-27044999 has 10 genes.\n",
      "TAD: chr08:27135001-27504999 has 39 genes.\n",
      "TAD: chr08:27640001-27739999 has 13 genes.\n",
      "TAD: chr08:27790001-28059999 has 34 genes.\n",
      "TAD: chr08:28130001-28739999 has 102 genes.\n",
      "TAD: chr08:28875001-29168739 has 41 genes.\n",
      "TAD: chr08:3920001-4759999 has 103 genes.\n",
      "TAD: chr08:5195001-5534999 has 37 genes.\n",
      "TAD: chr08:5755001-6389999 has 55 genes.\n",
      "TAD: chr08:6425001-7489999 has 49 genes.\n",
      "TAD: chr08:7605001-7849999 has 22 genes.\n",
      "TAD: chr08:7915001-8159999 has 17 genes.\n",
      "TAD: chr08:8330001-8459999 has 9 genes.\n",
      "TAD: chr08:8485001-9014999 has 44 genes.\n",
      "TAD: chr08:9462501-9774999 has 21 genes.\n",
      "TAD: chr08:9965001-10494999 has 28 genes.\n",
      "TAD: chr09:1-2874999 has 161 genes.\n",
      "TAD: chr09:10880001-11504999 has 51 genes.\n",
      "TAD: chr09:11785001-12524999 has 68 genes.\n",
      "TAD: chr09:12585001-12884999 has 34 genes.\n",
      "TAD: chr09:12925001-13004999 has 8 genes.\n",
      "TAD: chr09:13365001-13834999 has 28 genes.\n",
      "TAD: chr09:14185001-14649999 has 32 genes.\n",
      "TAD: chr09:14730001-14934999 has 20 genes.\n",
      "TAD: chr09:14970001-15299999 has 31 genes.\n",
      "TAD: chr09:15495001-15889999 has 38 genes.\n",
      "TAD: chr09:16080001-16384999 has 30 genes.\n",
      "TAD: chr09:16655001-16759999 has 12 genes.\n",
      "TAD: chr09:16845001-17634999 has 94 genes.\n",
      "TAD: chr09:17670001-17864999 has 16 genes.\n",
      "TAD: chr09:17975001-18149999 has 16 genes.\n",
      "TAD: chr09:18255001-18474999 has 25 genes.\n",
      "TAD: chr09:18620001-19089999 has 66 genes.\n",
      "TAD: chr09:19220001-19264999 has 4 genes.\n",
      "TAD: chr09:19535001-19809999 has 41 genes.\n",
      "TAD: chr09:20200001-20274999 has 15 genes.\n",
      "TAD: chr09:20450001-20589999 has 16 genes.\n",
      "TAD: chr09:20625001-20959999 has 39 genes.\n",
      "TAD: chr09:20995001-21364999 has 56 genes.\n",
      "TAD: chr09:21435001-21984999 has 75 genes.\n",
      "TAD: chr09:22020001-22359999 has 53 genes.\n",
      "TAD: chr09:22610001-23125300 has 77 genes.\n",
      "TAD: chr09:2980001-3949999 has 24 genes.\n",
      "TAD: chr09:4420001-4554999 has 12 genes.\n",
      "TAD: chr09:5185001-6279999 has 71 genes.\n",
      "TAD: chr09:6540001-6729999 has 17 genes.\n",
      "TAD: chr09:6860001-7924999 has 63 genes.\n",
      "TAD: chr09:8060001-8459999 has 14 genes.\n",
      "TAD: chr09:8605001-8879999 has 18 genes.\n",
      "TAD: chr09:8975001-10759999 has 113 genes.\n",
      "TAD: chr10:1-1569999 has 93 genes.\n",
      "TAD: chr10:10810001-11739999 has 56 genes.\n",
      "TAD: chr10:12040001-12299999 has 21 genes.\n",
      "TAD: chr10:12655001-13354999 has 65 genes.\n",
      "TAD: chr10:14005001-14244999 has 14 genes.\n",
      "TAD: chr10:14460001-16764999 has 220 genes.\n",
      "TAD: chr10:17495001-18914999 has 126 genes.\n",
      "TAD: chr10:1790001-2119999 has 26 genes.\n",
      "TAD: chr10:18980001-19569999 has 56 genes.\n",
      "TAD: chr10:19865001-19919999 has 6 genes.\n",
      "TAD: chr10:20055001-20344999 has 40 genes.\n",
      "TAD: chr10:20400001-20499999 has 16 genes.\n",
      "TAD: chr10:20670001-20934999 has 35 genes.\n",
      "TAD: chr10:21045001-21399999 has 44 genes.\n",
      "TAD: chr10:21550001-21794999 has 26 genes.\n",
      "TAD: chr10:2180001-2934999 has 82 genes.\n",
      "TAD: chr10:21835001-21909999 has 13 genes.\n",
      "TAD: chr10:22060001-22449999 has 40 genes.\n",
      "TAD: chr10:22630001-22754999 has 17 genes.\n",
      "TAD: chr10:22965001-23104999 has 20 genes.\n",
      "TAD: chr10:23190001-23999999 has 111 genes.\n",
      "TAD: chr10:24280001-24404999 has 15 genes.\n",
      "TAD: chr10:24470001-24584999 has 12 genes.\n",
      "TAD: chr10:24640001-25095745 has 43 genes.\n",
      "TAD: chr10:3105001-3324999 has 21 genes.\n",
      "TAD: chr10:3370001-4594999 has 89 genes.\n",
      "TAD: chr10:4765001-4884999 has 12 genes.\n",
      "TAD: chr10:5135001-7759999 has 152 genes.\n",
      "TAD: chr10:8145001-8499999 has 26 genes.\n",
      "TAD: chr10:8602501-10254999 has 76 genes.\n",
      "TAD: chr11:10585001-13894999 has 130 genes.\n",
      "TAD: chr11:14315001-14954999 has 37 genes.\n",
      "TAD: chr11:15070001-15654999 has 34 genes.\n",
      "TAD: chr11:15765001-16374999 has 37 genes.\n",
      "TAD: chr11:16470001-17114999 has 20 genes.\n",
      "TAD: chr11:1650001-1724999 has 6 genes.\n",
      "TAD: chr11:18200001-18589999 has 19 genes.\n",
      "TAD: chr11:19000001-19179999 has 12 genes.\n",
      "TAD: chr11:19425001-19849999 has 50 genes.\n",
      "TAD: chr11:20205001-20889999 has 51 genes.\n",
      "TAD: chr11:2040001-2089999 has 14 genes.\n",
      "TAD: chr11:21045001-21499999 has 46 genes.\n",
      "TAD: chr11:21887501-21989999 has 12 genes.\n",
      "TAD: chr11:22050001-22489999 has 40 genes.\n",
      "TAD: chr11:2255001-2279999 has 5 genes.\n",
      "TAD: chr11:22555001-23069999 has 46 genes.\n",
      "TAD: chr11:23125001-23539999 has 37 genes.\n",
      "TAD: chr11:23665001-24154999 has 52 genes.\n",
      "TAD: chr11:2385001-2564999 has 22 genes.\n",
      "TAD: chr11:24265001-24364999 has 8 genes.\n",
      "TAD: chr11:24605001-25499999 has 65 genes.\n",
      "TAD: chr11:25580001-26149999 has 53 genes.\n",
      "TAD: chr11:2610001-3014999 has 43 genes.\n",
      "TAD: chr11:26317501-26589999 has 23 genes.\n",
      "TAD: chr11:27120001-30181525 has 218 genes.\n",
      "TAD: chr11:285001-689999 has 64 genes.\n",
      "TAD: chr11:3080001-4394999 has 134 genes.\n",
      "TAD: chr11:4485001-4869999 has 34 genes.\n",
      "TAD: chr11:4960001-5769999 has 68 genes.\n",
      "TAD: chr11:5820001-6177499 has 39 genes.\n",
      "TAD: chr11:6490001-6994999 has 60 genes.\n",
      "TAD: chr11:7315001-7479999 has 11 genes.\n",
      "TAD: chr11:7545001-7689999 has 13 genes.\n",
      "TAD: chr11:7910001-10534999 has 161 genes.\n",
      "TAD: chr11:845001-1469999 has 83 genes.\n",
      "TAD: chr12:1-644999 has 87 genes.\n",
      "TAD: chr12:10355001-11274999 has 67 genes.\n",
      "TAD: chr12:11332501-12674999 has 54 genes.\n",
      "TAD: chr12:1255001-1604999 has 47 genes.\n",
      "TAD: chr12:13040001-13209999 has 10 genes.\n",
      "TAD: chr12:13655001-15074999 has 55 genes.\n",
      "TAD: chr12:15142501-15309999 has 8 genes.\n",
      "TAD: chr12:15380001-17667499 has 97 genes.\n",
      "TAD: chr12:1690001-2099999 has 59 genes.\n",
      "TAD: chr12:17860001-18069999 has 13 genes.\n",
      "TAD: chr12:18105001-18339999 has 18 genes.\n",
      "TAD: chr12:18425001-18954999 has 35 genes.\n",
      "TAD: chr12:19005001-19344999 has 24 genes.\n",
      "TAD: chr12:19380001-20219999 has 95 genes.\n",
      "TAD: chr12:20285001-21494999 has 106 genes.\n",
      "TAD: chr12:21550001-22304999 has 55 genes.\n",
      "TAD: chr12:2200001-2269999 has 8 genes.\n",
      "TAD: chr12:22420001-22789999 has 30 genes.\n",
      "TAD: chr12:22880001-23029999 has 14 genes.\n",
      "TAD: chr12:23365001-23464999 has 11 genes.\n",
      "TAD: chr12:23945001-25444999 has 133 genes.\n",
      "TAD: chr12:25545001-25854999 has 41 genes.\n",
      "TAD: chr12:25900001-26034999 has 16 genes.\n",
      "TAD: chr12:26070001-26394999 has 35 genes.\n",
      "TAD: chr12:26560001-26854999 has 30 genes.\n",
      "TAD: chr12:27045001-27119999 has 10 genes.\n",
      "TAD: chr12:27215001-27487302 has 36 genes.\n",
      "TAD: chr12:3135001-3384999 has 33 genes.\n",
      "TAD: chr12:3810001-4474999 has 72 genes.\n",
      "TAD: chr12:4790001-5939999 has 89 genes.\n",
      "TAD: chr12:6220001-6394999 has 9 genes.\n",
      "TAD: chr12:6920001-8502499 has 62 genes.\n",
      "TAD: chr12:800001-1144999 has 51 genes.\n",
      "TAD: chr12:9070001-9164999 has 9 genes.\n",
      "TAD: chr12:9245001-9934999 has 41 genes.\n",
      "Median distances calculated and saved to nontad_exp_dist.bed.\n"
     ]
    }
   ],
   "source": [
    "#calculate median distance between genes per nontad\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "\n",
    "# Load intersected TAD and gene data\n",
    "tad_gene_file = 'nontads_exp_genes.bed'\n",
    "output_file = 'nontad_exp_dist.bed'\n",
    "\n",
    "# Dictionary to hold TADs and their corresponding gene start-end coordinates\n",
    "tad_genes = defaultdict(list)\n",
    "\n",
    "# Read the TAD-gene intersections\n",
    "with open(tad_gene_file, 'r') as f:\n",
    "    for line in f:\n",
    "        parts = line.strip().split('\\t')\n",
    "        if len(parts) < 6:\n",
    "            print(f\"Skipping line, not enough columns: {line.strip()}\")\n",
    "            continue  # Skip lines that don't have the expected number of columns\n",
    "        \n",
    "        # Extract TAD information and gene coordinates\n",
    "        tad_chr = parts[0]\n",
    "        tad_start = int(parts[1])\n",
    "        tad_end = int(parts[2])\n",
    "        gene_start = int(parts[4])\n",
    "        gene_end = int(parts[5])\n",
    "        \n",
    "        # Group genes by TAD coordinates\n",
    "        tad_genes[(tad_chr, tad_start, tad_end)].append((gene_start, gene_end))\n",
    "\n",
    "# Debug: Print number of TADs found\n",
    "print(f\"Found {len(tad_genes)} unique TADs.\")\n",
    "\n",
    "# Prepare to write output\n",
    "with open(output_file, 'w') as out:\n",
    "    # For each unique TAD, calculate the median distance between genes\n",
    "    for (tad_chr, tad_start, tad_end), genes in tad_genes.items():\n",
    "        # Sort gene starts and ends\n",
    "        gene_starts = sorted(gene[0] for gene in genes)\n",
    "        gene_ends = sorted(gene[1] for gene in genes)\n",
    "\n",
    "        # Debug: Print genes associated with the current TAD\n",
    "        print(f\"TAD: {tad_chr}:{tad_start}-{tad_end} has {len(genes)} genes.\")\n",
    "\n",
    "        # Calculate distances between adjacent genes\n",
    "        distances = []\n",
    "        for i in range(len(gene_starts) - 1):\n",
    "            distance = gene_starts[i + 1] - gene_ends[i]  # distance = gene2_start - gene1_end\n",
    "            if distance > 0:  # Only consider positive distances\n",
    "                distances.append(distance)\n",
    "        \n",
    "        # Calculate median distance\n",
    "        if distances:  # Check if there are distances to calculate median\n",
    "            median_distance = np.median(distances)\n",
    "        else:\n",
    "            median_distance = 0  # If no distances, set to 0 or you can skip this TAD\n",
    "            print(f\"No valid distances for TAD {tad_chr}:{tad_start}-{tad_end}. Setting median distance to 0.\")\n",
    "\n",
    "        # Write the results to the output file\n",
    "        out.write(f\"{tad_chr}\\t{tad_start}\\t{tad_end}\\t{median_distance}\\n\")\n",
    "\n",
    "print(f\"Median distances calculated and saved to {output_file}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5b5fce61-7cdf-44b0-afac-fee1ae4e4034",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAq8AAAHUCAYAAAAUbMECAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABRRElEQVR4nO39d3wU5f7//z83kGwKIZRACpAQkRKqFGnqoRk6KuhRRBQEzxEVFQEL8hWCIlgRPR7FRnsjggUQpEiQYgGULmBEPAYCkhBDD4EAyfX7w1/mw6aRhE02A4/77ba3G3PNtTOvuXayebKZudZhjDECAAAAbMDL0wUAAAAAhUV4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4BQAAgG0QXgEAAGAbhFcAAADYBuEVAAAAtkF4LUNmzpwph8NhPXx9fRUaGqpOnTpp8uTJSklJyfWc2NhYORyOIu0nPT1dsbGxWrt2bZGel9e+ateurd69exdpO5cyd+5cTZ06Nc91DodDsbGxbt2fu33zzTdq1aqVAgIC5HA4tGjRIk+XlK+OHTuqY8eO1vK+ffvkcDg0c+ZMj9WUU37n3eDBg4u0nfXr1ys2NlbHjx93X3FlXMeOHdW4cWO3bvOdd94pU+cHSl5e7wsF/TwV9vfC2rVr5XA4ivy7qKzJPo7CPC7Wr18/ORwODR8+vFDb9fHxUbVq1XTDDTdo7Nix2r9/f2kcXplU3tMFILcZM2aoQYMGOn/+vFJSUvT999/r5Zdf1muvvab58+fr5ptvtvo+8MAD6t69e5G2n56ergkTJkiSS3C5lOLsqzjmzp2rXbt2acSIEbnWbdiwQTVr1izxGorLGKM777xT9erV0+LFixUQEKD69et7uqxCCwsL04YNG1SnTh1Pl1KghQsXqmLFikV6zvr16zVhwgQNHjxYlSpVKpnCrgLvvPOOgoODi/yfB9hXXu8L7vh5atGihTZs2KCGDRu6qVLPyD6Oi/Xt21d16tTRa6+9ludzUlJS9NVXX0mSPv74Y7322mvy9fXNs++kSZPUqVMnZWZm6siRI/rxxx81ffp0vfHGG/rggw90zz33uPeAbIDwWgY1btxYrVq1spZvv/12PfHEE7rxxhvVr18/7d27VyEhIZKkmjVrlniYS09Pl7+/f6ns61Latm3r0f1fyqFDh3T06FH17dtXXbp08XQ5ReZ0Osv8GEtS8+bNPV0CcNUoqfeFihUr2uL95lLyOg6n06lKlSrle3yzZ8/W+fPn1atXLy1dulQLFizQgAED8uxbt25dl+3ccsstGjVqlG6++WYNHjxYTZs2VZMmTdx3QDbAZQM2ERERoddff12nTp3Se++9Z7Xn9SfV1atXq2PHjqpatar8/PwUERGh22+/Xenp6dq3b5+qVasmSZowYYL154jsT1Gyt7d161bdcccdqly5svW/7YIuUVi4cKGaNm0qX19fXXPNNXrrrbdc1mdfErFv3z6X9px/NurYsaOWLl2q/fv35/mnlrwuG9i1a5duvfVWVa5cWb6+vrruuus0a9asPPfzySefaOzYsQoPD1fFihV18803a8+ePfkP/EW+//57denSRYGBgfL391f79u21dOlSa31sbKwV7p9++mk5HA7Vrl073+1l1zR37lw9/fTTCgsLU4UKFdSnTx8dPnxYp06d0r///W8FBwcrODhY999/v9LS0ly2YYzRO++8o+uuu05+fn6qXLmy7rjjDv3xxx+5+r3yyiuKjIyUr6+vWrRooeXLl+eqKa8/D/7++++6//77VbduXfn7+6tGjRrq06ePdu7c6fYxXrp0qa677jo5nU5FRUXl+6lFzssGsrKyNHHiRNWvX19+fn6qVKmSmjZtqjfffFPS36/Nk08+KUmKioqyzqvs827+/Pnq2rWrwsLC5Ofnp+joaD3zzDM6ffq0y34HDx6sChUq6Pfff1fPnj1VoUIF1apVS6NGjVJGRoZL34yMDD3//POKjo6Wr6+vqlatqk6dOmn9+vVWn8K+ftu2bVPv3r1VvXp1OZ1OhYeHq1evXjp48GChxvW7775T27Zt5efnpxo1aui5555TZmamS59z585p4sSJatCggZxOp6pVq6b7779ff/31l8u47969W+vWrbPGsHbt2jLGKCQkRI888ojVNzMzU5UrV5aXl5cOHz5stU+ZMkXly5d3+XPz5s2bdcstt6hKlSry9fVV8+bN9emnn+Y6juTkZD344IOqWbOmfHx8FBUVpQkTJujChQtWn+xz+LXXXtOUKVMUFRWlChUqqF27dtq4cWOhxuv7779Xu3bt5Ovra43Xhx9+mOd72Pz589WuXTsFBASoQoUK6tatm7Zt2+bSpyjnTWFeB6ng9/n8PPnkkwoKCnJ57R999FE5HA69+uqrVtuRI0fk5eWl//znPy5jmv2+cKmfp2wrVqxQixYt5OfnpwYNGmj69Oku6/O6bKAoY5WX7EsWLrVvqfR+d+Rl+vTpCgkJ0axZs+Tn55dnfQWpUqWK3nvvPV24cEFvvPGG1f7XX3/p3//+t2rVqmWdPzfccINWrVpV7FrLJIMyY8aMGUaS2bRpU57r09LSTLly5UyXLl2stvHjx5uLX8aEhATj6+trYmJizKJFi8zatWvNxx9/bO69915z7Ngxc/bsWbNixQojyQwdOtRs2LDBbNiwwfz+++8u24uMjDRPP/20iYuLM4sWLcpzX8YYExkZaWrUqGEiIiLM9OnTzbJly8w999xjJJlXX30117ElJCS4PH/NmjVGklmzZo0xxpjdu3ebG264wYSGhlq1bdiwweovyYwfP95a/vXXX01gYKCpU6eOmT17tlm6dKm5++67jSTz8ssv59pP7dq1zT333GOWLl1qPvnkExMREWHq1q1rLly4UOBrs3btWuPt7W1atmxp5s+fbxYtWmS6du1qHA6HmTdvnjHGmAMHDpgFCxYYSebRRx81GzZsMFu3bs13m9k1RUZGmsGDB5sVK1aYadOmmQoVKphOnTqZmJgYM3r0aLNy5Urz8ssvm3LlyplHH33UZRv/+te/jLe3txk1apRZsWKFmTt3rmnQoIEJCQkxycnJVr/s127o0KFm+fLl5v333zc1atQwoaGhpkOHDla/hIQEI8nMmDHDalu3bp0ZNWqU+fzzz826devMwoULzW233Wb8/PzMr7/+6rYxXrVqlSlXrpy58cYbzYIFC8xnn31mrr/+ehMREZHneTdo0CBrefLkyaZcuXJm/Pjx5ptvvjErVqwwU6dONbGxsdZr8+ijjxpJZsGCBdZ5deLECWOMMS+88IJ54403zNKlS83atWvNtGnTTFRUlOnUqZPLfgcNGmR8fHxMdHS0ee2118yqVavMuHHjjMPhMBMmTLD6nT9/3nTq1MmUL1/ejB492ixbtswsXrzYPPvss+aTTz4p0uuXlpZmqlatalq1amU+/fRTs27dOjN//nwzbNgw88svvxQ4ph06dDBVq1Y14eHh5q233jJff/21eeyxx4wk88gjj1j9MjMzTffu3U1AQICZMGGCiYuLMx9++KGpUaOGadiwoUlPTzfGGLN161ZzzTXXmObNm1tjmH2O9+/f39SrV8/a5saNG40k4+fnZz7++GOrvUePHqZ169bW8urVq42Pj4+56aabzPz5882KFSvM4MGDc52HSUlJplatWiYyMtK89957ZtWqVeaFF14wTqfTDB482OqXfQ7Xrl3bdO/e3SxatMgsWrTINGnSxFSuXNkcP368wDHbsWOH8fX1NU2bNjXz5s0zixcvNj179jS1a9fO9R724osvGofDYYYMGWK++uors2DBAtOuXTsTEBBgdu/ebfUr7HlT2NfhUu/z+cl+71+/fr3V1qBBA+Pn52diYmKstvnz5xtJ1vmV833hUj9PkZGRpmbNmqZhw4Zm9uzZ5uuvvzb//Oc/jSSzbt06az853/+LMlb5Key+S/p3R2RkpOnVq1ee63744QcjyTz55JPGGGMGDhxoHA6H+eOPP1z6Ze/7s88+y/d4w8LCTJ06dazlbt26mWrVqpn333/frF271ixatMiMGzfO+j11pSC8liGXCq/GGBMSEmKio6Ot5ZyB8vPPPzeSzPbt2/Pdxl9//ZUrBObc3rhx4/Jdd7HIyEjjcDhy7S8mJsZUrFjRnD592uXYLhVejTGmV69eJjIyMs/ac9bdv39/43Q6TWJioku/Hj16GH9/f+sXVfZ+evbs6dLv008/NZJcAnJe2rZta6pXr25OnTpltV24cME0btzY1KxZ02RlZRlj/t+b/MXBPT/ZNfXp08elfcSIEUaSeeyxx1zab7vtNlOlShVrecOGDUaSef311136HThwwPj5+ZmnnnrKGGPMsWPHjK+vr+nbt69Lv+w30EuF15wuXLhgzp07Z+rWrWueeOKJXMdT3DFu06aNCQ8PN2fOnLHaTp48aapUqXLJ8Nq7d29z3XXXFbj9V199Nc9zMKesrCxz/vx5s27dOiPJ7Nixw1o3aNAgI8l8+umnLs/p2bOnqV+/vrU8e/ZsI8l88MEH+e6nsK/f5s2bjSTrP5FF0aFDByPJfPnlly7t//rXv4yXl5fZv3+/McaYTz75xEgyX3zxhUu/TZs2GUnmnXfesdoaNWrkcs5k+/DDD40k62dx4sSJpkGDBuaWW24x999/vzHGmHPnzpmAgADz7LPPWs9r0KCBad68uTl//rzL9nr37m3CwsJMZmamMcaYBx980FSoUMGqOdtrr71mJFlhMfscbtKkiUuw+Omnn4wkl/885OWf//ynCQgIMH/99ZfVlpmZaRo2bOhy/iQmJpry5cvn+g/lqVOnTGhoqLnzzjuttsKeN4V9HQrzPp+X06dPGx8fH/P8888bY4w5ePCgkWSefvpp4+fnZ86ePWuM+fv8CA8Pt56X1/tCQT9PkZGRxtfX1+W1OnPmjKlSpYp58MEHrbb8wmthxio/hd13Sf/uKCi8DhkyxEgy8fHxLvt47rnnXPoVJry2adPG+Pn5WcsVKlQwI0aMyLf/lYLLBmzGGFPg+uuuu04+Pj7697//rVmzZuX682Nh3X777YXu26hRIzVr1sylbcCAATp58qS2bt1arP0X1urVq9WlSxfVqlXLpX3w4MFKT0/PdRH9Lbfc4rLctGlTSSrwrs3Tp0/rxx9/1B133KEKFSpY7eXKldO9996rgwcPXtafj3LelRsdHS1J6tWrV672o0ePWpcOfPXVV3I4HBo4cKAuXLhgPUJDQ9WsWTPrT3EbNmzQ2bNnc13U3759e0VGRl6yvgsXLmjSpElq2LChfHx8VL58efn4+Gjv3r2Kj4/P1b+4Y7xp0yb169fP5aaFwMBA9enT55I1tm7dWjt27NDDDz+sr7/+WidPnrzkcy72xx9/aMCAAQoNDVW5cuXk7e2tDh06SFKuY3Q4HLlqatq0qcvxLV++XL6+vhoyZEi++yzs63fttdeqcuXKevrppzVt2jT98ssvRTq2wMDAXK/JgAEDlJWVpW+//daqpVKlSurTp49LLdddd51CQ0MLdTd49o2k2X+ejIuLU0xMjG6++WbFxcVJ+vtcPH36tNX3999/16+//mqdmxfvu2fPnkpKSrJ+tr766it16tRJ4eHhLv169OghSVq3bp1LPb169VK5cuWs5cKch9nb6dy5s4KDg602Ly8v3XnnnS79vv76a124cEH33XefSz2+vr7q0KFDrjErzHlT2NehuO/z/v7+ateunctrVKlSJT355JM6d+6cvv/+e0l/v4YX3xhcHNddd50iIiKsZV9fX9WrV69Qd8gXZqwud9+l8bsjL2lpafr000/Vvn17NWjQQJLUoUMH1alTRzNnzlRWVlaRtpczE7Ru3VozZ87UxIkTtXHjRp0/f75I27MLwquNnD59WkeOHFF4eHi+ferUqaNVq1apevXqeuSRR1SnTh3VqVPHuvavsMLCwgrdNzQ0NN+2I0eOFGm/RXXkyJE8a80eo5z7r1q1qsuy0+mUJJ05cybffRw7dkzGmCLtpyiqVKnisuzj41Ng+9mzZyVJhw8ftq419Pb2dnls3LhRqampLrUV9DoVZOTIkXruued02223acmSJfrxxx+1adMmNWvWLM9xK+4YZ2VlFbvGMWPG6LXXXtPGjRvVo0cPVa1aVV26dNHmzZsv+dy0tDTddNNN+vHHHzVx4kStXbtWmzZt0oIFC/Ks29/fP9ddwU6n03pdpL+vOwsPD5eXV/5vsYV9/YKCgrRu3Tpdd911evbZZ9WoUSOFh4dr/PjxhfrFlH1z58Vy/nwePnxYx48fl4+PT65akpOTrVoKEhkZab3/ZP/yzw6v2f/BW7Vqlfz8/NS+fXtrv5I0evToXPt9+OGHJcna9+HDh7VkyZJc/Ro1auTSL1txzsPsMclrzHK2Zdd+/fXX56pp/vz5ueopzHlT2Nfhct7nb775Zm3cuFGnT5/WqlWr1LlzZ1WtWlUtW7bUqlWrlJCQoISEhMsOrznHP/t4LzX+UuHG6nL3XRq/O/Iyf/58paWl6c4779Tx48d1/PhxnThxQnfeeacOHDhg/UevsBITE10ywfz58zVo0CB9+OGHateunapUqaL77rtPycnJRdpuWcdsAzaydOlSZWZmXnJ6q5tuukk33XSTMjMztXnzZv3nP//RiBEjFBISov79+xdqX0WZOzavH4rstuwf+Ow3opwX3Bfml2JBqlatqqSkpFzthw4dkiSXT0+KK/umk5LeT1EFBwfL4XDou+++s95IL5bdlv0a5Pc6FXRTmSTNmTNH9913nyZNmuTSnpqa6rYppypXriyHw1HguVSQ8uXLa+TIkRo5cqSOHz+uVatW6dlnn1W3bt104MAB+fv75/vc1atX69ChQ1q7dq31aauky5oPtlq1avr++++VlZWVb4At7OsnSU2aNNG8efNkjNHPP/+smTNn6vnnn5efn5+eeeaZAmu5+GapbDl/PoODg1W1alWtWLEiz20EBgYWuI9sXbp00Zdffql169YpKytLHTt2VGBgoMLDwxUXF6dVq1bppptuso4t++dmzJgx6tevX57bzJ5qLjg4WE2bNtWLL76YZ7+C/lNfFFWrVi1wzLJl1/75558X6i8YhVGU16G47/NdunTRc889p2+//VbffPONxo8fb7WvXLlSUVFR1vKVrDR+d+Tlo48+kiSNGDEiz+kgP/roI3Xr1q1Q2/rpp5+UnJysoUOHWm3BwcGaOnWqpk6dqsTERC1evFjPPPOMUlJS8j2v7IhPXm0iMTFRo0ePVlBQkB588MFCPadcuXJq06aN/vvf/0qS9Sf84v6PMT+7d+/Wjh07XNrmzp2rwMBAtWjRQpKsgPTzzz+79Fu8eHGu7RX2f+fS32+w2eHjYrNnz5a/v79bpmEJCAhQmzZttGDBApe6srKyNGfOHNWsWVP16tW77P0UVe/evWWM0Z9//qlWrVrlemRPndK2bVv5+vrq448/dnn++vXrC/0nvJzhaunSpfrzzz/ddiwBAQFq3bq1FixY4PLpyqlTp7RkyZIibatSpUq644479Mgjj+jo0aPW3eH5nffZ/1HLeYwXz+pRVD169NDZs2cLnMy/sK9fzlqbNWumN954Q5UqVSrUZTmnTp3K9XM2d+5ceXl56R//+IdVy5EjR5SZmZlnLRfPVVzQz+fNN9+sw4cPa+rUqWrbtq0Vtrp06aKFCxdq06ZNLp/o1a9fX3Xr1tWOHTvy3G+rVq2sbfTu3Vu7du1SnTp18uznrvDaoUMHrV692uU/1llZWfrss89c+nXr1k3ly5fX//73v3xrL6qivA7Z8nufz0/r1q1VsWJFTZ06VcnJyYqJiZH092u3bds2ffrpp2rYsOElx9Pdv0dKW2n87sgpPj5eGzZs0O233641a9bkemT/568wf8k7evSohg0bJm9vbz3xxBN59omIiNDw4cMVExNT4pfwlTY+eS2Ddu3aZV3rlJKSou+++04zZsxQuXLltHDhQmuqq7xMmzZNq1evVq9evRQREaGzZ89aU3Bk/9IIDAxUZGSkvvzyS3Xp0kVVqlRRcHDwJT+By094eLhuueUWxcbGKiwsTHPmzFFcXJxefvll6xOv66+/XvXr19fo0aN14cIFVa5cWQsXLrSusbpYkyZNtGDBAr377rtq2bKlvLy88v1FMH78eOtauHHjxqlKlSr6+OOPtXTpUr3yyisKCgoq1jHlNHnyZMXExKhTp04aPXq0fHx89M4772jXrl365JNPivwtZ+5www036N///rfuv/9+bd68Wf/4xz8UEBCgpKQkff/992rSpIkeeughVa5cWaNHj9bEiRP1wAMP6J///KcOHDig2NjYQv1Jvnfv3po5c6YaNGigpk2basuWLXr11VfdPufvCy+8oO7duysmJkajRo1SZmamXn75ZQUEBOjo0aMFPrdPnz7W/MjVqlXT/v37NXXqVEVGRqpu3bqSZIXBN998U4MGDZK3t7fq16+v9u3bq3Llyho2bJjGjx8vb29vffzxx7n+Q1YUd999t2bMmKFhw4Zpz5496tSpk7KysvTjjz8qOjpa/fv3L/Tr99VXX+mdd97RbbfdpmuuuUbGGC1YsEDHjx+3gkdBqlatqoceekiJiYmqV6+eli1bpg8++EAPPfSQdV1g//799fHHH6tnz556/PHH1bp1a3l7e+vgwYNas2aNbr31VvXt29cax3nz5mn+/Pm65ppr5Ovra41t586d5XA4tHLlSuuLUKS/33sGDRpk/fti7733nnr06KFu3bpp8ODBqlGjho4ePar4+Hht3brVCo3PP/+84uLi1L59ez322GOqX7++zp49q3379mnZsmWaNm2aW87JsWPHasmSJerSpYvGjh0rPz8/TZs2zZo2LfuT9Nq1a+v555/X2LFj9ccff6h79+6qXLmyDh8+rJ9++kkBAQEuY1AYhX0dCvM+n59y5cqpQ4cOWrJkiaKioqypEG+44QY5nU598803euyxxy5Za34/T4X9lN7TSut3x8WyP3V96qmn1Lp161zrT506pW+++UZz5szR448/brXv3btXGzduVFZWlvUlBR999JFOnjyp2bNnW5fOnDhxQp06ddKAAQPUoEEDBQYGatOmTVqxYkW+f9mwLY/dKoZcsu/Iz374+PiY6tWrmw4dOphJkyaZlJSUXM/JOQPAhg0bTN++fU1kZKRxOp2matWqpkOHDmbx4sUuz1u1apVp3ry5cTqdRpJ153b29i6+0za/fRnz/+6o/Pzzz02jRo2Mj4+PqV27tpkyZUqu5//222+ma9eupmLFiqZatWrm0UcfNUuXLs11t+nRo0fNHXfcYSpVqmQcDofLPpXHLAk7d+40ffr0MUFBQcbHx8c0a9Ys193y+d21WZi767N99913pnPnziYgIMD4+fmZtm3bmiVLluS5vaLMNpCzpvxmncjvtZk+fbpp06aNVVedOnXMfffdZzZv3mz1ycrKMpMnTza1atUyPj4+pmnTpmbJkiWmQ4cOl5xt4NixY2bo0KGmevXqxt/f39x4443mu+++y/Vcd4zx4sWLTdOmTY2Pj4+JiIgwL730Ur7n3cWzDbz++uumffv2Jjg42Hru0KFDzb59+1yeN2bMGBMeHm68vLxczrv169ebdu3aGX9/f1OtWjXzwAMPmK1bt+aqe9CgQSYgICBX3XnVeObMGTNu3DhTt25d4+PjY6pWrWo6d+7sMk2RMZd+/X799Vdz9913mzp16hg/Pz8TFBRkWrdubWbOnHnJ8ezQoYNp1KiRWbt2rWnVqpVxOp0mLCzMPPvss7nu7j9//rx57bXXTLNmzYyvr6+pUKGCadCggXnwwQfN3r17rX779u0zXbt2NYGBgdZUbxdr3ry5kWR++OEHq+3PP/80kkzVqlWtmTkutmPHDnPnnXea6tWrG29vbxMaGmo6d+5spk2b5tLvr7/+Mo899piJiooy3t7epkqVKqZly5Zm7NixJi0tzRhT8M9gXu8fefnuu+9MmzZtjNPpNKGhoebJJ580L7/8spGUa6qtRYsWmU6dOpmKFSsap9NpIiMjzR133GFWrVpl9SnKeVOY16Gw7/P5efPNN40k869//culPSYmxkjKtZ38fobz+3nK7077/N4zcs42UNixykth921Myf7uyFnHuXPnTPXq1QucFeXChQumZs2apkmTJi77zn6UL1/eVK1a1bRr1848++yzud7fzp49a4YNG2aaNm1qKlasaPz8/Ez9+vXN+PHjrZl/rhQOYy5x+zoAAFe5rl27at++ffrtt988XQpw1eOyAQAALjJy5Eg1b95ctWrV0tGjR/Xxxx8rLi7O+rMvAM8ivAIAcJHMzEyNGzdOycnJcjgcatiwof7v//5PAwcO9HRpACRx2QAAAABsg6myAAAAYBuEVwAAANgG4RUAAAC2ccXfsJWVlaVDhw4pMDDQIxPJAwAAoGDGGJ06dUrh4eH5fq12tis+vB46dEi1atXydBkAAAC4hAMHDlzy2/Ku+PCa/VV1Bw4cUMWKFT1cDQAAAHI6efKkatWqVaivGL7iw2v2pQIVK1YkvAIAAJRhhbnEkxu2AAAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANhGeU8XgKJJTExUamqqS1twcLAiIiI8VBEAAEDpIbzaSGJioqKjo5Wenu7S7u/vr/j4eAIsAAC44hFebSQ1NVXp6emaM/Z5RUdGSZLi9ydo4IvjlJqaSngFAABXPMKrDUVHRqlFvQaeLgMAAKDUccMWAAAAbIPwCgAAANsgvAIAAMA2uOb1CpbXtFoSU2sBAAD7IrxeofKbVktiai0AAGBfhNcrVF7TaklMrQUAAOyN8HqFY1otAABwJfHoDVuxsbFyOBwuj9DQUGu9MUaxsbEKDw+Xn5+fOnbsqN27d3uwYgAAAHiSx2cbaNSokZKSkqzHzp07rXWvvPKKpkyZorffflubNm1SaGioYmJidOrUKQ9WDAAAAE/xeHgtX768QkNDrUe1atUk/f2p69SpUzV27Fj169dPjRs31qxZs5Senq65c+d6uGoAAAB4gsfD6969exUeHq6oqCj1799ff/zxhyQpISFBycnJ6tq1q9XX6XSqQ4cOWr9+fb7by8jI0MmTJ10eAAAAuDJ4NLy2adNGs2fP1tdff60PPvhAycnJat++vY4cOaLk5GRJUkhIiMtzQkJCrHV5mTx5soKCgqxHrVq1SvQYAAAAUHo8Gl579Oih22+/XU2aNNHNN9+spUuXSpJmzZpl9XE4HC7PMcbkarvYmDFjdOLECetx4MCBkikeAAAApc7jlw1cLCAgQE2aNNHevXutWQdyfsqakpKS69PYizmdTlWsWNHlAQAAgCtDmQqvGRkZio+PV1hYmKKiohQaGqq4uDhr/blz57Ru3Tq1b9/eg1UCAADAUzz6JQWjR49Wnz59FBERoZSUFE2cOFEnT57UoEGD5HA4NGLECE2aNEl169ZV3bp1NWnSJPn7+2vAgAGeLBsAAAAe4tHwevDgQd19991KTU1VtWrV1LZtW23cuFGRkZGSpKeeekpnzpzRww8/rGPHjqlNmzZauXKlAgMDPVk2AAAAPMSj4XXevHkFrnc4HIqNjVVsbGzpFAQAAIAyrUxd8woAAAAUhPAKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALCN8p4uAPaUmJio1NRUl7bg4GBFRER4qCIAAHA1ILyiyBITExUdHa309HSXdn9/f8XHxxNgAQBAiSG8oshSU1OVnp6uOWOfV3RklCQpfn+CBr44TqmpqYRXAABQYgivKLboyCi1qNfA02UAAICrSJm5YWvy5MlyOBwaMWKE1WaMUWxsrMLDw+Xn56eOHTtq9+7dnisSAAAAHlUmwuumTZv0/vvvq2nTpi7tr7zyiqZMmaK3335bmzZtUmhoqGJiYnTq1CkPVQoAAABP8nh4TUtL0z333KMPPvhAlStXttqNMZo6darGjh2rfv36qXHjxpo1a5bS09M1d+5cD1YMAAAAT/F4eH3kkUfUq1cv3XzzzS7tCQkJSk5OVteuXa02p9OpDh06aP369fluLyMjQydPnnR5AAAA4Mrg0Ru25s2bp61bt2rTpk251iUnJ0uSQkJCXNpDQkK0f//+fLc5efJkTZgwwb2FAgAAoEzw2CevBw4c0OOPP645c+bI19c3334Oh8Nl2RiTq+1iY8aM0YkTJ6zHgQMH3FYzAAAAPMtjn7xu2bJFKSkpatmypdWWmZmpb7/9Vm+//bb27Nkj6e9PYMPCwqw+KSkpuT6NvZjT6ZTT6Sy5wgEAAOAxHvvktUuXLtq5c6e2b99uPVq1aqV77rlH27dv1zXXXKPQ0FDFxcVZzzl37pzWrVun9u3be6psAAAAeJDHPnkNDAxU48aNXdoCAgJUtWpVq33EiBGaNGmS6tatq7p162rSpEny9/fXgAEDPFEyAAAAPKxMf8PWU089pTNnzujhhx/WsWPH1KZNG61cuVKBgYGeLg0AAAAeUKbC69q1a12WHQ6HYmNjFRsb65F6riaJiYlKTU3N1R4cHKyIiAgPVAQAAJBbmQqv8IzExERFR0crPT091zp/f3/Fx8cTYAEAQJlAeIVSU1OVnp6uOWOfV3RklNUevz9BA18cp9TUVMIrAAAoEwivsERHRqlFvQaeLgMAACBfHv96WAAAAKCwCK8AAACwDcIrAAAAbIPwCgAAANsgvAIAAMA2CK8AAACwDcIrAAAAbIPwCgAAANsgvAIAAMA2CK8AAACwDb4e9goRHx9f4DIAAMCVgPBqc0lHUuXl5aWBAwd6uhQAAIASR3i1ueNpacrKytKcsc8rOjLKal+28Qc9N32aBysDAABwP8LrFSI6Mkot6jWwluP37/NcMQAAACWEG7YAAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtlPd0AfCM+Pj4PP8NAABQlhFerzJJR1Ll5eWlgQMHeroUAACAIiO8XmWOp6UpKytLc8Y+r+jIKEnSso0/6Lnp0zxcGQAAwKURXq9S0ZFRalGvgSQpfv8+zxYDAABQSNywBQAAANsgvAIAAMA2CK8AAACwDa55xSXlnEqLqbUAAICnEF6RL6bVAgAAZU2xwmtCQoKioqLcXQvKmLym1ZKYWgsAAHhOsa55vfbaa9WpUyfNmTNHZ8+edXdNKGOyp9XKfkSF1fB0SQAA4CpVrPC6Y8cONW/eXKNGjVJoaKgefPBB/fTTT+6uDQAAAHBRrPDauHFjTZkyRX/++admzJih5ORk3XjjjWrUqJGmTJmiv/76y911AgAAAJc3VVb58uXVt29fffrpp3r55Zf1v//9T6NHj1bNmjV13333KSkpyV11AgAAAJc328DmzZs1ffp0zZs3TwEBARo9erSGDh2qQ4cOady4cbr11lu5nOAqk9c0WsHBwYqIiPBANQAA4EpTrPA6ZcoUzZgxQ3v27FHPnj01e/Zs9ezZU15ef3+QGxUVpffee08NGjRwa7EouwqaVsvf31/x8fEEWAAAcNmKFV7fffddDRkyRPfff79CQ0Pz7BMREaGPPvrosoqDfeQ3rVb8/gQNfHGcUlNTCa8AAOCyFSu87t2795J9fHx8NGjQoOJsHjaWPa0WAABASSjWDVszZszQZ599lqv9s88+06xZswq9nXfffVdNmzZVxYoVVbFiRbVr107Lly+31htjFBsbq/DwcPn5+aljx47avXt3cUoGAADAFaBY4fWll15ScHBwrvbq1atr0qRJhd5OzZo19dJLL2nz5s3avHmzOnfurFtvvdUKqK+88oqmTJmit99+W5s2bVJoaKhiYmJ06tSp4pQNAAAAmytWeN2/f3+eXw8bGRmpxMTEQm+nT58+6tmzp+rVq6d69erpxRdfVIUKFbRx40YZYzR16lSNHTtW/fr1U+PGjTVr1iylp6dr7ty5xSkbAAAANles8Fq9enX9/PPPudp37NihqlWrFquQzMxMzZs3T6dPn1a7du2UkJCg5ORkde3a1erjdDrVoUMHrV+/Pt/tZGRk6OTJky4PAAAAXBmKFV779++vxx57TGvWrFFmZqYyMzO1evVqPf744+rfv3+RtrVz505VqFBBTqdTw4YN08KFC9WwYUMlJydLkkJCQlz6h4SEWOvyMnnyZAUFBVmPWrVqFf0AAQAAUCYVa7aBiRMnav/+/erSpYvKl/97E1lZWbrvvvuKdM2rJNWvX1/bt2/X8ePH9cUXX2jQoEFat26dtd7hcLj0N8bkarvYmDFjNHLkSGv55MmTBFgAAIArRLHCq4+Pj+bPn68XXnhBO3bskJ+fn5o0aaLIyMhibevaa6+VJLVq1UqbNm3Sm2++qaefflqSlJycrLCwMKt/SkpKrk9jL+Z0OuV0OotcBwAAAMq+y/p62OwbrdzJGKOMjAxFRUUpNDRUcXFxat68uSTp3LlzWrdunV5++WW37hMAAAD2UKzwmpmZqZkzZ+qbb75RSkqKsrKyXNavXr26UNt59tln1aNHD9WqVUunTp3SvHnztHbtWq1YsUIOh0MjRozQpEmTVLduXdWtW1eTJk2Sv7+/BgwYUJyyAQAAYHPFCq+PP/64Zs6cqV69eqlx48YFXoNakMOHD+vee+9VUlKSgoKC1LRpU61YsUIxMTGSpKeeekpnzpzRww8/rGPHjqlNmzZauXKlAgMDi7U/AAAA2Fuxwuu8efP06aefqmfPnpe1848++qjA9Q6HQ7GxsYqNjb2s/QAAAODKUKypsi6+yQoAAAAoLcUKr6NGjdKbb74pY4y76wEAAADyVazLBr7//nutWbNGy5cvV6NGjeTt7e2yfsGCBW4pDgAAALhYscJrpUqV1LdvX3fXAgAAABSoWOF1xowZ7q4DAAAAuKRiXfMqSRcuXNCqVav03nvv6dSpU5KkQ4cOKS0tzW3FAQAAABcr1iev+/fvV/fu3ZWYmKiMjAzFxMQoMDBQr7zyis6ePatp06a5u04AAACgeJ+8Pv7442rVqpWOHTsmPz8/q71v37765ptv3FYcAAAAcLFizzbwww8/yMfHx6U9MjJSf/75p1sKAwAAAHIq1ievWVlZyszMzNV+8OBBvroVAAAAJaZY4TUmJkZTp061lh0Oh9LS0jR+/PjL/spYAAAAID/FumzgjTfeUKdOndSwYUOdPXtWAwYM0N69exUcHKxPPvnE3TUCAAAAkooZXsPDw7V9+3Z98skn2rp1q7KysjR06FDdc889LjdwAQAAAO5UrPAqSX5+fhoyZIiGDBniznpwlUhMTFRqamqu9uDgYEVERHigIgAAYAfFCq+zZ88ucP19991XrGJwdUhMTFR0dLTS09NzrfP391d8fDwBFgAA5KlY4fXxxx93WT5//rzS09Pl4+Mjf39/wisKlJqaqvT0dM0Z+7yiI6Os9vj9CRr44jilpqYSXgEAQJ6KFV6PHTuWq23v3r166KGH9OSTT152Ubg6REdGqUW9Bp4uAwAA2EixpsrKS926dfXSSy/l+lQWAAAAcBe3hVdJKleunA4dOuTOTQIAAACWYl02sHjxYpdlY4ySkpL09ttv64YbbnBLYQAAAEBOxQqvt912m8uyw+FQtWrV1LlzZ73++uvuqAsAAADIpVjhNSsry911AAAAAJfk1mteAQAAgJJUrE9eR44cWei+U6ZMKc4uAAAAgFyKFV63bdumrVu36sKFC6pfv74k6bffflO5cuXUokULq5/D4XBPlQAAAICKGV779OmjwMBAzZo1S5UrV5b09xcX3H///brppps0atQotxYJAAAASMW85vX111/X5MmTreAqSZUrV9bEiROZbQAAAAAlpljh9eTJkzp8+HCu9pSUFJ06deqyiwIAAADyUqzw2rdvX91///36/PPPdfDgQR08eFCff/65hg4dqn79+rm7RgAAAEBSMa95nTZtmkaPHq2BAwfq/Pnzf2+ofHkNHTpUr776qlsLBAAAALIVK7z6+/vrnXfe0auvvqr//e9/Msbo2muvVUBAgLvrAwAAACyX9SUFSUlJSkpKUr169RQQECBjjLvqAgAAAHIpVng9cuSIunTponr16qlnz55KSkqSJD3wwANMkwUAAIASU6zw+sQTT8jb21uJiYny9/e32u+66y6tWLHCbcUBAAAAFyvWNa8rV67U119/rZo1a7q0161bV/v373dLYQAAAEBOxfrk9fTp0y6fuGZLTU2V0+m87KIAAACAvBQrvP7jH//Q7NmzrWWHw6GsrCy9+uqr6tSpk9uKAwAAAC5WrMsGXn31VXXs2FGbN2/WuXPn9NRTT2n37t06evSofvjhB3fXCAAAAEgq5ievDRs21M8//6zWrVsrJiZGp0+fVr9+/bRt2zbVqVPH3TUCAAAAkorxyev58+fVtWtXvffee5owYUJJ1AQAAADkqcifvHp7e2vXrl1yOBwlUQ8AAACQr2JdNnDffffpo48+cnctAAAAQIGKdcPWuXPn9OGHHyouLk6tWrVSQECAy/opU6a4pTgAAADgYkUKr3/88Ydq166tXbt2qUWLFpKk3377zaUPlxMAAACgpBQpvNatW1dJSUlas2aNpL+/Dvatt95SSEhIiRQHAAAAXKxI17waY1yWly9frtOnT7u1IAAAACA/xbphK1vOMAsAAACUpCKFV4fDkeuaVq5xBQAAQGkp0jWvxhgNHjxYTqdTknT27FkNGzYs12wDCxYscF+FAAAAwP9fkcLroEGDXJYHDhzo1mKuRomJiUpNTc3VnpGRYf0nIVt8fHxplQUAAFAmFSm8zpgxo6TquColJiYqOjpa6enpudaV8/JSZlaWB6oCAAAou4r1JQVwj9TUVKWnp2vO2OcVHRlltS/b+IOemz4t33YAAICrFeG1DIiOjFKLeg2s5fj9+wpsBwAAuFpd1lRZAAAAQGkivAIAAMA2CK8AAACwDcIrAAAAbIPwCgAAANsgvAIAAMA2CK8AAACwDY+G18mTJ+v6669XYGCgqlevrttuu0179uxx6WOMUWxsrMLDw+Xn56eOHTtq9+7dHqoYAAAAnuTR8Lpu3To98sgj2rhxo+Li4nThwgV17dpVp0+ftvq88sormjJlit5++21t2rRJoaGhiomJ0alTpzxYOQAAADzBo9+wtWLFCpflGTNmqHr16tqyZYv+8Y9/yBijqVOnauzYserXr58kadasWQoJCdHcuXP14IMPeqJsAAAAeEiZ+nrYEydOSJKqVKkiSUpISFBycrK6du1q9XE6nerQoYPWr1+fZ3jNyMhQRkaGtXzy5MkSrhruFh8fn6stODhYERERHqgGAACUJWUmvBpjNHLkSN14441q3LixJCk5OVmSFBIS4tI3JCRE+/fvz3M7kydP1oQJE0q2WJSIpCOp8vLy0sCBA3Ot8/f3V3x8PAEWAICrXJkJr8OHD9fPP/+s77//Ptc6h8PhsmyMydWWbcyYMRo5cqS1fPLkSdWqVcu9xaJEHE9LU1ZWluaMfV7RkVFWe/z+BA18cZxSU1MJrwAAXOXKRHh99NFHtXjxYn377beqWbOm1R4aGirp709gw8LCrPaUlJRcn8ZmczqdcjqdJVswSlR0ZJRa1Gvg6TIAAEAZ5NHZBowxGj58uBYsWKDVq1crKirKZX1UVJRCQ0MVFxdntZ07d07r1q1T+/btS7tcAAAAeJhHP3l95JFHNHfuXH355ZcKDAy0rnENCgqSn5+fHA6HRowYoUmTJqlu3bqqW7euJk2aJH9/fw0YMMCTpQMAAMADPBpe3333XUlSx44dXdpnzJihwYMHS5KeeuopnTlzRg8//LCOHTumNm3aaOXKlQoMDCzlagEAAOBpHg2vxphL9nE4HIqNjVVsbGzJFwQAAIAyzaPXvAIAAABFQXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2Ud7TBeDqEB8fn+e/i7sNSQoODlZERMRl1QUAAOyF8IoSlXQkVV5eXho4cKDbt+Hv76/4+HgCLAAAVxHCK0rU8bQ0ZWVlac7Y5xUdGSVJWrbxBz03fdplbSN+f4IGvjhOqamphFcAAK4ihFeUiujIKLWo10CSFL9/32VvAwAAXJ24YQsAAAC2QXgFAACAbRBeAQAAYBtc84orTmJiolJTU3O1M7UWAAD2R3jFFSUxMVHR0dFKT0/PtY6ptQAAsD/CK64oqampSk9Pd5lWS2JqLQAArhSEV1yRmFYLAIArEzdsAQAAwDYIrwAAALANwisAAABsg2tegSLIaxoupuACAKD0EF6BQspvGi6m4AIAoPQQXoFCymsaLqbgAgCgdBFegSJiGi4AADyHG7YAAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtlPd0AVeTxMREpaamWsvx8fEerAYAAMB+CK+lJDExUdHR0UpPT/d0KQAAALZFeC0lqampSk9P15yxzys6MkqStGzjD3pu+jQPVwYAAGAfhNdSFh0ZpRb1GkiS4vfv82wxAAAANuPRG7a+/fZb9enTR+Hh4XI4HFq0aJHLemOMYmNjFR4eLj8/P3Xs2FG7d+/2TLEAAADwOI+G19OnT6tZs2Z6++2381z/yiuvaMqUKXr77be1adMmhYaGKiYmRqdOnSrlSgEAAFAWePSygR49eqhHjx55rjPGaOrUqRo7dqz69esnSZo1a5ZCQkI0d+5cPfjgg6VZKgAAAMqAMjvPa0JCgpKTk9W1a1erzel0qkOHDlq/fn2+z8vIyNDJkyddHgAAALgylNnwmpycLEkKCQlxaQ8JCbHW5WXy5MkKCgqyHrVq1SrROgEAAFB6ymx4zeZwOFyWjTG52i42ZswYnThxwnocOHCgpEsEAABAKSmzU2WFhoZK+vsT2LCwMKs9JSUl16exF3M6nXI6nSVeHwAAAEpfmf3kNSoqSqGhoYqLi7Pazp07p3Xr1ql9+/YerAwAAACe4tFPXtPS0vT7779bywkJCdq+fbuqVKmiiIgIjRgxQpMmTVLdunVVt25dTZo0Sf7+/howYIAHqwYAAICneDS8bt68WZ06dbKWR44cKUkaNGiQZs6cqaeeekpnzpzRww8/rGPHjqlNmzZauXKlAgMDPVUyAAAAPMij4bVjx44yxuS73uFwKDY2VrGxsaVXFAAAAMqsMnvNKwAAAJAT4RUAAAC2QXgFAACAbRBeAQAAYBuEVwAAANgG4RUAAAC2UWa/HhYoCfHx8S7LGRkZeX6dcHBwsCIiIkqrLAAAUEiEV1wVko6kysvLSwMHDnRpL+flpcysrFz9/f39FR8fT4AFAKCMIbziqnA8LU1ZWVmaM/Z5RUdGSZKWbfxBz02f5tImSfH7EzTwxXFKTU0lvAIAUMYQXnFViY6MUot6DSRJ8fv35WoDAABlGzdsAQAAwDYIrwAAALANwisAAABsg2tegXzknFYr53JxJCYmKjU1NVc7U3MBAFA4hFcgh/ym1bpciYmJio6OVnp6eq51TM0FAEDhEF6BHPKaVkv6f1NrFVdqaqrS09OZmgsAgMtAeAXykXMKreyptdy9XQAAUHjcsAUAAADbILwCAADANgivAAAAsA2ueYWtlcR0Vu6Sc1qsslQbAAB2RXiFLZXUdFbuUtC0WAAAoPgIr7ClkprOyl3ymharrNQGAICdEV5hayU1nZW7XFxfWasNAAA74oYtAAAA2AbhFQAAALZBeAUAAIBtcM0rUEbkNZVWcHCwIiIiCvX8nFNzFWcbAACUdYRXwMMKmvbL399f8fHxlwyfBU3NVdhtAABgB4RXwMPym/Yrfn+CBr44TqmpqZcMnnlNzVXUbQAAYAeEV6CMyDntl6e2AQBAWcYNWwAAALANwisAAABsg/AKAAAA2yC8AgAAwDa4YasE5DXfZl5zeAIAAKBoCK9uVtB8mwAAALg8hFc3y2++zWUbf9Bz06d5sDIAAAD7I7yWkJzzbcbv3+e5YgAAAK4Q3LAFAAAA2yC8AgAAwDYIrwAAALANrnkF3CDnVGhXw9RoeU0JJ0nBwcGKiIiw/f4AAGUT4RW4DElHUuXl5aWBAwd6upRSVdCUcP7+/oqPj3droCzt/QEAyi7CK3AZjqelKSsr66qbGi2/KeHi9ydo4IvjlJqa6tYwWdr7AwCUXYRXwA2u1qnRch73lbY/AEDZww1bAAAAsA3CKwAAAGyD8AoAAADb4JpXoIzLOe1WRkaGnE5ngX0upajTTuXsf6n95VzPdFYAAHchvAJlVH7TcJXz8lJmVlaxt1vUaacK6l/YmpnOCgDgLoRXoIzKaxqu7Cm4LmdqrqJOO5VX//z2l1fNTGcFAHAnwitQxl08PVT2FFzumJqrqNNO5VWHu7YNAEBhccMWAAAAbIPwCgAAANsgvAIAAMA2uOYVgIuc01wVdRoud7jcqbmk0p+eK6/px4paQ1G2kd90Z3lNpeaOsSjq9Gqwt7LyepeVOq50dhtnwisASflPc1Xa3DE1l1S603PlV3NRaijKNgoao7ymUrvcsSjq9Gqwt7LyepeVOq50dhxnwisASXlPcyUVbRoud7jcqbmk0p+eK6+ai1pDUbaR33RneU2l5o6xKOr0arC3svJ6l5U6rnR2HGfCKwAX7piGy9112GVqLnfUUZRt5PdaldR4lJVxRukoK693WanjSmencbbFDVvvvPOOoqKi5Ovrq5YtW+q7777zdEkAAADwgDIfXufPn68RI0Zo7Nix2rZtm2666Sb16NFDiYmJni4NAAAApazMh9cpU6Zo6NCheuCBBxQdHa2pU6eqVq1aevfddz1dGgAAAEpZmb7m9dy5c9qyZYueeeYZl/auXbtq/fr1eT4nIyNDGRkZ1vKJEyckSSdPniy5Qi+SlpYmSdryW7zSzpyx2uP3J+Rqz6utqO123IYda7brce85sO/v9i1brHNzz549pVpHXjVk8/LyUlaOO+Pzqs8dx53f/oraXtia3XHc+W2jKK9hUevIqz2//ZX2OJf1bdix5rzay8rrXVbqcEffsryNS41zWlpaqWSo7H0YYy7d2ZRhf/75p5FkfvjhB5f2F1980dSrVy/P54wfP95I4sGDBw8ePHjw4GGzx4EDBy6ZD8v0J6/ZHA6Hy7IxJldbtjFjxmjkyJHWclZWlo4ePaqqVavm+5yCnDx5UrVq1dKBAwdUsWLFIj8fRcN4ly7Gu3Qx3qWL8S5djHfputLG2xijU6dOKTw8/JJ9y3R4DQ4OVrly5ZScnOzSnpKSopCQkDyf43Q6c327TKVKlS67looVK14RJ4ddMN6li/EuXYx36WK8SxfjXbqupPEOCgoqVL8yfcOWj4+PWrZsqbi4OJf2uLg4tW/f3kNVAQAAwFPK9CevkjRy5Ejde++9atWqldq1a6f3339fiYmJGjZsmKdLAwAAQCkr8+H1rrvu0pEjR/T8888rKSlJjRs31rJlyxQZGVkq+3c6nRo/fnyuSxFQMhjv0sV4ly7Gu3Qx3qWL8S5dV/N4O4wpzJwEAAAAgOeV6WteAQAAgIsRXgEAAGAbhFcAAADYBuEVAAAAtkF4vYR33nlHUVFR8vX1VcuWLfXdd995uqQyLzY2Vg6Hw+URGhpqrTfGKDY2VuHh4fLz81PHjh21e/dul21kZGTo0UcfVXBwsAICAnTLLbfo4MGDLn2OHTume++9V0FBQQoKCtK9996r48ePl8YhetS3336rPn36KDw8XA6HQ4sWLXJZX5rjm5iYqD59+iggIEDBwcF67LHHdO7cuZI4bI+51HgPHjw41/netm1blz6Md+FMnjxZ119/vQIDA1W9enXddttt1veuZ+P8dp/CjDfnt/u8++67atq0qfWlAu3atdPy5cut9ZzbRXDJL5C9is2bN894e3ubDz74wPzyyy/m8ccfNwEBAWb//v2eLq1MGz9+vGnUqJFJSkqyHikpKdb6l156yQQGBpovvvjC7Ny509x1110mLCzMnDx50uozbNgwU6NGDRMXF2e2bt1qOnXqZJo1a2YuXLhg9enevbtp3LixWb9+vVm/fr1p3Lix6d27d6keqycsW7bMjB071nzxxRdGklm4cKHL+tIa3wsXLpjGjRubTp06ma1bt5q4uDgTHh5uhg8fXuJjUJouNd6DBg0y3bt3dznfjxw54tKH8S6cbt26mRkzZphdu3aZ7du3m169epmIiAiTlpZm9eH8dp/CjDfnt/ssXrzYLF261OzZs8fs2bPHPPvss8bb29vs2rXLGMO5XRSE1wK0bt3aDBs2zKWtQYMG5plnnvFQRfYwfvx406xZszzXZWVlmdDQUPPSSy9ZbWfPnjVBQUFm2rRpxhhjjh8/bry9vc28efOsPn/++afx8vIyK1asMMYY88svvxhJZuPGjVafDRs2GEnm119/LYGjKptyhqnSHN9ly5YZLy8v8+eff1p9PvnkE+N0Os2JEydK5Hg9Lb/weuutt+b7HMa7+FJSUowks27dOmMM53dJyznexnB+l7TKlSubDz/8kHO7iLhsIB/nzp3Tli1b1LVrV5f2rl27av369R6qyj727t2r8PBwRUVFqX///vrjjz8kSQkJCUpOTnYZV6fTqQ4dOljjumXLFp0/f96lT3h4uBo3bmz12bBhg4KCgtSmTRurT9u2bRUUFHRVvz6lOb4bNmxQ48aNFR4ebvXp1q2bMjIytGXLlhI9zrJm7dq1ql69uurVq6d//etfSklJsdYx3sV34sQJSVKVKlUkcX6XtJzjnY3z2/0yMzM1b948nT59Wu3atePcLiLCaz5SU1OVmZmpkJAQl/aQkBAlJyd7qCp7aNOmjWbPnq2vv/5aH3zwgZKTk9W+fXsdOXLEGruCxjU5OVk+Pj6qXLlygX2qV6+ea9/Vq1e/ql+f0hzf5OTkXPupXLmyfHx8rqrXoEePHvr444+1evVqvf7669q0aZM6d+6sjIwMSYx3cRljNHLkSN14441q3LixJM7vkpTXeEuc3+62c+dOVahQQU6nU8OGDdPChQvVsGFDzu0iKvNfD+tpDofDZdkYk6sNrnr06GH9u0mTJmrXrp3q1KmjWbNmWRf6F2dcc/bJqz+vz99Ka3x5Df7+CutsjRs3VqtWrRQZGamlS5eqX79++T6P8S7Y8OHD9fPPP+v777/PtY7z2/3yG2/Ob/eqX7++tm/fruPHj+uLL77QoEGDtG7dOms953bh8MlrPoKDg1WuXLlc/wtJSUnJ9T8WFCwgIEBNmjTR3r17rVkHChrX0NBQnTt3TseOHSuwz+HDh3Pt66+//rqqX5/SHN/Q0NBc+zl27JjOnz9/Vb8GYWFhioyM1N69eyUx3sXx6KOPavHixVqzZo1q1qxptXN+l4z8xjsvnN+Xx8fHR9dee61atWqlyZMnq1mzZnrzzTc5t4uI8JoPHx8ftWzZUnFxcS7tcXFxat++vYeqsqeMjAzFx8crLCxMUVFRCg0NdRnXc+fOad26dda4tmzZUt7e3i59kpKStGvXLqtPu3btdOLECf30009Wnx9//FEnTpy4ql+f0hzfdu3aadeuXUpKSrL6rFy5Uk6nUy1btizR4yzLjhw5ogMHDigsLEwS410UxhgNHz5cCxYs0OrVqxUVFeWynvPbvS413nnh/HYvY4wyMjI4t4uqlG4Ms6XsqbI++ugj88svv5gRI0aYgIAAs2/fPk+XVqaNGjXKrF271vzxxx9m48aNpnfv3iYwMNAat5deeskEBQWZBQsWmJ07d5q77747z+lAatasaVatWmW2bt1qOnfunOd0IE2bNjUbNmwwGzZsME2aNLkqpso6deqU2bZtm9m2bZuRZKZMmWK2bdtmTeFWWuObPd1Kly5dzNatW82qVatMzZo1bTXdSmEUNN6nTp0yo0aNMuvXrzcJCQlmzZo1pl27dqZGjRqMdzE89NBDJigoyKxdu9Zlaqb09HSrD+e3+1xqvDm/3WvMmDHm22+/NQkJCebnn382zz77rPHy8jIrV640xnBuFwXh9RL++9//msjISOPj42NatGjhMoUI8pY9N523t7cJDw83/fr1M7t377bWZ2VlmfHjx5vQ0FDjdDrNP/7xD7Nz506XbZw5c8YMHz7cVKlSxfj5+ZnevXubxMRElz5Hjhwx99xzjwkMDDSBgYHmnnvuMceOHSuNQ/SoNWvWGEm5HoMGDTLGlO747t+/3/Tq1cv4+fmZKlWqmOHDh5uzZ8+W5OGXuoLGOz093XTt2tVUq1bNeHt7m4iICDNo0KBcY8l4F05e4yzJzJgxw+rD+e0+lxpvzm/3GjJkiJUnqlWrZrp06WIFV2M4t4vCYYwxpfc5LwAAAFB8XPMKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAAAA2yC8AgAAwDYIrwAAALANwisAAABsg/AKAMWwdu1aORwOHT9+XJI0c+ZMVapUyaM15eRwOLRo0SJPlwEAbkV4BXDFGTx4sBwOh4YNG5Zr3cMPPyyHw6HBgwe7dZ933XWXfvvtN7duMy/Zx+ZwOOTt7a2QkBDFxMRo+vTpysrKcumblJSkHj16FGq7BF0AdkF4BXBFqlWrlubNm6czZ85YbWfPntUnn3yiiIgIt+/Pz89P1atXd/t289K9e3clJSVp3759Wr58uTp16qTHH39cvXv31oULF6x+oaGhcjqdpVITAJQWwiuAK1KLFi0UERGhBQsWWG0LFixQrVq11Lx5c5e+xhi98soruuaaa+Tn56dmzZrp888/d+mzbNky1atXT35+furUqZP27dvnsj7nZQP/+9//dOuttyokJEQVKlTQ9ddfr1WrVrk8p3bt2po0aZKGDBmiwMBARURE6P3337/ksTmdToWGhqpGjRpq0aKFnn32WX355Zdavny5Zs6cafW7+NPUc+fOafjw4QoLC5Ovr69q166tyZMnW3VIUt++feVwOKxldx3DwYMH1b9/f1WpUkUBAQFq1aqVfvzxR2v9kiVL1LJlS/n6+uqaa67RhAkTXEI4AFyM8ArginX//fdrxowZ1vL06dM1ZMiQXP3+v//v/9OMGTP07rvvavfu3XriiSc0cOBArVu3TpJ04MAB9evXTz179tT27dv1wAMP6Jlnnilw32lpaerZs6dWrVqlbdu2qVu3burTp48SExNd+r3++utq1aqVtm3bpocfflgPPfSQfv311yIfa+fOndWsWTOXsH6xt956S4sXL9ann36qPXv2aM6cOVZI3bRpkyRpxowZSkpKspbdcQxpaWnq0KGDDh06pMWLF2vHjh166qmnrEscvv76aw0cOFCPPfaYfvnlF7333nuaOXOmXnzxxSKPAYCrhAGAK8ygQYPMrbfeav766y/jdDpNQkKC2bdvn/H19TV//fWXufXWW82gQYOMMcakpaUZX19fs379epdtDB061Nx9993GGGPGjBljoqOjTVZWlrX+6aefNpLMsWPHjDHGzJgxwwQFBRVYV8OGDc1//vMfazkyMtIMHDjQWs7KyjLVq1c377777iWPLS933XWXiY6OtpYlmYULFxpjjHn00UdN586dXY7hYhf3decxvPfeeyYwMNAcOXIkz+3ddNNNZtKkSS5t//d//2fCwsIuWQuAq1N5D2dnACgxwcHB6tWrl2bNmiVjjHr16qXg4GCXPr/88ovOnj2rmJgYl/Zz585ZlxfEx8erbdu2cjgc1vp27doVuO/Tp09rwoQJ+uqrr3To0CFduHBBZ86cyfWpZdOmTa1/OxwOhYaGKiUlpVjHa4xxqfFigwcPVkxMjOrXr6/u3burd+/e6tq1a4kfw/bt29W8eXNVqVIlz31s2bJFmzZtcvmkNTMzU2fPnlV6err8/f0LdewArh6EVwBXtCFDhmj48OGSpP/+97+51mf/+Xrp0qWqUaOGy7rsm52MMUXe75NPPqmvv/5ar732mq699lr5+fnpjjvu0Llz51z6eXt7uyw7HI5cswYUVnx8vKKiovJc16JFCyUkJGj58uVatWqV7rzzTt188825ru119zH4+fkVWHNWVpYmTJigfv365Vrn6+tb4HMBXJ0IrwCuaN27d7fCVrdu3XKtb9iwoZxOpxITE9WhQ4c8t9GwYcNc00ht3LixwP1+9913Gjx4sPr27Svp72s/c97k5U6rV6/Wzp079cQTT+Tbp2LFirrrrrt011136Y477lD37t119OhRValSRd7e3srMzHT7MTRt2lQffvihtZ+cWrRooT179ujaa68t0nYBXL0IrwCuaOXKlVN8fLz175wCAwM1evRoPfHEE8rKytKNN96okydPav369apQoYIGDRqkYcOG6fXXX9fIkSP14IMPasuWLS539efl2muv1YIFC9SnTx85HA4999xzxf5ENaeMjAwlJycrMzNThw8f1ooVKzR58mT17t1b9913X57PeeONNxQWFqbrrrtOXl5e+uyzzxQaGmrNkFC7dm198803uuGGG+R0OlW5cmW3HMPdd9+tSZMm6bbbbtPkyZMVFhambdu2KTw8XO3atdO4cePUu3dv1apVS//85z/l5eWln3/+WTt37tTEiRMvd6gAXIGYbQDAFa9ixYqqWLFivutfeOEFjRs3TpMnT1Z0dLS6deumJUuWWH+Cj4iI0BdffKElS5aoWbNmmjZtmiZNmlTgPt944w1VrlxZ7du3V58+fdStWze1aNHCLcezYsUKhYWFqXbt2urevbvWrFmjt956S19++WWeAV2SKlSooJdfflmtWrXS9ddfr3379mnZsmXy8vr718Drr7+uuLg4l6nE3HEMPj4+WrlypapXr66ePXuqSZMmeumll6w6u3Xrpq+++kpxcXG6/vrr1bZtW02ZMkWRkZGXMUIArmQOU5yLuQAAAAAP4JNXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBtEF4BAABgG4RXAAAA2AbhFQAAALZBeAUAAIBt/P8AcY74YgI2Q+AAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 800x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plotting distribution of median distances between genes per nontad\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Initialize a list to store the values from column 4\n",
    "distances = []\n",
    "\n",
    "# Read the file and extract column 4 values\n",
    "with open('nontad_exp_dist.bed', 'r') as file:\n",
    "    for line in file:\n",
    "        parts = line.strip().split('\\t')\n",
    "        if len(parts) >= 4:\n",
    "            distances.append(float(parts[3]))\n",
    "\n",
    "# Plot the histogram\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.hist(distances, bins=100, color='pink', edgecolor='black')\n",
    "plt.xlabel('Median Distance')\n",
    "plt.ylabel('Frequency')\n",
    "plt.title('Distribution of median distances between genes within nonTADs')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a7f5bbfb-23ec-4688-bd9c-207f268dc3c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data has been split into quartiles and saved to separate files: tad_exp_dist_Q1.bed, tad_exp_dist_Q2.bed, tad_exp_dist_Q3.bed, tad_exp_dist_Q4.bed.\n"
     ]
    }
   ],
   "source": [
    "#split TADs into 4 quartiles of median distances\n",
    "import pandas as pd\n",
    "\n",
    "# Load the data into a DataFrame\n",
    "data = pd.read_csv('nontad_exp_dist.bed', sep='\\t', header=None)\n",
    "data.columns = ['chr', 'tad_start', 'tad_end', 'median_distance']\n",
    "\n",
    "# Calculate the quartiles for the 'median_distance' column\n",
    "quartiles = data['median_distance'].quantile([0.25, 0.5, 0.75])\n",
    "\n",
    "# Assign each row to a quartile\n",
    "data['quartile'] = pd.cut(data['median_distance'], \n",
    "                          bins=[-float('inf'), quartiles[0.25], quartiles[0.5], quartiles[0.75], float('inf')],\n",
    "                          labels=['Q1', 'Q2', 'Q3', 'Q4'])\n",
    "\n",
    "# Write each quartile to a separate file\n",
    "for q in ['Q1', 'Q2', 'Q3', 'Q4']:\n",
    "    quartile_data = data[data['quartile'] == q]\n",
    "    quartile_data.drop(columns=['quartile']).to_csv(f'nontad_exp_dist_{q}.bed', sep='\\t', index=False, header=False)\n",
    "\n",
    "print(\"Data has been split into quartiles and saved to separate files: tad_exp_dist_Q1.bed, tad_exp_dist_Q2.bed, tad_exp_dist_Q3.bed, tad_exp_dist_Q4.bed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4e10149b-a87a-499e-b04a-6b16fb394093",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "143 nontad_exp_dist_Q1.bed\n",
      "142 nontad_exp_dist_Q2.bed\n",
      "142 nontad_exp_dist_Q3.bed\n",
      "142 nontad_exp_dist_Q4.bed\n"
     ]
    }
   ],
   "source": [
    "!wc -l nontad_exp_dist_Q1.bed\n",
    "!wc -l nontad_exp_dist_Q2.bed\n",
    "!wc -l nontad_exp_dist_Q3.bed\n",
    "!wc -l nontad_exp_dist_Q4.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2781804f-a91e-4df2-a344-e1ecf28e9c92",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
